// @audiotool/nexus/dist/api.js
var S = Object.defineProperty;
var T = (s, n, e) => n in s ? S(s, n, { enumerable: !0, configurable: !0, writable: !0, value: e }) : s[n] = e;
var a = (s, n, e) => T(s, typeof n != "symbol" ? n + "" : n, e);
import { P as p, S as d } from "./user_service_connect-BZIXD0EN.js";
import { A, a as O, aM as g, aN as j, ak as f, al as G, $ as L, a0 as v, au as F, C as x, m as K, M as H, N as Y, w, x as B, ag as J, ai as h, D as V, p as M, W as b, X as k, y, z, aq as $, ar as Q, aC as W, aD as X, c as Z, b as ee, d as se, e as ae, a3 as te, a4 as oe, k as ne, l as re, J as ie, K as Re, am as pe, an as me, aG as le, aH as ue, aA as de, aB as Ee, G as Ce, L as Ne, j as ce, u as Ue, v as Se, H as Te, I as _e, ab as Pe, ad as De, ae as Ie, af as qe, a1 as Ae, a2 as Oe, ay as ge, az as je, as as fe, at as Ge, Z as Le, _ as ve, h as Fe, f as xe, i as Ke, g as He, r as Ye, t as we, q as Be, F as Je, a5 as he, a6 as Ve, a7 as Me, a8 as be, ac as ke, aa as ye, ah as ze, a9 as $e, av as Qe, Y as We, T as Xe, U as Ze, o as es, R as ss, V as as, B as ts, E as os, ao as ns, ap as rs, aI as is, aJ as Rs, aE as ps, aF as ms, aK as ls, aL as us, O as ds, Q as Es, aj as Cs, aw as Ns, ax as cs, n as Us, s as Ss } from "./user_service_connect-BZIXD0EN.js";
import { B as _s, f as Ps, E as Ds, F as Is, d as qs, e as As, I as Os, b as gs, L as js, P as fs, a as Gs, S as Ls, T as vs, U as Fs, c as xs, g as Ks, j as Hs, m as Ys, k as ws, h as Bs, i as Js, l as hs, p as Vs, s as Ms, u as bs } from "./opt_pb-Dp9pF04Q.js";
import { C as ys, c as zs, D as $s, d as Qs, G as Ws, a as Xs, L as Zs, b as ea, O as sa, S as aa } from "./operation_pb-CYWjPOnU.js";
import { Message as m, proto3 as o, protoInt64 as c, Timestamp as U } from "@bufbuild/protobuf";
const r = class r extends m {
  constructor(e) {
    super();
    /**
     * Project
     *
     * @generated from field: audiotool.project.v1.Project project = 1;
     */
    a(this, "project");
    /**
     * The DocumentService URL for the session where the studio connects to.
     *
     * @generated from field: string document_service_url = 2;
     */
    a(this, "documentServiceUrl", "");
    /**
     * The prefix URL for studio (location of assets in the CDN).
     *
     * @generated from field: string studio_prefix_url = 3;
     */
    a(this, "studioPrefixUrl", "");
    /**
     * The prefix URL for document_service (location of assets in the CDN).
     *
     * Within this location is the document_validator.wasm
     *
     * @generated from field: string document_service_prefix_url = 4;
     */
    a(this, "documentServicePrefixUrl", "");
    /**
     * The prefix URL for the audio engine (location of assets in the CDN).
     *
     * @generated from field: string audio_engine_prefix_url = 5;
     */
    a(this, "audioEnginePrefixUrl", "");
    o.util.initPartial(e, this);
  }
  static fromBinary(e, t) {
    return new r().fromBinary(e, t);
  }
  static fromJson(e, t) {
    return new r().fromJson(e, t);
  }
  static fromJsonString(e, t) {
    return new r().fromJsonString(e, t);
  }
  static equals(e, t) {
    return o.util.equals(r, e, t);
  }
};
a(r, "runtime", o), a(r, "typeName", "audiotool.project.v1.Session"), a(r, "fields", o.util.newFieldList(() => [
  { no: 1, name: "project", kind: "message", T: p },
  {
    no: 2,
    name: "document_service_url",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  {
    no: 3,
    name: "studio_prefix_url",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  {
    no: 4,
    name: "document_service_prefix_url",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  {
    no: 5,
    name: "audio_engine_prefix_url",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  }
]));
let E = r;
var l = /* @__PURE__ */ ((s) => (s[s.UNSPECIFIED = 0] = "UNSPECIFIED", s[s.STARTED = 1] = "STARTED", s[s.AUDIO_RENDERING = 3] = "AUDIO_RENDERING", s[s.AUDIO_RENDERED = 4] = "AUDIO_RENDERED", s[s.AUDIO_CONVERTING = 5] = "AUDIO_CONVERTING", s[s.AUDIO_CONVERTED = 6] = "AUDIO_CONVERTED", s[s.FINISHED = 7] = "FINISHED", s))(l || {});
o.util.setEnumType(l, "audiotool.project.v1.SyncTrackStatus", [
  { no: 0, name: "SYNC_TRACK_STATUS_UNSPECIFIED" },
  { no: 1, name: "SYNC_TRACK_STATUS_STARTED" },
  { no: 3, name: "SYNC_TRACK_STATUS_AUDIO_RENDERING" },
  { no: 4, name: "SYNC_TRACK_STATUS_AUDIO_RENDERED" },
  { no: 5, name: "SYNC_TRACK_STATUS_AUDIO_CONVERTING" },
  { no: 6, name: "SYNC_TRACK_STATUS_AUDIO_CONVERTED" },
  { no: 7, name: "SYNC_TRACK_STATUS_FINISHED" }
]);
const i = class i extends m {
  constructor(e) {
    super();
    /**
     * The project that is synced.
     *
     * @generated from field: audiotool.project.v1.Project project = 1;
     */
    a(this, "project");
    /**
     * Required. The mode to sync the track.
     *
     * @generated from field: audiotool.project.v1.SyncTrackMode mode = 2;
     */
    a(this, "mode", d.UNSPECIFIED);
    /**
     * The commit index of the project.
     *
     * @generated from field: uint64 commit_index = 3;
     */
    a(this, "commitIndex", c.zero);
    /**
     * The desired track name.
     *
     * @generated from field: string track_name = 4;
     */
    a(this, "trackName", "");
    /**
     * The user who triggered the sync.
     *
     * @generated from field: string user_name = 5;
     */
    a(this, "userName", "");
    /**
     * The time when the operation was created.
     *
     * @generated from field: google.protobuf.Timestamp create_time = 6;
     */
    a(this, "createTime");
    /**
     * The current status of the sync track process.
     *
     * @generated from field: audiotool.project.v1.SyncTrackStatus status = 7;
     */
    a(this, "status", 0);
    o.util.initPartial(e, this);
  }
  static fromBinary(e, t) {
    return new i().fromBinary(e, t);
  }
  static fromJson(e, t) {
    return new i().fromJson(e, t);
  }
  static fromJsonString(e, t) {
    return new i().fromJsonString(e, t);
  }
  static equals(e, t) {
    return o.util.equals(i, e, t);
  }
};
a(i, "runtime", o), a(i, "typeName", "audiotool.project.v1.SyncTrackInfo"), a(i, "fields", o.util.newFieldList(() => [
  { no: 1, name: "project", kind: "message", T: p },
  { no: 2, name: "mode", kind: "enum", T: o.getEnumType(d) },
  {
    no: 3,
    name: "commit_index",
    kind: "scalar",
    T: 4
    /* ScalarType.UINT64 */
  },
  {
    no: 4,
    name: "track_name",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  {
    no: 5,
    name: "user_name",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  { no: 6, name: "create_time", kind: "message", T: U },
  { no: 7, name: "status", kind: "enum", T: o.getEnumType(l) }
]));
let C = i;
var u = /* @__PURE__ */ ((s) => (s[s.SYNC_TRACK_SCREENSHOT_ENUM_UNSPECIFIED = 0] = "SYNC_TRACK_SCREENSHOT_ENUM_UNSPECIFIED", s[s.SYNC_TRACK_SCREENSHOT_RUNNING = 1] = "SYNC_TRACK_SCREENSHOT_RUNNING", s[s.SYNC_TRACK_SCREENSHOT_FINISHED = 2] = "SYNC_TRACK_SCREENSHOT_FINISHED", s))(u || {});
o.util.setEnumType(u, "audiotool.project.v1.SyncTrackScreenshotStatus", [
  { no: 0, name: "SYNC_TRACK_SCREENSHOT_ENUM_UNSPECIFIED" },
  { no: 1, name: "SYNC_TRACK_SCREENSHOT_RUNNING" },
  { no: 2, name: "SYNC_TRACK_SCREENSHOT_FINISHED" }
]);
const R = class R extends m {
  constructor(e) {
    super();
    /**
     * The project that is synced.
     *
     * @generated from field: audiotool.project.v1.Project project = 1;
     */
    a(this, "project");
    /**
     * The commit index of the project.
     *
     * @generated from field: uint64 commit_index = 2;
     */
    a(this, "commitIndex", c.zero);
    /**
     * The desired track name.
     *
     * @generated from field: string track_name = 3;
     */
    a(this, "trackName", "");
    /**
     * The user who triggered the sync.
     *
     * @generated from field: string user_name = 4;
     */
    a(this, "userName", "");
    /**
     * The time when the operation was created.
     *
     * @generated from field: google.protobuf.Timestamp create_time = 5;
     */
    a(this, "createTime");
    /**
     * The current status of the sync track process.
     *
     * @generated from field: audiotool.project.v1.SyncTrackScreenshotStatus status = 6;
     */
    a(this, "status", 0);
    o.util.initPartial(e, this);
  }
  static fromBinary(e, t) {
    return new R().fromBinary(e, t);
  }
  static fromJson(e, t) {
    return new R().fromJson(e, t);
  }
  static fromJsonString(e, t) {
    return new R().fromJsonString(e, t);
  }
  static equals(e, t) {
    return o.util.equals(R, e, t);
  }
};
a(R, "runtime", o), a(R, "typeName", "audiotool.project.v1.SyncTrackScreenshotInfo"), a(R, "fields", o.util.newFieldList(() => [
  { no: 1, name: "project", kind: "message", T: p },
  {
    no: 2,
    name: "commit_index",
    kind: "scalar",
    T: 4
    /* ScalarType.UINT64 */
  },
  {
    no: 3,
    name: "track_name",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  {
    no: 4,
    name: "user_name",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  { no: 5, name: "create_time", kind: "message", T: U },
  { no: 6, name: "status", kind: "enum", T: o.getEnumType(u) }
]));
let N = R;
export {
  A as Audiograph,
  O as AudiographService,
  g as BatchGetUsersRequest,
  j as BatchGetUsersResponse,
  _s as Bool,
  Ps as Bytes,
  ys as CancelOperationRequest,
  zs as CancelOperationResponse,
  f as CancelSampleUploadRequest,
  G as CancelSampleUploadResponse,
  L as CloseSessionRequest,
  v as CloseSessionResponse,
  F as CommentMode,
  x as CreatePresetRequest,
  K as CreatePresetResponse,
  H as CreateProjectRequest,
  Y as CreateProjectResponse,
  w as CreateProjectRoleRequest,
  B as CreateProjectRoleResponse,
  J as CreateSampleRequest,
  h as CreateSampleResponse,
  $s as DeleteOperationRequest,
  Qs as DeleteOperationResponse,
  V as DeletePresetRequest,
  M as DeletePresetResponse,
  b as DeleteProjectRequest,
  k as DeleteProjectResponse,
  y as DeleteProjectRoleRequest,
  z as DeleteProjectRoleResponse,
  $ as DeleteSampleRequest,
  Q as DeleteSampleResponse,
  W as DeleteUserRequest,
  X as DeleteUserResponse,
  Ds as EntityOptions,
  Is as FieldOptions,
  qs as Float,
  As as FloatRange,
  Z as GetAudiographChannels,
  ee as GetAudiographResolution,
  se as GetAudiographsRequest,
  ae as GetAudiographsResponse,
  te as GetLatestVersionBundleRequest,
  oe as GetLatestVersionBundleResponse,
  Ws as GetOperationRequest,
  Xs as GetOperationResponse,
  ne as GetPresetRequest,
  re as GetPresetResponse,
  ie as GetProjectRequest,
  Re as GetProjectResponse,
  pe as GetSampleRequest,
  me as GetSampleResponse,
  le as GetSettingsRequest,
  ue as GetSettingsResponse,
  de as GetUserRequest,
  Ee as GetUserResponse,
  Ce as Graph,
  Os as Int32,
  gs as Int32Range,
  Zs as ListOperationsRequest,
  ea as ListOperationsResponse,
  js as ListOptions,
  Ne as ListPresetsRequest,
  ce as ListPresetsResponse,
  Ue as ListProjectRolesRequest,
  Se as ListProjectRolesResponse,
  Te as ListProjectsRequest,
  _e as ListProjectsResponse,
  Pe as ListSampleProcessesRequest,
  De as ListSampleProcessesResponse,
  Ie as ListSamplesRequest,
  qe as ListSamplesResponse,
  Ae as ListSessionsRequest,
  Oe as ListSessionsResponse,
  ge as ListUsersRequest,
  je as ListUsersResponse,
  fe as ListenRequest,
  Ge as ListenResponse,
  Le as OpenSessionRequest,
  ve as OpenSessionResponse,
  sa as Operation,
  fs as PointerOptions,
  Fe as Preset,
  xe as PresetDeviceType,
  Ke as PresetService,
  He as PresetUsage,
  p as Project,
  Ye as ProjectRole,
  we as ProjectRoleService,
  Be as ProjectRoleType,
  Je as ProjectService,
  he as ReconvertLegacyProjectRequest,
  Ve as ReconvertLegacyProjectResponse,
  Me as SampleConvertDoneErrorType,
  be as SampleConvertInfo,
  ke as SampleProcess,
  ye as SampleService,
  ze as SampleUploadEndpoint,
  $e as SampleUploadInfo,
  E as Session,
  Qe as Settings,
  aa as Status,
  Gs as String,
  Ls as StringFormat,
  C as SyncTrackInfo,
  d as SyncTrackMode,
  We as SyncTrackRequest,
  N as SyncTrackScreenshotInfo,
  u as SyncTrackScreenshotStatus,
  l as SyncTrackStatus,
  vs as TargetType,
  Xe as TrackLicense,
  Fs as UInt32,
  xs as UInt32Range,
  Ze as UpdatePresetRequest,
  es as UpdatePresetResponse,
  ss as UpdateProjectRequest,
  as as UpdateProjectResponse,
  ts as UpdateProjectRoleRequest,
  os as UpdateProjectRoleResponse,
  ns as UpdateSampleRequest,
  rs as UpdateSampleResponse,
  is as UpdateSettingsRequest,
  Rs as UpdateSettingsResponse,
  ps as UpdateUserRequest,
  ms as UpdateUserResponse,
  ls as UploadAvatarRequest,
  us as UploadAvatarResponse,
  ds as UploadCoverRequest,
  Es as UploadCoverResponse,
  Cs as UploadSampleFinishedRequest,
  Ns as User,
  cs as UserService,
  Ks as bool,
  Hs as bytes,
  Ys as entity,
  ws as field,
  Bs as float,
  Js as int32,
  hs as list,
  Us as neverThrowingFetch,
  Vs as pointer,
  Ss as sample,
  Ms as string$,
  bs as uint32
};
