// @audiotool/nexus/dist/index.js
import { Any as se, createRegistry as K } from "@bufbuild/protobuf";
import { a as N, P as q, b as te, N as oe, K as G, d as ie, i as ce } from "./synced-document-ywEybIAl.js";
import { c as ra } from "./synced-document-ywEybIAl.js";
import { i as ne, a7 as V, a8 as M, aa as de, n as v, aO as le, aP as pe, aQ as D, aR as R, a as ge, ax as fe, t as be, F as ue } from "./user_service_connect-BZIXD0EN.js";
import { s as Y } from "./lang-K-8hAzE4.js";
const U = (e) => {
  const a = /[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/i, s = e.match(a);
  return s == null ? new Error() : s[0];
}, me = {
  0: "presets/a0931d29-e023-4de4-9ff6-24adb6131a24",
  1: "presets/820a78bf-faae-4abd-8324-1bba1587e7b2",
  2: "presets/14960b89-6d7b-4fa1-8d96-7980ed73411a",
  3: "presets/06675de4-8cbb-4a90-a72a-6a72533809c7",
  4: "presets/36ef9001-51b1-4672-9754-4df245dce4ca",
  5: "presets/ea4909a7-bf2d-40c9-9b87-7db810f227cb",
  6: "presets/1b089fa6-d037-4dd6-875f-b102e27a856c",
  7: "presets/d6220705-1098-48bf-8285-c43c2ad6c518",
  8: "presets/d6f1683d-1e94-425e-b765-d65abe6b3936",
  9: "presets/bf29be77-cda7-416d-b2dd-ef3c69651a8d",
  10: "presets/f2884edc-36b2-4d70-b15c-85092d18198a",
  11: "presets/8a174394-e0f2-4dff-9b7d-f7e9b896989d",
  12: "presets/a3450aa2-5040-457e-9487-a691c5c65334",
  13: "presets/55f01a34-9b08-4b56-a2c1-8f5b9addb79d",
  14: "presets/b2f74cf6-dafc-4470-8fa0-54f3777298d5",
  15: "presets/4fcfbfcf-0e96-4f20-a8ae-d0a106ab9193",
  16: "presets/cdcdd0ae-132e-40c0-a8e2-ce16743fe458",
  17: "presets/7bb3404b-9059-4870-a3e4-3a5573b6287c",
  18: "presets/77494ec9-1e14-4dae-a34e-125dfc89e3b2",
  19: "presets/639c481c-0ed7-4511-86dc-c3c2ed83f44c",
  20: "presets/a732efd9-408e-489e-8455-002696ea97e1",
  21: "presets/46700c95-d85f-496c-8dfa-f0866d5467a3",
  22: "presets/82696cc3-5d04-44e7-a01d-0fbaa00bdb6e",
  23: "presets/e26f1313-000d-4deb-8721-3fbfa43f5cb3",
  24: "presets/b6c5ecf7-030e-4d0a-8297-09e0f0541c26",
  25: "presets/f02ab80c-4f10-4f6e-9efc-2f1c72f0e60d",
  26: "presets/f43b259a-a743-4cb0-921f-44656833bb2a",
  27: "presets/ebacba3a-ffbe-4515-b7b7-c2507e77a3ce",
  28: "presets/f55c3487-726e-43b0-b2d9-890c3428ea09",
  29: "presets/e6d5a691-4c6f-40d7-89fa-d62a3ce292ef",
  30: "presets/4dedadf7-39ef-47b8-8940-1a1c2625d5f1",
  31: "presets/973a7754-aa3b-48b5-8e5c-7dcb04fae6d6",
  32: "presets/2afb483c-2aaf-4863-a512-bb3dabd64ae0",
  33: "presets/859c09db-c948-49ec-96cd-e639716128f6",
  34: "presets/fe33dd5a-f585-4063-ad7b-d2ae964fc318",
  35: "presets/7f29cd8a-8c0e-44f2-ba6f-a32ab255f5c3",
  36: "presets/4222d3d1-7df3-483a-a208-93614abab1c4",
  37: "presets/7d3955b0-d0b8-4287-bfc8-db7f7c07ddef",
  38: "presets/0e6f20e6-9160-42a3-9509-fd4487223258",
  39: "presets/7748c722-018a-4219-9c53-b6949e634cd0",
  40: "presets/1b0efe86-22e9-4d0c-8963-f1e00cc331c8",
  41: "presets/0b01cd16-56b3-4508-a193-1123410b9718",
  42: "presets/351c16de-7f58-41f0-9995-a60ef48233f2",
  43: "presets/d74c0c4a-f5ee-429b-8274-b43f00643af9",
  44: "presets/f95280dd-2b04-48e9-a6b5-dabf9883209b",
  45: "presets/689a975a-e0a9-448b-91b8-c4c6898e10eb",
  46: "presets/8af7540d-5686-4467-bf83-eee081b1da43",
  47: "presets/2ebebdf9-3d1d-42e2-95ea-b116de82b827",
  48: "presets/c4f798f8-d931-446b-ab7f-b4eaa75182b2",
  49: "presets/46ecb0fc-8c64-4a46-9150-386c16d6918e",
  50: "presets/e15ce70a-e7ff-4532-b619-a30b17b97b57",
  51: "presets/8b3de6e0-8e86-47ba-934d-014a604cf289",
  52: "presets/16987293-10a7-4b8f-8426-ffa0dbadbf33",
  53: "presets/27605c9a-fc33-4eb8-982f-dafb209291f8",
  54: "presets/1b01e9ba-2677-440e-a055-9616ddf0da58",
  55: "presets/6642f9d9-22ba-4d38-ba8e-d143c31f5de1",
  56: "presets/c7dc2efa-90f3-4780-b326-d3e0374fc0eb",
  57: "presets/b81332d8-6ce2-491e-b4d2-107d24ba8dc3",
  58: "presets/3aaba4d6-a63b-4cee-b9b3-c39e1b3dff10",
  59: "presets/40f254ea-2617-4949-be1f-2255cfa5edbd",
  60: "presets/a0e633c2-4979-42e5-b4f9-c16b3e5c3866",
  61: "presets/c6006e20-496d-4625-b815-ee5831d14481",
  62: "presets/58bee540-5410-47eb-af97-e08c9d0b9f04",
  63: "presets/b8d549b3-ca15-421e-bbbd-b449b876907a",
  64: "presets/66c5d885-b784-40f5-a134-451c7627a71d",
  65: "presets/46af8d73-1844-4b27-af60-89f3407a4d50",
  66: "presets/96533625-f1c1-4762-8209-b055f6f678b5",
  67: "presets/11566846-86fd-48b6-b94f-562d4e53daf9",
  68: "presets/dd0a0ca3-7449-4728-86c4-53c8fdae2384",
  69: "presets/51f8a3e0-fdeb-460b-8b65-9473f158058c",
  70: "presets/4232177f-ab6c-435c-b73d-3bfd6ae4f567",
  71: "presets/23b891f7-e80b-47d3-a861-2c97f6e5e4d9",
  72: "presets/b532b371-6021-46dd-a4af-5588684ab77d",
  73: "presets/defb2e3b-d2e0-4eb9-8d71-7df36d31ec3d",
  74: "presets/f7ff13a4-a825-4bf4-a06e-ff5f081575ce",
  75: "presets/0316febc-3ff1-45ef-a369-8a95806f2646",
  76: "presets/edf04b25-728d-47ce-8a77-c455f6ae9cd4",
  77: "presets/be0b2d05-abbf-44c9-b403-c4ab921cfec0",
  78: "presets/98187def-976c-4650-85dc-02e0d8bcc092",
  79: "presets/b8ccc37a-395b-4a5c-accf-88798594d34a",
  80: "presets/8d165833-0fb5-4538-b357-417e0d21a592",
  81: "presets/1f7116dd-c242-43e1-9ab9-fc5f706a2158",
  82: "presets/bdb78d0b-d83d-43a3-8f79-b66a142c929f",
  83: "presets/7ce36f23-7b02-49fc-8bd1-ca7e12585eb8",
  84: "presets/7e01138c-e98d-4f62-a03d-3a316d7d5305",
  85: "presets/a76b3b24-03fa-4873-970f-7557ffa88749",
  86: "presets/025764de-ff81-4362-8414-5aafe1926a43",
  87: "presets/52e2e11e-ddb9-4785-9ffc-82dbf7af478a",
  88: "presets/0409f236-80ba-4a36-aaf7-d886604ff39c",
  89: "presets/060c37fe-4a10-4095-9530-3dc6c1c428c7",
  90: "presets/f951bad2-f378-4c1c-b5e5-f71e4105951f",
  91: "presets/a7d023a7-43b1-4503-86ff-412cff0a0fc6",
  92: "presets/39811b37-507d-46a0-bc2f-8f127606e0d4",
  93: "presets/c30af7a6-dca0-4948-8cd4-979ec0ee1b34",
  94: "presets/95faa5c8-51ae-434b-b34c-2a9a4e2f2443",
  95: "presets/cddab967-6a0c-4c2e-8a51-8e308e53ef99",
  96: "presets/7b993ed1-7075-41e3-b51c-adbf2674f3b9",
  97: "presets/a22397a4-68be-4b16-8bd2-d58ce43d0c4d",
  98: "presets/479c325e-e2f1-4f12-98d6-27ddc6306216",
  99: "presets/285a7dc1-cc01-4277-9dcd-a29a40bb5a60",
  100: "presets/d96df2aa-9a77-40b7-a439-d720f2252618",
  101: "presets/41f458e0-f696-4c03-b722-d52a20efced0",
  102: "presets/4ae465e7-8a9a-40a0-a066-13ccfb350ddf",
  103: "presets/127dca19-bbf4-4104-8e65-d20f8dbba0fd",
  104: "presets/05b6d94a-1a2e-4542-b3e9-da1bc534b975",
  105: "presets/a458046c-59e9-49f6-a628-22876927c897",
  106: "presets/23145472-516d-4f9c-90b4-9f63ff8ce455",
  107: "presets/02ca6de1-1a08-4994-a87c-23c261a43634",
  108: "presets/32779b50-653b-49c0-9f35-c5fa17570cbe",
  109: "presets/2aa96f1c-3c98-49c3-b4da-bc19d3ac64ac",
  110: "presets/9880f830-5184-47ec-88bb-0b268613853b",
  111: "presets/4ecb6171-87fd-4f63-9adc-c4ccc765ffcc",
  112: "presets/677f0ef2-692b-4474-9d11-e1610918d0b0",
  113: "presets/47241bb4-9eff-463d-84b4-620b55d8297b",
  114: "presets/84f0b350-e1b1-4ef3-a3a7-3dbee2a8e708",
  115: "presets/6023ccdc-fa78-495d-a22f-c9ac9ffabcf9",
  116: "presets/04bd3b71-a9c8-44c9-8bfe-f2e7182514b2",
  117: "presets/90ad8394-47cf-4089-88f8-3485d3d420c2",
  118: "presets/34c8332e-d188-4b08-8bab-989ab5027b35",
  119: "presets/92266fda-dfd3-4e16-88c7-1f0160d3b8d4",
  120: "presets/4792e4fa-4be4-4c15-ae4e-a1a89238aa6f",
  121: "presets/8ef100a5-5462-42bd-af07-5be8ff8b07ef",
  122: "presets/b0163790-e8c1-4f46-848c-9cb5f5cb08b5",
  123: "presets/0bed8f0f-f854-4100-b003-7be900bbed8c",
  124: "presets/de45f831-dca8-4bfe-b5d6-c7bf2a58c38b",
  125: "presets/7208226b-5271-4852-beb7-f0194bacb40d",
  126: "presets/d39b7b86-42a0-47fc-99ea-6b38a8d890a0",
  127: "presets/e859bb19-7e48-4831-8265-744446cef771"
}, ye = {
  0: "presets/6b49e6b6-09f1-43d0-b15d-adeccdb9eeab",
  8: "presets/284daead-69b1-459d-83f6-1e08d6d8cee9",
  16: "presets/e3ce588c-f237-481d-9f87-be3d423b29ae",
  24: "presets/18aa38cd-5fc5-4855-b7a9-7b5fca4f2dcd",
  25: "presets/d6c69f5f-8e34-424d-8927-1a9ffd68e06a",
  32: "presets/56f936cf-452d-4d5e-bf03-28da021f0025",
  40: "presets/c9973261-780e-445f-b8da-c43fcc422574",
  48: "presets/42e00fe8-e6f0-41e2-8982-174ac70d5e7d"
}, he = [
  {
    program: 0,
    slug: "acoustic-piano",
    displayName: "Acoustic Piano",
    category: "Keys",
    tags: ["keys", "piano", "acoustic", "warm", "classical"],
    description: "Classic full-bodied piano sound",
    id: "presets/a0931d29-e023-4de4-9ff6-24adb6131a24"
  },
  {
    program: 1,
    slug: "bright-grand-2",
    displayName: "Bright Grand 2",
    category: "Keys",
    tags: ["keys", "piano", "bright", "pop", "grand"],
    description: "Brighter, harder attack",
    id: "presets/820a78bf-faae-4abd-8324-1bba1587e7b2"
  },
  {
    program: 2,
    slug: "piano-2",
    displayName: "Piano 2",
    category: "Keys",
    tags: ["keys", "piano", "bright", "pop", "classical"],
    description: "Classic piano",
    id: "presets/14960b89-6d7b-4fa1-8d96-7980ed73411a"
  },
  {
    program: 3,
    slug: "honky-tonk-piano",
    displayName: "Honky-Tonk Piano",
    category: "Keys",
    tags: ["keys", "piano", "detuned", "vintage", "honky-tonk"],
    description: "Detuned barroom piano",
    id: "presets/06675de4-8cbb-4a90-a72a-6a72533809c7"
  },
  {
    program: 4,
    slug: "electronic-piano-1",
    displayName: "Electronic Piano 1",
    category: "Keys",
    tags: ["keys", "piano", "electronic", "electric", "soft", "jazz", "rhodes", "epiano"],
    description: "Rhodes-style smooth tone",
    id: "presets/36ef9001-51b1-4672-9754-4df245dce4ca"
  },
  {
    program: 5,
    slug: "electronic-piano-2",
    displayName: "Electronic Piano 2",
    category: "Keys",
    tags: ["keys", "piano", "electronic", "electric", "bright", "fm", "pop", "fm", "bells", "epiano"],
    description: "FM-type bell piano",
    id: "presets/ea4909a7-bf2d-40c9-9b87-7db810f227cb"
  },
  {
    program: 6,
    slug: "harpsichord",
    displayName: "Harpsichord",
    category: "Keys",
    tags: ["keys", "piano", "baroque", "plucked", "plucky", "classical", "harpsichord"],
    description: "Plucked baroque keyboard",
    id: "presets/1b089fa6-d037-4dd6-875f-b102e27a856c"
  },
  {
    program: 7,
    slug: "clavinet",
    displayName: "Clavinet",
    category: "Keys",
    tags: ["keys", "piano", "funk", "clavinet", "percussive", "electronic", "electric"],
    description: "Funky percussive electronic keyboard",
    id: "presets/d6220705-1098-48bf-8285-c43c2ad6c518"
  },
  {
    program: 8,
    slug: "celesta",
    displayName: "Celesta",
    category: "Mallets",
    tags: ["mallets", "bright", "bells", "orchestral", "chromati", "celesta"],
    description: "Bell-like piano",
    id: "presets/d6f1683d-1e94-425e-b765-d65abe6b3936"
  },
  {
    program: 9,
    slug: "glockenspiel",
    displayName: "Glockenspiel",
    category: "Mallets",
    tags: ["mallets", "bells", "metallic", "bright", "chromati", "glockenspiel"],
    description: "High metallic chime",
    id: "presets/bf29be77-cda7-416d-b2dd-ef3c69651a8d"
  },
  {
    program: 10,
    slug: "music-box",
    displayName: "Music Box",
    category: "Mallets",
    tags: ["mallets", "toy", "soft", "bells", "chromatic", "musicbox"],
    description: "Wind-up toy box tone",
    id: "presets/f2884edc-36b2-4d70-b15c-85092d18198a"
  },
  {
    program: 11,
    slug: "vibraphone",
    displayName: "Vibraphone",
    category: "Mallets",
    tags: ["mallets", "jazz", "vibrato", "chromatic", "vibraphone"],
    description: "Jazz mallet instrument",
    id: "presets/8a174394-e0f2-4dff-9b7d-f7e9b896989d"
  },
  {
    program: 12,
    slug: "marimba",
    displayName: "Marimba",
    category: "Mallets",
    tags: ["mallets", "warm", "wood", "chromatic", "marimba"],
    description: "Wooden mallet percussion",
    id: "presets/a3450aa2-5040-457e-9487-a691c5c65334"
  },
  {
    program: 13,
    slug: "xylophone",
    displayName: "Xylophone",
    category: "Mallets",
    tags: ["mallets", "bright", "wood", "chromatic", "xylophone"],
    description: "Bright wooden percussion",
    id: "presets/55f01a34-9b08-4b56-a2c1-8f5b9addb79d"
  },
  {
    program: 14,
    slug: "tubular-bells",
    displayName: "Tubular Bells",
    category: "Mallets",
    tags: ["mallets", "bells", "metallic", "cinematic", "chromatic", "church", "cathedral"],
    description: "Church bell tone",
    id: "presets/b2f74cf6-dafc-4470-8fa0-54f3777298d5"
  },
  {
    program: 15,
    slug: "dulcimer",
    displayName: "Dulcimer",
    category: "Mallets",
    tags: ["mallets", "folk", "string", "plucked", "plucky", "chromatic"],
    description: "Hammered string tone",
    id: "presets/4fcfbfcf-0e96-4f20-a8ae-d0a106ab9193"
  },
  {
    program: 16,
    slug: "jazz-organ",
    displayName: "Jazz Organ",
    category: "Keys",
    tags: ["keys", "organ", "attack", "bright", "hammond", "jazz"],
    description: "Hammond-style tonewheel",
    id: "presets/cdcdd0ae-132e-40c0-a8e2-ce16743fe458"
  },
  {
    program: 17,
    slug: "hammond-organ",
    displayName: "Hammond Organ",
    category: "Keys",
    tags: ["keys", "organ", "jazz", "rock", "hammond"],
    description: "Hammond-style tonewheel",
    id: "presets/7bb3404b-9059-4870-a3e4-3a5573b6287c"
  },
  {
    program: 18,
    slug: "rock-organ",
    displayName: "Rock Organ",
    category: "Keys",
    tags: ["keys", "organ", "rock", "gritty", "hammond"],
    description: "Hammond-style tonewheel",
    id: "presets/77494ec9-1e14-4dae-a34e-125dfc89e3b2"
  },
  {
    program: 19,
    slug: "church-organ",
    displayName: "Church Organ",
    category: "Keys",
    tags: ["keys", "organ", "pipe", "church", "majestic", "cathedral"],
    description: "Grand cathedral sound",
    id: "presets/639c481c-0ed7-4511-86dc-c3c2ed83f44c"
  },
  {
    program: 20,
    slug: "reed-organ",
    displayName: "Reed Organ",
    category: "Keys",
    tags: ["keys", "organ", "folk", "reed", "vintage"],
    description: "Harmonium / pump organ",
    id: "presets/a732efd9-408e-489e-8455-002696ea97e1"
  },
  {
    program: 21,
    slug: "accordion",
    displayName: "Accordion",
    category: "Misc",
    tags: ["misc", "keys", "organ", "folk", "accordion", "world", "traditional"],
    description: "Traditional accordion tone",
    id: "presets/46700c95-d85f-496c-8dfa-f0866d5467a3"
  },
  {
    program: 22,
    slug: "harmonica",
    displayName: "Harmonica",
    category: "Wind",
    tags: ["wind", "organ", "blues", "harmonica", "solo", "reed"],
    description: "Breath-based reed tone",
    id: "presets/82696cc3-5d04-44e7-a01d-0fbaa00bdb6e"
  },
  {
    program: 23,
    slug: "bandoneon",
    displayName: "Bandoneon",
    category: "Misc",
    tags: ["misc", "keys", "organ", "latin", "accordion", "dance", "reed"],
    description: "Reed-based button accordion",
    id: "presets/e26f1313-000d-4deb-8721-3fbfa43f5cb3"
  },
  {
    program: 24,
    slug: "nylon-guitar",
    displayName: "Nylon Guitar",
    category: "Guitar",
    tags: ["guitar", "acoustic", "warm", "classical", "nylon"],
    description: "Classical guitar tone",
    id: "presets/b6c5ecf7-030e-4d0a-8297-09e0f0541c26"
  },
  {
    program: 25,
    slug: "dark-steel-guitar",
    displayName: "Dark Steel Guitar",
    category: "Guitar",
    tags: ["guitar", "folk", "pop", "bright", "steel"],
    description: "Bright steel-string sound",
    id: "presets/f02ab80c-4f10-4f6e-9efc-2f1c72f0e60d"
  },
  {
    program: 26,
    slug: "jazz-guitar",
    displayName: "Jazz Guitar",
    category: "Guitar",
    tags: ["guitar", "jazz", "clean", "electronic", "electric", "smooth"],
    description: "Smooth hollow-body tone",
    id: "presets/f43b259a-a743-4cb0-921f-44656833bb2a"
  },
  {
    program: 27,
    slug: "clean-guitar",
    displayName: "Clean Guitar",
    category: "Guitar",
    tags: ["guitar", "electronic", "electric", "pop", "clean", "balanced"],
    description: "Balanced clean tone",
    id: "presets/ebacba3a-ffbe-4515-b7b7-c2507e77a3ce"
  },
  {
    program: 28,
    slug: "muted-guitar",
    displayName: "Muted Guitar",
    category: "Guitar",
    tags: ["guitar", "funk", "rhythm", "tight", "muted", "plucky"],
    description: "Palm-muted pluck",
    id: "presets/f55c3487-726e-43b0-b2d9-890c3428ea09"
  },
  {
    program: 29,
    slug: "overdriven-guitar",
    displayName: "Overdriven Guitar",
    category: "Guitar",
    tags: ["guitar", "rock", "gritty", "lead", "crunchy", "distorted", "overdrive"],
    description: "Crunchy rock tone",
    id: "presets/e6d5a691-4c6f-40d7-89fa-d62a3ce292ef"
  },
  {
    program: 30,
    slug: "distortion-guitar",
    displayName: "Distortion Guitar",
    category: "Guitar",
    tags: ["guitar", "metal", "aggressive", "lead", "distorted"],
    description: "Heavy metal tone",
    id: "presets/4dedadf7-39ef-47b8-8940-1a1c2625d5f1"
  },
  {
    program: 31,
    slug: "guitar-harmonics",
    displayName: "Guitar Harmonics",
    category: "Guitar",
    tags: ["guitar", "ambient", "harmonic", "bells"],
    description: "Bell-like guitar harmonics",
    id: "presets/973a7754-aa3b-48b5-8e5c-7dcb04fae6d6"
  },
  {
    program: 32,
    slug: "acoustic-bass",
    displayName: "Acoustic Bass",
    category: "Bass",
    tags: ["bass", "jazz", "acoustic", "warm", "upright"],
    description: "Upright double bass",
    id: "presets/2afb483c-2aaf-4863-a512-bb3dabd64ae0"
  },
  {
    program: 33,
    slug: "fingered-bass",
    displayName: "Fingered Bass",
    category: "Bass",
    tags: ["bass", "pop", "groove", "soft", "smooth", "electronic", "electric", "fingered"],
    description: "Smooth electronic bass",
    id: "presets/859c09db-c948-49ec-96cd-e639716128f6"
  },
  {
    program: 34,
    slug: "picked-bass",
    displayName: "Picked Bass",
    category: "Bass",
    tags: ["bass", "rock", "pick", "bright", "sharp", "picked"],
    description: "Sharper bass tone",
    id: "presets/fe33dd5a-f585-4063-ad7b-d2ae964fc318"
  },
  {
    program: 35,
    slug: "fretless-bass",
    displayName: "Fretless Bass",
    category: "Bass",
    tags: ["bass", "jazz", "smooth", "expressive", "slide", "glide"],
    description: "Slides and glides",
    id: "presets/7f29cd8a-8c0e-44f2-ba6f-a32ab255f5c3"
  },
  {
    program: 36,
    slug: "slap-bass-1",
    displayName: "Slap Bass 1",
    category: "Bass",
    tags: ["bass", "funk", "slap", "bright"],
    description: "Funky thumb slap",
    id: "presets/4222d3d1-7df3-483a-a208-93614abab1c4"
  },
  {
    program: 37,
    slug: "slap-bass-2",
    displayName: "Slap Bass 2",
    category: "Bass",
    tags: ["bass", "funk", "aggressive", "slap"],
    description: "Heavier slap sound",
    id: "presets/7d3955b0-d0b8-4287-bfc8-db7f7c07ddef"
  },
  {
    program: 38,
    slug: "synth-bass-1",
    displayName: "Synth Bass 1",
    category: "Bass",
    tags: ["synth", "bass", "analog", "bright"],
    description: "Analog-style bass bright",
    id: "presets/0e6f20e6-9160-42a3-9509-fd4487223258"
  },
  {
    program: 39,
    slug: "synth-bass-2",
    displayName: "Synth Bass 2",
    category: "Bass",
    tags: ["synth", "bass", "analog", "dark"],
    description: "Analog-style bass dark",
    id: "presets/7748c722-018a-4219-9c53-b6949e634cd0"
  },
  {
    program: 40,
    slug: "violin",
    displayName: "Violin",
    category: "Strings",
    tags: ["strings", "solo", "classical", "violin"],
    description: "Solo bowed string",
    id: "presets/1b0efe86-22e9-4d0c-8963-f1e00cc331c8"
  },
  {
    program: 41,
    slug: "viola",
    displayName: "Viola",
    category: "Strings",
    tags: ["strings", "warm", "ensemble", "violin", "viola", "classical"],
    description: "Warm mid-range string",
    id: "presets/0b01cd16-56b3-4508-a193-1123410b9718"
  },
  {
    program: 42,
    slug: "cello",
    displayName: "Cello",
    category: "Strings",
    tags: ["strings", "warm", "solo", "cello", "deep", "emotional", "classical"],
    description: "Deep, emotional tone",
    id: "presets/351c16de-7f58-41f0-9995-a60ef48233f2"
  },
  {
    program: 43,
    slug: "contrabass",
    displayName: "Contrabass",
    category: "Strings",
    tags: ["strings", "bass", "orchestral", "contrabass", "classical"],
    description: "Orchestral bass",
    id: "presets/d74c0c4a-f5ee-429b-8274-b43f00643af9"
  },
  {
    program: 44,
    slug: "tremolo-strings",
    displayName: "Tremolo Strings",
    category: "Strings",
    tags: ["strings", "tremolo", "orchestral", "tremolo", "classical"],
    description: "Fast repeated bowing",
    id: "presets/f95280dd-2b04-48e9-a6b5-dabf9883209b"
  },
  {
    program: 45,
    slug: "pizzicato-strings",
    displayName: "Pizzicato Strings",
    category: "Strings",
    tags: ["strings", "plucks", "plucked", "plucky", "staccato", "orchestral", "pizzicato", "classical"],
    description: "Plucked orchestral strings",
    id: "presets/689a975a-e0a9-448b-91b8-c4c6898e10eb"
  },
  {
    program: 46,
    slug: "harp",
    displayName: "Harp",
    category: "Strings",
    tags: ["strings", "harp", "arpeggio", "arpeggiated", "sweeping"],
    description: "Sweeping harp sound",
    id: "presets/8af7540d-5686-4467-bf83-eee081b1da43"
  },
  {
    program: 47,
    slug: "timpani",
    displayName: "Timpani",
    category: "Perc",
    tags: ["percussion", "percussive", "drums", "orchestral"],
    description: "Tuned orchestral drum",
    id: "presets/2ebebdf9-3d1d-42e2-95ea-b116de82b827"
  },
  {
    program: 48,
    slug: "string-section",
    displayName: "String Section",
    category: "Strings",
    tags: ["strings", "ensemble", "attack", "classical"],
    description: "Full string section",
    id: "presets/c4f798f8-d931-446b-ab7f-b4eaa75182b2"
  },
  {
    program: 49,
    slug: "string-ensemble",
    displayName: "String Ensemble",
    category: "Strings",
    tags: ["strings", "ensemble", "lush", "classical"],
    description: "Full string section with low attack",
    id: "presets/46ecb0fc-8c64-4a46-9150-386c16d6918e"
  },
  {
    program: 50,
    slug: "synth-strings-1",
    displayName: "Synth Strings 1",
    category: "Strings",
    tags: ["synth", "strings", "pad", "ensemble", "analog"],
    description: "Classic analog pad",
    id: "presets/e15ce70a-e7ff-4532-b619-a30b17b97b57"
  },
  {
    program: 51,
    slug: "synth-strings-2",
    displayName: "Synth Strings 2",
    category: "Strings",
    tags: ["synth", "strings", "pad", "bright", "ensemble", "digital"],
    description: "Bright digital pad",
    id: "presets/8b3de6e0-8e86-47ba-934d-014a604cf289"
  },
  {
    program: 52,
    slug: "choir-aahs",
    displayName: "Choir Aahs",
    category: "Pad",
    tags: ["pad", "ensemble", "choir", "vocals", "voices", "classical"],
    description: "Vocal ensemble “Aah”",
    id: "presets/16987293-10a7-4b8f-8426-ffa0dbadbf33"
  },
  {
    program: 53,
    slug: "choir-oohs",
    displayName: "Choir Oohs",
    category: "Pad",
    tags: ["pad", "ensemble", "choir", "vocals", "voices", "soft", "classical"],
    description: "Softer vocal “Ooh”",
    id: "presets/27605c9a-fc33-4eb8-982f-dafb209291f8"
  },
  {
    program: 54,
    slug: "synth-voice",
    displayName: "Synth Voice",
    category: "Pad",
    tags: ["synth", "vocals", "pad", "ensemble", "voices"],
    description: "Vocal-like synth tone",
    id: "presets/1b01e9ba-2677-440e-a055-9616ddf0da58"
  },
  {
    program: 55,
    slug: "orchestra-hit",
    displayName: "Orchestra Hit",
    category: "FX",
    tags: ["fx", "ensemble", "orchestral", "hit", "retro", "classical", "staccato"],
    description: "Classic staccato hit",
    id: "presets/6642f9d9-22ba-4d38-ba8e-d143c31f5de1"
  },
  {
    program: 56,
    slug: "trumpet",
    displayName: "Trumpet",
    category: "Wind",
    tags: ["wind", "brass", "solo", "bright", "trumpet"],
    description: "Bright brass solo",
    id: "presets/c7dc2efa-90f3-4780-b326-d3e0374fc0eb"
  },
  {
    program: 57,
    slug: "trombone",
    displayName: "Trombone",
    category: "Wind",
    tags: ["wind", "brass", "warm", "orchestral", "trombone"],
    description: "Warm low brass",
    id: "presets/b81332d8-6ce2-491e-b4d2-107d24ba8dc3"
  },
  {
    program: 58,
    slug: "tuba",
    displayName: "Tuba",
    category: "Wind",
    tags: ["wind", "brass", "bass", "orchestral", "tuba"],
    description: "Deep orchestral brass",
    id: "presets/3aaba4d6-a63b-4cee-b9b3-c39e1b3dff10"
  },
  {
    program: 59,
    slug: "muted-trumpet",
    displayName: "Muted Trumpet",
    category: "Wind",
    tags: ["wind", "brass", "jazz", "muted", "trumpet", "muted"],
    description: "Jazz trumpet with mute",
    id: "presets/40f254ea-2617-4949-be1f-2255cfa5edbd"
  },
  {
    program: 60,
    slug: "french-horn",
    displayName: "French Horn",
    category: "Wind",
    tags: ["wind", "brass", "orchestral", "warm", "smooth", "horn"],
    description: "Smooth orchestral horn",
    id: "presets/a0e633c2-4979-42e5-b4f9-c16b3e5c3866"
  },
  {
    program: 61,
    slug: "brass-section",
    displayName: "Brass Section",
    category: "Wind",
    tags: ["wind", "brass", "ensemble", "cinematic", "powerful"],
    description: "Powerful ensemble",
    id: "presets/c6006e20-496d-4625-b815-ee5831d14481"
  },
  {
    program: 62,
    slug: "synth-brass-1",
    displayName: "Synth Brass 1",
    category: "Wind",
    tags: ["synth", "brass", "bright", "wind", "digital"],
    description: "Digital brass synth",
    id: "presets/58bee540-5410-47eb-af97-e08c9d0b9f04"
  },
  {
    program: 63,
    slug: "synth-brass-2",
    displayName: "Synth Brass 2",
    category: "Wind",
    tags: ["synth", "brass", "analog", "wind"],
    description: "Analog brass synth",
    id: "presets/b8d549b3-ca15-421e-bbbd-b449b876907a"
  },
  {
    program: 64,
    slug: "soprano-sax",
    displayName: "Soprano Sax",
    category: "Wind",
    tags: ["wind", "reed", "saxophone", "bright", "solo", "high"],
    description: "Bright, high sax",
    id: "presets/66c5d885-b784-40f5-a134-451c7627a71d"
  },
  {
    program: 65,
    slug: "alto-sax",
    displayName: "Alto Sax",
    category: "Wind",
    tags: ["wind", "reed", "saxophone", "jazz", "warm", "smooth", "middle"],
    description: "Smooth middle sax",
    id: "presets/46af8d73-1844-4b27-af60-89f3407a4d50"
  },
  {
    program: 66,
    slug: "tenor-sax",
    displayName: "Tenor Sax",
    category: "Wind",
    tags: ["wind", "reed", "saxophone", "warm", "jazz", "deep"],
    description: "Deep expressive sax",
    id: "presets/96533625-f1c1-4762-8209-b055f6f678b5"
  },
  {
    program: 67,
    slug: "baritone-sax",
    displayName: "Baritone Sax",
    category: "Wind",
    tags: ["wind", "reed", "saxophone", "low", "brass"],
    description: "Fat low sax sound",
    id: "presets/11566846-86fd-48b6-b94f-562d4e53daf9"
  },
  {
    program: 68,
    slug: "oboe",
    displayName: "Oboe",
    category: "Wind",
    tags: ["wind", "reed", "woodwind", "oboe", "solo", "nasal"],
    description: "Nasal woodwind tone",
    id: "presets/dd0a0ca3-7449-4728-86c4-53c8fdae2384"
  },
  {
    program: 69,
    slug: "english-horn",
    displayName: "English Horn",
    category: "Wind",
    tags: ["wind", "reed", "woodwind", "warm", "solo", "mellow", "oboe", "horn"],
    description: "Mellow oboe variant",
    id: "presets/51f8a3e0-fdeb-460b-8b65-9473f158058c"
  },
  {
    program: 70,
    slug: "bassoon",
    displayName: "Bassoon",
    category: "Wind",
    tags: ["wind", "reed", "woodwind", "low", "classical", "deep"],
    description: "Deep reed woodwind",
    id: "presets/4232177f-ab6c-435c-b73d-3bfd6ae4f567"
  },
  {
    program: 71,
    slug: "clarinet",
    displayName: "Clarinet",
    category: "Wind",
    tags: ["wind", "reed", "woodwind", "clarinet", "warm", "smooth"],
    description: "Smooth woodwind",
    id: "presets/23b891f7-e80b-47d3-a861-2c97f6e5e4d9"
  },
  {
    program: 72,
    slug: "piccolo",
    displayName: "Piccolo",
    category: "Wind",
    tags: ["wind", "pipe", "flute", "bright", "orchestral", "high"],
    description: "Bright, high flute",
    id: "presets/b532b371-6021-46dd-a4af-5588684ab77d"
  },
  {
    program: 73,
    slug: "flute",
    displayName: "Flute",
    category: "Wind",
    tags: ["wind", "pipe", "flute", "woodwind", "solo", "breathy", "clear"],
    description: "Clear breathy tone",
    id: "presets/defb2e3b-d2e0-4eb9-8d71-7df36d31ec3d"
  },
  {
    program: 74,
    slug: "recorder",
    displayName: "Recorder",
    category: "Wind",
    tags: ["wind", "pipe", "folk", "flute", "soft", "sweet"],
    description: "Sweet folk flute",
    id: "presets/f7ff13a4-a825-4bf4-a06e-ff5f081575ce"
  },
  {
    program: 75,
    slug: "pan-flute",
    displayName: "Pan Flute",
    category: "Wind",
    tags: ["wind", "pipe", "folk", "panflute", "world", "airy", "bamboo", "flute"],
    description: "Airy bamboo pipes",
    id: "presets/0316febc-3ff1-45ef-a369-8a95806f2646"
  },
  {
    program: 76,
    slug: "blown-bottle",
    displayName: "Blown Bottle",
    category: "Wind",
    tags: ["ambient", "wind", "fx", "pipe", "whistle", "airy"],
    description: "Airy whistle tone",
    id: "presets/edf04b25-728d-47ce-8a77-c455f6ae9cd4"
  },
  {
    program: 77,
    slug: "shakuhachi",
    displayName: "Shakuhachi",
    category: "Wind",
    tags: ["wind", "pipe", "world", "asian", "flute", "japanese", "bamboo"],
    description: "Japanese bamboo flute",
    id: "presets/be0b2d05-abbf-44c9-b403-c4ab921cfec0"
  },
  {
    program: 78,
    slug: "whistle",
    displayName: "Whistle",
    category: "Wind",
    tags: ["fx", "bright", "high", "pipe", "wind", "whistle", "clear", "sharp"],
    description: "Clear sharp whistle",
    id: "presets/98187def-976c-4650-85dc-02e0d8bcc092"
  },
  {
    program: 79,
    slug: "ocarina",
    displayName: "Ocarina",
    category: "Wind",
    tags: ["folk", "ocarina", "world", "pipe", "wind", "flute", "soft"],
    description: "Soft clay flute",
    id: "presets/b8ccc37a-395b-4a5c-accf-88798594d34a"
  },
  {
    program: 80,
    slug: "square-lead",
    displayName: "Square Lead",
    category: "Lead",
    tags: ["synth", "lead", "retro", "square"],
    description: "Basic square wave lead",
    id: "presets/8d165833-0fb5-4538-b357-417e0d21a592"
  },
  {
    program: 81,
    slug: "saw-lead",
    displayName: "Saw Lead",
    category: "Lead",
    tags: ["synth", "lead", "bright", "sawtooth"],
    description: "Classic saw lead",
    id: "presets/1f7116dd-c242-43e1-9ab9-fc5f706a2158"
  },
  {
    program: 82,
    slug: "calliope-lead",
    displayName: "Calliope Lead",
    category: "Lead",
    tags: ["lead", "synth", "whimsical", "organ"],
    description: "Organ-like synth",
    id: "presets/bdb78d0b-d83d-43a3-8f79-b66a142c929f"
  },
  {
    program: 83,
    slug: "chiffer-lead",
    displayName: "Chiffer Lead",
    category: "Lead",
    tags: ["lead", "synth", "airy", "soft"],
    description: "Breath noise synth",
    id: "presets/7ce36f23-7b02-49fc-8bd1-ca7e12585eb8"
  },
  {
    program: 84,
    slug: "charang",
    displayName: "Charang",
    category: "Lead",
    tags: ["lead", "synth", "gritty", "guitar"],
    description: "Guitar-like synth",
    id: "presets/7e01138c-e98d-4f62-a03d-3a316d7d5305"
  },
  {
    program: 85,
    slug: "solo-voice",
    displayName: "Solo Voice",
    category: "Lead",
    tags: ["lead", "synth", "vocals", "voices"],
    description: "Vocal-ish lead tone",
    id: "presets/a76b3b24-03fa-4873-970f-7557ffa88749"
  },
  {
    program: 86,
    slug: "fifth-sawtooth",
    displayName: "Fifth Sawtooth",
    category: "Lead",
    tags: ["lead", "synth", "harmonic", "sawtooth"],
    description: "Dual oscillator lead",
    id: "presets/025764de-ff81-4362-8414-5aafe1926a43"
  },
  {
    program: 87,
    slug: "bass-lead",
    displayName: "Bass & Lead",
    category: "Bass",
    tags: ["bass", "synth", "lead"],
    description: "Hybrid lead/bass",
    id: "presets/52e2e11e-ddb9-4785-9ffc-82dbf7af478a"
  },
  {
    program: 88,
    slug: "fantasia-pad",
    displayName: "Fantasia Pad",
    category: "Pad",
    tags: ["pad", "synth", "ambient", "evolving", "soft", "motion", "evolving"],
    description: "Soft evolving pad",
    id: "presets/0409f236-80ba-4a36-aaf7-d886604ff39c"
  },
  {
    program: 89,
    slug: "warm-pad",
    displayName: "Warm Pad",
    category: "Pad",
    tags: ["synth", "pad", "warm", "gentle", "analog"],
    description: "Gentle analog pad",
    id: "presets/060c37fe-4a10-4095-9530-3dc6c1c428c7"
  },
  {
    program: 90,
    slug: "polysynth",
    displayName: "Polysynth",
    category: "Pad",
    tags: ["synth", "pad", "thick", "layered"],
    description: "Rich layered pad",
    id: "presets/f951bad2-f378-4c1c-b5e5-f71e4105951f"
  },
  {
    program: 91,
    slug: "space-voice",
    displayName: "Space Voice",
    category: "Pad",
    tags: ["pad", "synth", "choir", "vocals", "voices"],
    description: "Vocal-like pad",
    id: "presets/a7d023a7-43b1-4503-86ff-412cff0a0fc6"
  },
  {
    program: 92,
    slug: "bowed-glass",
    displayName: "Bowed Glass",
    category: "Pad",
    tags: ["pad", "synth", "strings", "glassy"],
    description: "Bowed string pad",
    id: "presets/39811b37-507d-46a0-bc2f-8f127606e0d4"
  },
  {
    program: 93,
    slug: "metal",
    displayName: "Metal",
    category: "Pad",
    tags: ["pad", "synth", "metallic", "bells"],
    description: "Bell-like pad",
    id: "presets/c30af7a6-dca0-4948-8cd4-979ec0ee1b34"
  },
  {
    program: 94,
    slug: "halo",
    displayName: "Halo",
    category: "Pad",
    tags: ["pad", "synth", "bright", "airy"],
    description: "Airy shimmering pad",
    id: "presets/95faa5c8-51ae-434b-b34c-2a9a4e2f2443"
  },
  {
    program: 95,
    slug: "sweep",
    displayName: "Sweep",
    category: "Pad",
    tags: ["pad", "synth", "motion", "filtered", "sweeping"],
    description: "Filter-sweep pad",
    id: "presets/cddab967-6a0c-4c2e-8a51-8e308e53ef99"
  },
  {
    program: 96,
    slug: "rain",
    displayName: "Rain",
    category: "FX",
    tags: ["fx", "ambient", "rain", "nature", "texture"],
    description: "Nature-like pad",
    id: "presets/7b993ed1-7075-41e3-b51c-adbf2674f3b9"
  },
  {
    program: 97,
    slug: "soundtrack",
    displayName: "Soundtrack",
    category: "Pad",
    tags: ["fx", "cinematic", "pad", "synth", "texture"],
    description: "Cinematic pad",
    id: "presets/a22397a4-68be-4b16-8bd2-d58ce43d0c4d"
  },
  {
    program: 98,
    slug: "crystal",
    displayName: "Crystal",
    category: "Keys",
    tags: ["keys", "fx", "bright", "bells", "glassy", "crystal"],
    description: "Glassy sparkle",
    id: "presets/479c325e-e2f1-4f12-98d6-27ddc6306216"
  },
  {
    program: 99,
    slug: "atmosphere",
    displayName: "Atmosphere",
    category: "Pad",
    tags: ["keys", "fx", "space", "ambient", "pad", "texture", "motion"],
    description: "Evolving space pad",
    id: "presets/285a7dc1-cc01-4277-9dcd-a29a40bb5a60"
  },
  {
    program: 100,
    slug: "brightness",
    displayName: "Brightness",
    category: "Pad",
    tags: ["fx", "bright", "synth", "pad"],
    description: "High resonant pad",
    id: "presets/d96df2aa-9a77-40b7-a439-d720f2252618"
  },
  {
    program: 101,
    slug: "goblins",
    displayName: "Goblins",
    category: "FX",
    tags: ["fx", "weird", "creepy", "synth", "strange"],
    description: "Weird sound effect",
    id: "presets/41f458e0-f696-4c03-b722-d52a20efced0"
  },
  {
    program: 102,
    slug: "echo-drops",
    displayName: "Echo Drops",
    category: "FX",
    tags: ["fx", "echo", "space", "synth", "delayed", "texture"],
    description: "Delayed texture",
    id: "presets/4ae465e7-8a9a-40a0-a066-13ccfb350ddf"
  },
  {
    program: 103,
    slug: "sci-fi",
    displayName: "Sci-Fi",
    category: "Pad",
    tags: ["keys", "fx", "synth", "space", "pad", "scifi", "sci-fi"],
    description: "Sci-fi-style pad",
    id: "presets/127dca19-bbf4-4104-8e65-d20f8dbba0fd"
  },
  {
    program: 104,
    slug: "sitar",
    displayName: "Sitar",
    category: "Misc",
    tags: ["misc", "world", "indian", "plucked", "plucky", "sitar"],
    description: "Indian plucked string",
    id: "presets/05b6d94a-1a2e-4542-b3e9-da1bc534b975"
  },
  {
    program: 105,
    slug: "banjo",
    displayName: "Banjo",
    category: "Misc",
    tags: ["misc", "folk", "banjo", "country", "strings", "bright"],
    description: "Bright folk string",
    id: "presets/a458046c-59e9-49f6-a628-22876927c897"
  },
  {
    program: 106,
    slug: "shamisen",
    displayName: "Shamisen",
    category: "Misc",
    tags: ["misc", "world", "asian", "plucked", "plucky"],
    description: "Japanese 3-string lute",
    id: "presets/23145472-516d-4f9c-90b4-9f63ff8ce455"
  },
  {
    program: 107,
    slug: "koto",
    displayName: "Koto",
    category: "Misc",
    tags: ["misc", "world", "asian", "string"],
    description: "Japanese zither",
    id: "presets/02ca6de1-1a08-4994-a87c-23c261a43634"
  },
  {
    program: 108,
    slug: "kalimba",
    displayName: "Kalimba",
    category: "Misc",
    tags: ["misc", "world", "african", "percussion", "percussive", "piano", "kalimba"],
    description: "Thumb piano",
    id: "presets/32779b50-653b-49c0-9f35-c5fa17570cbe"
  },
  {
    program: 109,
    slug: "bagpipe",
    displayName: "Bagpipe",
    category: "Misc",
    tags: ["misc", "folk", "bagpipe", "wind", "pipe"],
    description: "Drone pipe sound",
    id: "presets/2aa96f1c-3c98-49c3-b4da-bc19d3ac64ac"
  },
  {
    program: 110,
    slug: "fiddle",
    displayName: "Fiddle",
    category: "Misc",
    tags: ["misc", "folk", "string", "solo", "violin", "strings", "fiddle"],
    description: "Folk violin",
    id: "presets/9880f830-5184-47ec-88bb-0b268613853b"
  },
  {
    program: 111,
    slug: "shanai",
    displayName: "Shanai",
    category: "Misc",
    tags: ["misc", "world", "indian", "wind", "reed", "pipe"],
    description: "Indian reed pipe",
    id: "presets/4ecb6171-87fd-4f63-9adc-c4ccc765ffcc"
  },
  {
    program: 112,
    slug: "tinkle-bells",
    displayName: "Tinkle Bells",
    category: "Perc",
    tags: ["percussion", "percussive", "bells", "bright", "fx"],
    description: "Bright chime",
    id: "presets/677f0ef2-692b-4474-9d11-e1610918d0b0"
  },
  {
    program: 113,
    slug: "agogo",
    displayName: "Agogo",
    category: "Perc",
    tags: ["percussion", "percussive", "latin", "bright", "cowbell"],
    description: "Twin cowbell tones",
    id: "presets/47241bb4-9eff-463d-84b4-620b55d8297b"
  },
  {
    program: 114,
    slug: "steel-drums",
    displayName: "Steel Drums",
    category: "Perc",
    tags: ["percussion", "percussive", "world", "steel", "tropical", "drums", "misc"],
    description: "Caribbean tone",
    id: "presets/84f0b350-e1b1-4ef3-a3a7-3dbee2a8e708"
  },
  {
    program: 115,
    slug: "woodblock",
    displayName: "Woodblock",
    category: "Perc",
    tags: ["percussion", "percussive", "wood", "short", "woodblock"],
    description: "Percussive wood click",
    id: "presets/6023ccdc-fa78-495d-a22f-c9ac9ffabcf9"
  },
  {
    program: 116,
    slug: "taiko-drum",
    displayName: "Taiko Drum",
    category: "Perc",
    tags: ["percussion", "percussive", "world", "taiko", "drums", "japanese", "asian"],
    description: "Deep Japanese drum",
    id: "presets/04bd3b71-a9c8-44c9-8bfe-f2e7182514b2"
  },
  {
    program: 117,
    slug: "melodic-drum",
    displayName: "Melodic Drum",
    category: "Perc",
    tags: ["percussion", "percussive", "drums", "melodic"],
    description: "Tuned toms",
    id: "presets/90ad8394-47cf-4089-88f8-3485d3d420c2"
  },
  {
    program: 118,
    slug: "synth-tom",
    displayName: "Synth Tom",
    category: "Perc",
    tags: ["percussion", "percussive", "synth", "drums", "electronic", "electric"],
    description: "Electronic tom hit",
    id: "presets/34c8332e-d188-4b08-8bab-989ab5027b35"
  },
  {
    program: 119,
    slug: "reverse-cymbal",
    displayName: "Reverse Cymbal",
    category: "Perc",
    tags: ["percussion", "percussive", "fx", "reverse", "cymbal", "drums"],
    description: "Reversed swell",
    id: "presets/92266fda-dfd3-4e16-88c7-1f0160d3b8d4"
  },
  {
    program: 120,
    slug: "guitar-fret-noise",
    displayName: "Guitar Fret Noise",
    category: "FX",
    tags: ["fx", "guitar", "noise", "sliding", "texture"],
    description: "Sliding fret sound",
    id: "presets/4792e4fa-4be4-4c15-ae4e-a1a89238aa6f"
  },
  {
    program: 121,
    slug: "breath-noise",
    displayName: "Breath Noise",
    category: "FX",
    tags: ["fx", "vocals", "air", "texture"],
    description: "Inhale/exhale effect",
    id: "presets/8ef100a5-5462-42bd-af07-5be8ff8b07ef"
  },
  {
    program: 122,
    slug: "seashore",
    displayName: "Seashore",
    category: "FX",
    tags: ["fx", "ambient", "nature", "texture"],
    description: "Ocean wave sound",
    id: "presets/b0163790-e8c1-4f46-848c-9cb5f5cb08b5"
  },
  {
    program: 123,
    slug: "bird-tweet",
    displayName: "Bird Tweet",
    category: "FX",
    tags: ["fx", "nature", "ambient", "texture"],
    description: "Bird chirping",
    id: "presets/0bed8f0f-f854-4100-b003-7be900bbed8c"
  },
  {
    program: 124,
    slug: "telephone-ring",
    displayName: "Telephone Ring",
    category: "FX",
    tags: ["fx", "phone", "retro"],
    description: "Old-style ring",
    id: "presets/de45f831-dca8-4bfe-b5d6-c7bf2a58c38b"
  },
  {
    program: 125,
    slug: "helicopter",
    displayName: "Helicopter",
    category: "FX",
    tags: ["fx", "machine", "transport"],
    description: "Rotor blade sound",
    id: "presets/7208226b-5271-4852-beb7-f0194bacb40d"
  },
  {
    program: 126,
    slug: "applause",
    displayName: "Applause",
    category: "FX",
    tags: ["fx", "crowd", "human", "texture"],
    description: "Audience clapping",
    id: "presets/d39b7b86-42a0-47fc-99ea-6b38a8d890a0"
  },
  {
    program: 127,
    slug: "gunshot",
    displayName: "Gunshot",
    category: "FX",
    tags: ["fx", "weapon", "impact"],
    description: "Pistol/bang sound",
    id: "presets/e859bb19-7e48-4831-8265-744446cef771"
  }
], Ee = [
  {
    program: 0,
    slug: "standard-kit",
    displayName: "Standard Drum Kit",
    category: "Drums",
    tags: ["drums", "clean", "soft"],
    id: "presets/6b49e6b6-09f1-43d0-b15d-adeccdb9eeab"
  },
  {
    program: 8,
    slug: "room-kit",
    displayName: "Room Drum Kit",
    category: "Drums",
    tags: ["drums", "clean", "effected"],
    id: "presets/284daead-69b1-459d-83f6-1e08d6d8cee9"
  },
  {
    program: 16,
    slug: "power-kit",
    displayName: "Power Drum Kit",
    category: "Drums",
    tags: ["drums", "hard", "aggressive"],
    id: "presets/e3ce588c-f237-481d-9f87-be3d423b29ae"
  },
  {
    program: 24,
    slug: "electronic-kit",
    displayName: "Electronic Drum Kit",
    category: "Drums",
    tags: ["drums", "electronic", "hard"],
    id: "presets/18aa38cd-5fc5-4855-b7a9-7b5fca4f2dcd"
  },
  {
    program: 25,
    slug: "analog-kit",
    displayName: "Analog Drum Kit",
    category: "Drums",
    tags: ["drums", "electronic", "soft"],
    id: "presets/d6c69f5f-8e34-424d-8927-1a9ffd68e06a"
  },
  {
    program: 32,
    slug: "jazz-kit",
    displayName: "Jazz Drum Kit",
    category: "Drums",
    tags: ["drums", "acoustic", "soft", "jazz"],
    id: "presets/56f936cf-452d-4d5e-bf03-28da021f0025"
  },
  {
    program: 40,
    slug: "brush-kit",
    displayName: "Brush Drum Kit",
    category: "Drums",
    tags: ["drums", "acoustic", "soft"],
    id: "presets/c9973261-780e-445f-b8da-c43fcc422574"
  },
  {
    program: 48,
    slug: "orchestra-kit",
    displayName: "Orchestra Drum Kit",
    category: "Drums",
    tags: ["drums", "orchestral", "clean", "classical"],
    id: "presets/42e00fe8-e6f0-41e2-8982-174ac70d5e7d"
  }
], we = {
  "acoustic-piano": 0,
  "bright-grand-2": 1,
  "piano-2": 2,
  "honky-tonk-piano": 3,
  "electronic-piano-1": 4,
  "electronic-piano-2": 5,
  harpsichord: 6,
  clavinet: 7,
  celesta: 8,
  glockenspiel: 9,
  "music-box": 10,
  vibraphone: 11,
  marimba: 12,
  xylophone: 13,
  "tubular-bells": 14,
  dulcimer: 15,
  "jazz-organ": 16,
  "hammond-organ": 17,
  "rock-organ": 18,
  "church-organ": 19,
  "reed-organ": 20,
  accordion: 21,
  harmonica: 22,
  bandoneon: 23,
  "nylon-guitar": 24,
  "dark-steel-guitar": 25,
  "jazz-guitar": 26,
  "clean-guitar": 27,
  "muted-guitar": 28,
  "overdriven-guitar": 29,
  "distortion-guitar": 30,
  "guitar-harmonics": 31,
  "acoustic-bass": 32,
  "fingered-bass": 33,
  "picked-bass": 34,
  "fretless-bass": 35,
  "slap-bass-1": 36,
  "slap-bass-2": 37,
  "synth-bass-1": 38,
  "synth-bass-2": 39,
  violin: 40,
  viola: 41,
  cello: 42,
  contrabass: 43,
  "tremolo-strings": 44,
  "pizzicato-strings": 45,
  harp: 46,
  timpani: 47,
  "string-section": 48,
  "string-ensemble": 49,
  "synth-strings-1": 50,
  "synth-strings-2": 51,
  "choir-aahs": 52,
  "choir-oohs": 53,
  "synth-voice": 54,
  "orchestra-hit": 55,
  trumpet: 56,
  trombone: 57,
  tuba: 58,
  "muted-trumpet": 59,
  "french-horn": 60,
  "brass-section": 61,
  "synth-brass-1": 62,
  "synth-brass-2": 63,
  "soprano-sax": 64,
  "alto-sax": 65,
  "tenor-sax": 66,
  "baritone-sax": 67,
  oboe: 68,
  "english-horn": 69,
  bassoon: 70,
  clarinet: 71,
  piccolo: 72,
  flute: 73,
  recorder: 74,
  "pan-flute": 75,
  "blown-bottle": 76,
  shakuhachi: 77,
  whistle: 78,
  ocarina: 79,
  "square-lead": 80,
  "saw-lead": 81,
  "calliope-lead": 82,
  "chiffer-lead": 83,
  charang: 84,
  "solo-voice": 85,
  "fifth-sawtooth": 86,
  "bass-lead": 87,
  "fantasia-pad": 88,
  "warm-pad": 89,
  polysynth: 90,
  "space-voice": 91,
  "bowed-glass": 92,
  metal: 93,
  halo: 94,
  sweep: 95,
  rain: 96,
  soundtrack: 97,
  crystal: 98,
  atmosphere: 99,
  brightness: 100,
  goblins: 101,
  "echo-drops": 102,
  "sci-fi": 103,
  sitar: 104,
  banjo: 105,
  shamisen: 106,
  koto: 107,
  kalimba: 108,
  bagpipe: 109,
  fiddle: 110,
  shanai: 111,
  "tinkle-bells": 112,
  agogo: 113,
  "steel-drums": 114,
  woodblock: 115,
  "taiko-drum": 116,
  "melodic-drum": 117,
  "synth-tom": 118,
  "reverse-cymbal": 119,
  "guitar-fret-noise": 120,
  "breath-noise": 121,
  seashore: 122,
  "bird-tweet": 123,
  "telephone-ring": 124,
  helicopter: 125,
  applause: 126,
  gunshot: 127
}, Se = {
  "standard-kit": 0,
  "room-kit": 8,
  "power-kit": 16,
  "electronic-kit": 24,
  "analog-kit": 25,
  "jazz-kit": 32,
  "brush-kit": 40,
  "orchestra-kit": 48
}, _e = (e) => {
  const a = N(ne, e), s = async (r) => {
    const o = await a.getPreset({
      name: `presets/${U(r)}`
    });
    if (o instanceof Error)
      throw o;
    if (o.preset === void 0)
      throw new Error(`No preset found for UUID "${U(r)}"`);
    return H(o.preset);
  };
  return {
    search: async (r, o) => {
      const i = `preset.device_type == '${ke[r]}'`;
      o !== void 0 && o.length > 1 && (o = `${o.replace(/['"]/g, "").toLowerCase().split(" ").filter((c) => c.length > 1).join(" & ")}`);
      const d = await a.listPresets({
        filter: i,
        textSearch: o
      });
      if (d instanceof Error)
        throw d;
      return await Promise.all(
        d.presets.map(
          (l) => H(l)
        )
      );
    },
    get: s,
    getInstrument: async (r) => {
      const o = typeof r == "number" ? r : typeof r == "string" ? we[r] : r.program, i = me[o];
      if (i === void 0)
        throw new Error(
          `No GM instrument preset for program ${String(r)}. Expected 0..127, a GmInstrumentSlug, or a GmInstrument.`
        );
      return await s(i);
    },
    getDrums: async (r) => {
      const o = typeof r == "number" ? r : typeof r == "string" ? Se[r] : r.program, i = ye[o];
      if (i === void 0)
        throw new Error(
          `No GM drum kit preset for program ${String(r)}. Expected 0, 8, 16, 24, 25, 32, 40, 48, a GmDrumSlug, or a GmDrum.`
        );
      return await s(i);
    },
    gmInstruments: he,
    gmDrums: Ee
  };
}, ke = {
  autoFilter: "PRESET_DEVICE_TYPE_AUTO_FILTER",
  bandSplitter: "PRESET_DEVICE_TYPE_BAND_SPLITTER",
  bassline: "PRESET_DEVICE_TYPE_BASSLINE",
  beatbox8: "PRESET_DEVICE_TYPE_BEATBOX8",
  beatbox9: "PRESET_DEVICE_TYPE_BEATBOX9",
  crossfader: "PRESET_DEVICE_TYPE_CROSSFADER",
  curve: "PRESET_DEVICE_TYPE_CURVE",
  exciter: "PRESET_DEVICE_TYPE_EXCITER",
  graphicalEQ: "PRESET_DEVICE_TYPE_GRAPHICAL_EQ",
  gravity: "PRESET_DEVICE_TYPE_GRAVITY",
  heisenberg: "PRESET_DEVICE_TYPE_HEISENBERG",
  helmholtz: "PRESET_DEVICE_TYPE_HELMHOLTZ",
  machiniste: "PRESET_DEVICE_TYPE_MACHINISTE",
  matrixArpeggiator: "PRESET_DEVICE_TYPE_MATRIX",
  gakki: "PRESET_DEVICE_TYPE_GAKKI",
  noteSplitter: "PRESET_DEVICE_TYPE_NOTE_SPLITTER",
  panorama: "PRESET_DEVICE_TYPE_PANORAMA",
  pulsar: "PRESET_DEVICE_TYPE_PULSAR",
  pulverisateur: "PRESET_DEVICE_TYPE_PULVERISATEUR",
  quantum: "PRESET_DEVICE_TYPE_QUANTUM",
  quasar: "PRESET_DEVICE_TYPE_QUASAR",
  rasselbock: "PRESET_DEVICE_TYPE_RASSELBOCK",
  space: "PRESET_DEVICE_TYPE_SPACE",
  stereoEnhancer: "PRESET_DEVICE_TYPE_STEREO_ENHANCER",
  stompboxChorus: "PRESET_DEVICE_TYPE_STOMPBOX_CHORUS",
  stompboxCompressor: "PRESET_DEVICE_TYPE_STOMPBOX_COMPRESSOR",
  stompboxCrusher: "PRESET_DEVICE_TYPE_STOMPBOX_CRUSHER",
  stompboxDelay: "PRESET_DEVICE_TYPE_STOMPBOX_DELAY",
  stompboxFlanger: "PRESET_DEVICE_TYPE_STOMPBOX_FLANGER",
  stompboxGate: "PRESET_DEVICE_TYPE_STOMPBOX_GATE",
  stompboxParametricEqualizer: "PRESET_DEVICE_TYPE_STOMPBOX_PARAMETRIC_EQUALIZER",
  stompboxPhaser: "PRESET_DEVICE_TYPE_STOMPBOX_PHASER",
  stompboxPitchDelay: "PRESET_DEVICE_TYPE_STOMPBOX_PITCH_DELAY",
  stompboxReverb: "PRESET_DEVICE_TYPE_STOMPBOX_REVERB",
  stompboxSlope: "PRESET_DEVICE_TYPE_STOMPBOX_SLOPE",
  stompboxStereoDetune: "PRESET_DEVICE_TYPE_STOMPBOX_STEREO_DETUNE",
  stompboxTube: "PRESET_DEVICE_TYPE_STOMPBOX_TUBE",
  tonematrix: "PRESET_DEVICE_TYPE_TONEMATRIX",
  waveshaper: "PRESET_DEVICE_TYPE_WAVESHAPER"
}, H = async (e) => {
  const a = e, s = await fetch(e.dataUrl, {
    credentials: "omit"
  }).then((d) => d.arrayBuffer()), r = new q();
  if (!se.fromBinary(new Uint8Array(s)).unpackTo(r))
    throw new Error("Failed to unpack preset data");
  if (r.target === void 0)
    throw new Error("Preset data does not contain a target");
  const i = te(r.target);
  if (i === void 0)
    throw new Error("Preset data does not contain an entity");
  return {
    meta: a,
    data: r,
    entityType: i,
    _presetName: e.name
  };
}, Te = (e, a = 5e3) => {
  const s = /* @__PURE__ */ new Map();
  let r = new AbortController();
  const o = async () => {
    r.abort(), r = new AbortController();
    const i = r;
    if (s.size === 0)
      return;
    const d = e.listen(
      { names: [...s.keys()] },
      { signal: i.signal }
    )[Symbol.asyncIterator]();
    for (; ; ) {
      if (i.signal.aborted)
        return;
      let l;
      try {
        l = await d.next();
      } catch (f) {
        if (i.signal.aborted || s.size === 0)
          return;
        console.warn(
          `sampleService.listen iterator error, retrying in ${a}ms`,
          f
        ), await Y(a), o();
        return;
      }
      if (l.done ?? !1) {
        if (s.size === 0)
          return;
        console.warn(
          `sampleService.listen iterator done, retrying in ${a}ms`
        ), await Y(a), o();
        return;
      }
      const t = l.value.operation;
      if (t === void 0) {
        console.warn("received empty sampleService.listen operation, ignoring");
        continue;
      }
      const c = ve(t.metadata);
      if (c === void 0 || c.sample === void 0) {
        console.warn(
          "received sampleService.listen operation without valid SampleConvertInfo, ignoring"
        );
        continue;
      }
      const n = c.sample.name, p = s.get(n);
      if (p !== void 0 && t.done) {
        if (t.result.case === "error") {
          if (p(
            new Error(`Operation error: ${t.result.value.message}`)
          ), s.delete(n), s.size === 0)
            return;
          continue;
        }
        if (c.error !== V.NONE && c.error !== V.UNSPECIFIED) {
          if (p(
            new Error(
              `SampleConvertDone event with error: ${V[c.error] ?? c.error}`
            )
          ), s.delete(n), s.size === 0)
            return;
          continue;
        }
        if (p(c.sample), s.delete(n), s.size === 0)
          return;
      }
    }
  };
  return {
    listenForSample: async (i, d) => {
      if (s.has(i))
        return new Error(
          `listenForSample called multiple times for sampleId: ${i}`
        );
      if (d != null && d.aborted)
        return new Error("Aborted");
      const { promise: l, resolve: t } = Promise.withResolvers();
      return s.set(i, t), d == null || d.addEventListener(
        "abort",
        () => {
          s.has(i) && (s.delete(i), o(), t(new Error("Aborted")));
        },
        { once: !0 }
      ), o(), l;
    }
  };
}, Pe = K(M), ve = (e) => {
  if (e !== void 0 && e.is(M))
    return e.unpack(Pe);
}, Ne = {
  "one-shot": R.ONE_SHOT,
  loop: R.LOOP
}, xe = {
  [R.UNSPECIFIED]: "one-shot",
  [R.ONE_SHOT]: "one-shot",
  [R.LOOP]: "loop"
}, Ce = {
  public: D.PUBLIC,
  unlisted: D.UNLISTED
}, De = {
  [D.UNSPECIFIED]: "unlisted",
  [D.PUBLIC]: "public",
  [D.UNLISTED]: "unlisted"
}, X = (e) => {
  if (e instanceof oe)
    return e.fields.sampleName.value;
  if (typeof e == "string") {
    const a = U(e);
    return a instanceof Error ? e.startsWith("samples/") ? e : `samples/${e}` : `samples/${a}`;
  }
  return e.name;
}, J = (e) => {
  var a, s;
  return {
    name: e.name,
    displayName: e.displayName,
    description: e.description,
    ownerName: e.ownerName,
    favoritedByUser: e.favoritedByUser,
    numFavorites: e.numFavorites,
    numUsages: e.numUsages,
    bpm: e.bpm,
    kind: xe[e.sampleType],
    visibility: De[e.usage],
    tags: e.tags,
    createTime: (a = e.createTime) == null ? void 0 : a.toDate(),
    updateTime: (s = e.updateTime) == null ? void 0 : s.toDate()
  };
}, Re = "3840_both.json.gz", Ie = (e) => (a) => {
  if (e === "") return "";
  const s = (a == null ? void 0 : a.resolution) ?? 1920, r = (a == null ? void 0 : a.channel) ?? "both", o = `${s}_${r}.json.gz`;
  return e.replace(Re, o);
}, z = (e) => ({
  ...J(e),
  durationSeconds: e.playDuration ? Number(e.playDuration.seconds) + e.playDuration.nanos / 1e9 : 0,
  mp3Url: e.mp3Url,
  wavUrl: e.wavUrl,
  flacUrl: e.flacUrl,
  previewMp3Url: e.previewMp3Url,
  getWaveformUrl: Ie(e.waveformUrl)
}), Be = (e) => {
  const a = N(de, e), s = Te(a), r = async (t, c) => {
    const n = X(t), p = await a.getSample({ name: n }, { signal: c });
    return p instanceof Error ? p : p.sample ? z(p.sample) : new Error(`Sample not found: ${n}`);
  };
  return {
    upload: async (t, c) => {
      const n = t.file instanceof ArrayBuffer ? new Blob([t.file], { type: "application/octet-stream" }) : t.file, p = new le({
        displayName: t.displayName,
        description: t.description ?? "",
        bpm: t.bpm ?? 0,
        sampleType: Ne[t.kind ?? "one-shot"],
        usage: Ce[t.visibility ?? "unlisted"],
        tags: t.tags ?? ["sample"],
        clearance: pe.UNSAFE
      }), f = await a.createSample(
        { sample: p },
        { signal: c }
      );
      if (f instanceof Error)
        return f;
      const { sample: E, uploadEndpoint: k } = f;
      if (!E)
        return new Error("Server did not return sample metadata");
      if (!k)
        return new Error("Server did not return upload endpoint");
      const u = J(E), _ = new AbortController(), g = _.signal;
      c && c.addEventListener("abort", () => _.abort(), {
        once: !0
      });
      let b = !1;
      const w = (m) => {
        if (!b && t.preventTabClose)
          return m.preventDefault(), m.returnValue = "", "";
      }, x = typeof window < "u" && typeof document < "u";
      x && t.preventTabClose && window.addEventListener("beforeunload", w);
      const B = () => {
        x && t.preventTabClose && window.removeEventListener("beforeunload", w);
      };
      let C = () => {
      };
      const S = new Promise((m) => {
        C = m;
      });
      let y = !1;
      const h = (m) => {
        y || (y = !0, C(m));
      }, re = (async () => {
        try {
          const m = await v(
            k.uploadUrl,
            {
              method: "PUT",
              headers: k.headers,
              body: n,
              signal: g
            }
          );
          if (m instanceof Error)
            return await a.cancelSampleUpload(
              { name: E.name },
              { signal: g }
            ), h(m), m;
          if (!m.ok) {
            await a.cancelSampleUpload(
              { name: E.name },
              { signal: g }
            );
            const O = new Error(`Upload failed: ${m.statusText}`);
            return h(O), O;
          }
          const T = await a.uploadSampleFinished(
            { name: E.name },
            { signal: g }
          );
          if (T instanceof Error)
            return h(T), T;
          if (h(void 0), T.done)
            return T.result.case === "error" ? new Error(
              `Processing failed: ${T.result.value.message}`
            ) : await r(E.name, g);
          const A = await s.listenForSample(
            E.name,
            g
          );
          return A instanceof Error ? new Error(`Processing failed: ${A.message}`, {
            cause: A
          }) : z(A);
        } catch (m) {
          const T = m instanceof Error ? m : new Error("Sample upload failed", { cause: m });
          return h(T), T;
        } finally {
          h(
            new Error("Sample upload ended without resolving `uploaded`")
          ), b = !0, B();
        }
      })();
      return {
        ...u,
        uploaded: S,
        ready: re
      };
    },
    download: async (t, c, n) => {
      const p = (c == null ? void 0 : c.format) ?? "flac", f = await r(t, n);
      if (f instanceof Error)
        return f;
      const k = {
        wav: f.wavUrl,
        mp3: f.mp3Url,
        flac: f.flacUrl,
        preview: f.previewMp3Url
      }[p];
      if (!k)
        return new Error(
          `Sample ${f.name} does not have a ${p} URL available yet`
        );
      const u = await v(k, { signal: n });
      if (u instanceof Error)
        return u;
      if (!u.ok)
        return new Error(`Failed to download sample: ${u.statusText}`);
      try {
        return await u.blob();
      } catch (_) {
        return new Error("Failed to read sample data", { cause: _ });
      }
    },
    get: r,
    list: async (t, c) => {
      const n = await a.listSamples(
        {
          filter: (t == null ? void 0 : t.filter) ?? "",
          textSearch: (t == null ? void 0 : t.textSearch) ?? "",
          orderBy: (t == null ? void 0 : t.orderBy) ?? "",
          pageSize: (t == null ? void 0 : t.pageSize) ?? 20,
          pageToken: (t == null ? void 0 : t.pageToken) ?? ""
        },
        { signal: c }
      );
      return n instanceof Error ? n : {
        samples: n.samples.map(z),
        nextPageToken: n.nextPageToken
      };
    },
    delete: async (t, c) => {
      const n = X(t), p = await a.deleteSample({ name: n }, { signal: c });
      if (p instanceof Error)
        return p;
    }
  };
}, Ae = (e) => {
  const a = e.startsWith("Bearer ") ? e : `Bearer ${e}`;
  return {
    getToken: async () => a
  };
}, Ue = (e) => {
  const a = new Headers(e), s = a.get(G) === "true";
  return a.delete(G), { headers: a, keepalive: s };
}, $ = () => ({
  createTransport: async (e) => {
    const { createConnectTransport: a } = await import("@connectrpc/connect-web");
    return a({
      baseUrl: e.baseUrl,
      useBinaryFormat: e.useBinaryFormat,
      fetch: async (s, r) => {
        const { headers: o, keepalive: i } = Ue(r == null ? void 0 : r.headers);
        return o.set("Authorization", await e.getToken()), fetch(s, {
          credentials: "omit",
          ...r,
          headers: o,
          keepalive: i
        });
      },
      jsonOptions: {
        typeRegistry: e.typeRegistry
      }
    });
  }
}), W = async ({
  getToken: e,
  transportFactory: a
}) => {
  const s = await a.createTransport({
    baseUrl: "https://rpc.audiotool.com/",
    useBinaryFormat: !1,
    typeRegistry: K(q, M),
    getToken: e
  }), r = N(ue, s), o = N(
    be,
    s
  ), i = N(fe, s), d = Be(s), l = _e(s), t = N(
    ge,
    s
  );
  return {
    projects: r,
    projectRoles: o,
    users: i,
    samples: d,
    presets: l,
    audioGraph: t,
    authorizedFetch: async (n, p) => v(n, {
      credentials: "omit",
      ...p
    }),
    fetch: v,
    open: async (n) => {
      const p = Oe(n);
      return await ie(
        r,
        p,
        e,
        a
      );
    }
  };
}, qe = async ({
  auth: e,
  transport: a,
  wasm: s
}) => {
  const r = typeof e == "string" ? Ae(e) : e, o = a ?? $();
  return s && ce(s), W({
    getToken: () => r.getToken(),
    transportFactory: o
  });
}, Oe = (e) => {
  const a = U(e);
  if (a instanceof Error)
    throw new Error(
      `couldn't extract project uuid from string: ${e}, should be URL/UUID/project name`
    );
  return `projects/${a}`;
}, Ve = "https://oauth.audiotool.com/oauth2/auth", j = "https://oauth.audiotool.com/oauth2/token", ze = "https://rpc.audiotool.com", Fe = "https://accounts.audiotool.com/oauth/popup/start", Q = "https://accounts.audiotool.com", Le = "audiotool-oauth", Me = "popup,width=480,height=720,noopener=no", Je = async ({
  clientId: e,
  redirectUrl: a,
  scope: s
}) => {
  const r = {
    accessToken: `oidc_${e}_oidc_access_token`,
    refreshToken: `oidc_${e}_oidc_refresh_token`,
    expiresAt: `oidc_${e}_oidc_expires_at`,
    codeVerifier: `oidc_${e}_oidc_code_verifier`,
    userName: `oidc_${e}_oidc_user_name`,
    state: `oidc_${e}_oidc_state`
  }, o = () => {
    I(r), $e(r, e, s, a);
  }, i = new URLSearchParams(window.location.search), d = i.get("error"), l = i.get("error_description");
  if (d != null)
    return F(), I(r), {
      status: "unauthenticated",
      login: o,
      error: P(d, l)
    };
  const t = i.get("code");
  if (t != null) {
    const g = i.get("state"), b = localStorage.getItem(r.state);
    if (g == null || b == null || g !== b)
      return {
        status: "unauthenticated",
        login: o,
        error: P(
          "Invalid state URL parameter.",
          "Clean the URL of stale query parameters and try again."
        )
      };
    const w = localStorage.getItem(r.codeVerifier);
    if (w == null)
      return {
        status: "unauthenticated",
        login: o,
        error: P("Code verifier not found. Restart the auth flow.")
      };
    const x = await fetch(j, {
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded"
      },
      body: new URLSearchParams({
        client_id: e,
        grant_type: "authorization_code",
        code: t,
        redirect_uri: a,
        code_verifier: w,
        token_endpoint_auth_method: "none"
      })
    }), {
      access_token: B,
      refresh_token: C,
      expires_in: S,
      error: y,
      error_description: h
    } = await x.json();
    if (y)
      return {
        status: "unauthenticated",
        login: o,
        error: P(y, h)
      };
    localStorage.setItem(r.accessToken, B), localStorage.setItem(r.refreshToken, C), localStorage.setItem(
      r.expiresAt,
      (Date.now() + S * 1e3).toString()
    ), F(), localStorage.removeItem(r.codeVerifier);
  }
  localStorage.removeItem(r.state);
  const c = await L(r, e);
  if (c === void 0)
    return { status: "unauthenticated", login: o };
  if (c instanceof Error)
    return { status: "unauthenticated", login: o, error: c };
  let n;
  const p = async () => {
    if (n === void 0) {
      n = (async () => {
        const w = await L(r, e);
        return w === void 0 ? new Error("User not logged in.") : w;
      })();
      const b = await n;
      if (n = void 0, b instanceof Error)
        throw b;
      return b.startsWith("Bearer ") ? b : `Bearer ${b}`;
    }
    const g = await n;
    if (g instanceof Error)
      throw g;
    return g.startsWith("Bearer ") ? g : `Bearer ${g}`;
  }, f = await Z(p);
  if (f instanceof Error)
    return { status: "unauthenticated", login: o, error: f };
  const E = $();
  return {
    ...await W({
      getToken: p,
      transportFactory: E
    }),
    status: "authenticated",
    userName: f,
    logout: () => {
      I(r), F(), window.location.reload();
    },
    exportTokens: () => {
      const g = localStorage.getItem(r.accessToken), b = localStorage.getItem(r.refreshToken), w = localStorage.getItem(r.expiresAt);
      if (!g || !b || !w)
        throw new Error("Cannot export tokens: not authenticated");
      return {
        accessToken: g,
        refreshToken: b,
        expiresAt: parseInt(w, 10)
      };
    }
  };
}, Z = async (e) => {
  const a = await e(), s = await v(
    `${ze}/audiotool.auth.v1.AuthService/GetWhoami`,
    {
      method: "POST",
      headers: {
        Authorization: a,
        "Content-Type": "application/json"
      },
      body: JSON.stringify({})
    }
  );
  if (s instanceof Error)
    return s;
  if (!s.ok)
    return new Error(`Failed to get user info: ${s.statusText}`);
  const { whoami: r } = await s.json();
  return (r == null ? void 0 : r.userName) ?? "Unknown User";
}, I = (e) => {
  localStorage.removeItem(e.accessToken), localStorage.removeItem(e.refreshToken), localStorage.removeItem(e.expiresAt), localStorage.removeItem(e.codeVerifier), localStorage.removeItem(e.userName), localStorage.removeItem(e.state);
}, F = () => {
  const e = new URL(window.location.href);
  e.searchParams.delete("code"), e.searchParams.delete("scope"), e.searchParams.delete("state"), e.searchParams.delete("error"), e.searchParams.delete("error_description"), window.history.replaceState(
    {},
    document.title,
    e.search ? e.href : e.href.replace("?", "")
  );
}, $e = async (e, a, s, r) => {
  const o = ee();
  localStorage.setItem(e.codeVerifier, o);
  const i = await ae(o), d = crypto.getRandomValues(new Uint8Array(16)).join("");
  localStorage.setItem(e.state, d);
  const l = new URL(Ve);
  l.search = new URLSearchParams({
    response_type: "code",
    client_id: a,
    scope: s,
    code_challenge_method: "S256",
    code_challenge: i,
    redirect_uri: r,
    state: d
  }).toString(), window.location.href = l.toString();
}, P = (e, a) => (a = a ? `: ${a}` : "", new Error(`${e}${a}`)), ee = () => {
  const e = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
  return crypto.getRandomValues(new Uint8Array(64)).reduce(
    (s, r) => s + e[r % e.length],
    ""
  );
}, ae = async (e) => {
  const a = new TextEncoder().encode(e), s = await crypto.subtle.digest("SHA-256", a);
  return btoa(String.fromCharCode(...new Uint8Array(s))).replace(/=/g, "").replace(/\+/g, "-").replace(/\//g, "_");
}, L = async (e, a) => {
  const s = localStorage.getItem(e.accessToken);
  if (s == null)
    return;
  let r = !0;
  {
    const p = localStorage.getItem(e.expiresAt);
    r = p == null ? !0 : Date.now() >= parseInt(p) - 6e4;
  }
  if (!r)
    return s;
  const o = localStorage.getItem(e.refreshToken);
  if (o == null) {
    I(e);
    return;
  }
  const i = await v(j, {
    method: "POST",
    headers: {
      "Content-Type": "application/x-www-form-urlencoded"
    },
    body: new URLSearchParams({
      client_id: a,
      grant_type: "refresh_token",
      refresh_token: o
    })
  });
  if (i instanceof Error)
    return new Error(`Error during refresh token request: ${i.name}`, {
      cause: i.message
    });
  const { error: d, error_description: l, access_token: t, refresh_token: c, expires_in: n } = await i.json();
  return d ? P(d, l) : (localStorage.setItem(e.accessToken, t), localStorage.setItem(e.refreshToken, c), localStorage.setItem(
    e.expiresAt,
    (Date.now() + n * 1e3).toString()
  ), t);
}, Qe = async ({
  clientId: e,
  scope: a,
  targetOrigin: s = window.location.origin,
  windowFeatures: r = Me
}) => {
  const o = window.open("about:blank", Le, r);
  if (o == null)
    return {
      status: "unauthenticated",
      error: new Error(
        "Popup was blocked. audiotoolPopup() must be called from a user-gesture handler."
      )
    };
  try {
    return await We({ popup: o, clientId: e, scope: a, targetOrigin: s });
  } catch (i) {
    return {
      status: "unauthenticated",
      error: i instanceof Error ? i : new Error(String(i))
    };
  }
}, We = async ({
  popup: e,
  clientId: a,
  scope: s,
  targetOrigin: r
}) => {
  const o = ee(), i = await ae(o), d = crypto.getRandomValues(new Uint8Array(16)).join(""), l = new URL(Fe);
  l.search = new URLSearchParams({
    client_id: a,
    scope: s,
    state: d,
    code_challenge: i,
    code_challenge_method: "S256",
    target_origin: r
  }).toString(), e.location.href = l.toString();
  const t = await je(e, d);
  if (t.kind === "cancelled")
    return {
      status: "unauthenticated",
      error: new Error("Popup was closed before login completed.")
    };
  if (t.kind === "error")
    return {
      status: "unauthenticated",
      error: P(t.error, t.errorDescription)
    };
  const c = await fetch(j, {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: new URLSearchParams({
      client_id: a,
      grant_type: "authorization_code",
      code: t.code,
      // For popup-mode the redirect_uri is the accounts popup-callback
      // endpoint; Hydra expects it to match exactly what was sent to the
      // /oauth2/auth call.
      redirect_uri: `${Q}/oauth/popup/callback`,
      code_verifier: o,
      token_endpoint_auth_method: "none"
    })
  }), {
    access_token: n,
    refresh_token: p,
    expires_in: f,
    error: E,
    error_description: k
  } = await c.json();
  if (E)
    return {
      status: "unauthenticated",
      error: P(E, k)
    };
  const u = {
    accessToken: `oidc_${a}_oidc_access_token`,
    refreshToken: `oidc_${a}_oidc_refresh_token`,
    expiresAt: `oidc_${a}_oidc_expires_at`,
    codeVerifier: `oidc_${a}_oidc_code_verifier`,
    userName: `oidc_${a}_oidc_user_name`,
    state: `oidc_${a}_oidc_state`
  };
  localStorage.setItem(u.accessToken, n), localStorage.setItem(u.refreshToken, p), localStorage.setItem(
    u.expiresAt,
    (Date.now() + f * 1e3).toString()
  );
  let _;
  const g = async () => {
    if (_ === void 0) {
      _ = (async () => {
        const h = await L(u, a);
        return h === void 0 ? new Error("User not logged in.") : h;
      })();
      const y = await _;
      if (_ = void 0, y instanceof Error) throw y;
      return y.startsWith("Bearer ") ? y : `Bearer ${y}`;
    }
    const S = await _;
    if (S instanceof Error) throw S;
    return S.startsWith("Bearer ") ? S : `Bearer ${S}`;
  }, b = await Z(g);
  if (b instanceof Error)
    return { status: "unauthenticated", error: b };
  const w = $();
  return {
    ...await W({
      getToken: g,
      transportFactory: w
    }),
    status: "authenticated",
    userName: b,
    logout: () => {
      I(u), window.location.reload();
    },
    exportTokens: () => {
      const S = localStorage.getItem(u.accessToken), y = localStorage.getItem(u.refreshToken), h = localStorage.getItem(u.expiresAt);
      if (!S || !y || !h)
        throw new Error("Cannot export tokens: not authenticated");
      return {
        accessToken: S,
        refreshToken: y,
        expiresAt: parseInt(h, 10)
      };
    }
  };
}, je = (e, a) => new Promise((s) => {
  const r = () => {
    window.removeEventListener("message", o), window.clearInterval(i);
  }, o = (d) => {
    if (d.origin !== Q) return;
    const l = d.data;
    if (!(l == null || typeof l != "object") && l.type === "audiotool-oauth") {
      if (r(), typeof l.error == "string" && l.error.length > 0) {
        s({
          kind: "error",
          error: l.error,
          errorDescription: typeof l.error_description == "string" ? l.error_description : void 0
        });
        return;
      }
      if (typeof l.code != "string" || typeof l.state != "string") {
        s({
          kind: "error",
          error: "invalid_response",
          errorDescription: "Popup returned an unexpected message shape."
        });
        return;
      }
      if (l.state !== a) {
        s({
          kind: "error",
          error: "invalid_state",
          errorDescription: "State mismatch — possible CSRF or mismatched popup session."
        });
        return;
      }
      s({ kind: "ok", code: l.code, state: l.state });
    }
  }, i = window.setInterval(() => {
    e.closed && (r(), s({ kind: "cancelled" }));
  }, 500);
  window.addEventListener("message", o);
}), Ge = "https://oauth.audiotool.com/oauth2/token", Ze = (e) => {
  let { accessToken: a, refreshToken: s, expiresAt: r } = e;
  const { clientId: o, onTokenRefresh: i } = e;
  let d;
  const l = async () => Date.now() >= r - 6e4 ? d || (d = (async () => {
    try {
      const c = await v(Ge, {
        method: "POST",
        headers: {
          "Content-Type": "application/x-www-form-urlencoded"
        },
        body: new URLSearchParams({
          client_id: o,
          grant_type: "refresh_token",
          refresh_token: s
        })
      });
      if (c instanceof Error)
        throw new Error(`Token refresh failed: ${c.message}`, {
          cause: c
        });
      if (!c.ok)
        throw new Error(`Token refresh failed: ${c.statusText}`);
      const n = await c.json();
      if (n.error)
        throw new Error(`Token refresh failed: ${n.error}`, {
          cause: n.error_description
        });
      return a = n.access_token, s = n.refresh_token, r = Date.now() + n.expires_in * 1e3, i == null || i({
        accessToken: a,
        refreshToken: s,
        expiresAt: r
      }), a;
    } finally {
      d = void 0;
    }
  })(), d) : a;
  return {
    getToken: async () => {
      const t = await l();
      return t.startsWith("Bearer ") ? t : `Bearer ${t}`;
    }
  };
};
export {
  Je as audiotool,
  Qe as audiotoolPopup,
  qe as createAudiotoolClient,
  ra as createOfflineDocument,
  Ae as createPATAuth,
  Ze as createServerAuth
};
