// @audiotool/nexus/dist/node.js
import { K as KEEPALIVE_HEADER, c as createOfflineDocument$1, i as initWasmLoader } from "./synced-document-ywEybIAl.js";
import { t as throw_ } from "./lang-K-8hAzE4.js";
const wasmUrl = "/document_validator.wasm", wasmJsUrl = "/wasm_exec.js", createDiskWasmLoader = () => ({
  async executeRuntime() {
    const path = await import("node:path"), url = await import("node:url"), fs = await import("node:fs");
    let filePath;
    filePath = path.resolve(
      path.dirname(url.fileURLToPath(import.meta.url)),
      `.${wasmJsUrl}`
    );
    const { promise, resolve } = Promise.withResolvers();
    return fs.readFile(
      filePath,
      "utf-8",
      (err, data) => err != null ? throw_(err) : (eval(data), resolve())
    ), await promise;
  },
  async loadModule() {
    const e = await import("node:fs"), r = await import("node:path"), a = await import("node:url");
    let t;
    return t = r.resolve(
      r.dirname(a.fileURLToPath(import.meta.url)),
      `.${wasmUrl}`
    ), WebAssembly.compile(new Uint8Array(e.readFileSync(t)));
  }
}), createNodeTransport = () => ({
  createTransport: async (e) => {
    const { createConnectTransport: r } = await import("@connectrpc/connect-node");
    return r({
      baseUrl: e.baseUrl,
      httpVersion: "2",
      jsonOptions: {
        typeRegistry: e.typeRegistry
      },
      useBinaryFormat: e.useBinaryFormat,
      interceptors: [
        (a) => async (t) => (t.header.delete(KEEPALIVE_HEADER), t.header.set("Authorization", await e.getToken()), a(t))
      ]
    });
  }
}), createOfflineDocument = async (e) => (initWasmLoader(createDiskWasmLoader()), createOfflineDocument$1(e));
export {
  createDiskWasmLoader,
  createNodeTransport,
  createOfflineDocument
};
