// @audiotool/nexus/dist/operation_pb-CYWjPOnU.js
var q = Object.defineProperty;
var L = (J, a, t) => a in J ? q(J, a, { enumerable: !0, configurable: !0, writable: !0, value: t }) : J[a] = t;
var r = (J, a, t) => L(J, typeof a != "symbol" ? a + "" : a, t);
import { Message as s, proto3 as i, Any as k, Timestamp as z } from "@bufbuild/protobuf";
const e = class e extends s {
  constructor(t) {
    super();
    /**
     * The server-assigned name, which is only unique within the same service that
     * originally returns it. If you use the default HTTP mapping, the
     * `name` should be a resource name ending with `operations/{unique_id}`.
     *
     * @generated from field: string name = 1;
     */
    r(this, "name", "");
    /**
     * Service-specific metadata associated with the operation.  It typically
     * contains progress information and common metadata such as create time.
     * Some services might not provide such metadata.  Any method that returns a
     * long-running operation should document the metadata type, if any.
     *
     * @generated from field: google.protobuf.Any metadata = 2;
     */
    r(this, "metadata");
    /**
     * If the value is `false`, it means the operation is still in progress.
     * If `true`, the operation is completed, and either `error` or `response` is
     * available.
     *
     * @generated from field: bool done = 3;
     */
    r(this, "done", !1);
    /**
     * The operation result, which can be either an `error` or a valid `response`.
     * If `done` == `false`, neither `error` nor `response` is set.
     * If `done` == `true`, exactly one of `error` or `response` is set.
     *
     * @generated from oneof audiotool.longrunning.v1.Operation.result
     */
    r(this, "result", { case: void 0 });
    /**
     * List of owners for this Operation
     *
     * @generated from field: repeated string owners = 6;
     */
    r(this, "owners", []);
    /**
     * The time the operation was created.
     *
     * @generated from field: google.protobuf.Timestamp create_time = 7;
     */
    r(this, "createTime");
    i.util.initPartial(t, this);
  }
  static fromBinary(t, n) {
    return new e().fromBinary(t, n);
  }
  static fromJson(t, n) {
    return new e().fromJson(t, n);
  }
  static fromJsonString(t, n) {
    return new e().fromJsonString(t, n);
  }
  static equals(t, n) {
    return i.util.equals(e, t, n);
  }
};
r(e, "runtime", i), r(e, "typeName", "audiotool.longrunning.v1.Operation"), r(e, "fields", i.util.newFieldList(() => [
  {
    no: 1,
    name: "name",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  { no: 2, name: "metadata", kind: "message", T: k },
  {
    no: 3,
    name: "done",
    kind: "scalar",
    T: 8
    /* ScalarType.BOOL */
  },
  { no: 4, name: "error", kind: "message", T, oneof: "result" },
  { no: 5, name: "response", kind: "message", T: k, oneof: "result" },
  { no: 6, name: "owners", kind: "scalar", T: 9, repeated: !0 },
  { no: 7, name: "create_time", kind: "message", T: z }
]));
let y = e;
const o = class o extends s {
  constructor(t) {
    super();
    /**
     * The name of the operation resource.
     *
     * @generated from field: string name = 1;
     */
    r(this, "name", "");
    i.util.initPartial(t, this);
  }
  static fromBinary(t, n) {
    return new o().fromBinary(t, n);
  }
  static fromJson(t, n) {
    return new o().fromJson(t, n);
  }
  static fromJsonString(t, n) {
    return new o().fromJsonString(t, n);
  }
  static equals(t, n) {
    return i.util.equals(o, t, n);
  }
};
r(o, "runtime", i), r(o, "typeName", "audiotool.longrunning.v1.GetOperationRequest"), r(o, "fields", i.util.newFieldList(() => [
  {
    no: 1,
    name: "name",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  }
]));
let B = o;
const u = class u extends s {
  constructor(t) {
    super();
    /**
     * The operation resource.
     *
     * @generated from field: audiotool.longrunning.v1.Operation operation = 1;
     */
    r(this, "operation");
    i.util.initPartial(t, this);
  }
  static fromBinary(t, n) {
    return new u().fromBinary(t, n);
  }
  static fromJson(t, n) {
    return new u().fromJson(t, n);
  }
  static fromJsonString(t, n) {
    return new u().fromJsonString(t, n);
  }
  static equals(t, n) {
    return i.util.equals(u, t, n);
  }
};
r(u, "runtime", i), r(u, "typeName", "audiotool.longrunning.v1.GetOperationResponse"), r(u, "fields", i.util.newFieldList(() => [
  { no: 1, name: "operation", kind: "message", T: y }
]));
let S = u;
const m = class m extends s {
  constructor(t) {
    super();
    /**
     * The name of the operation's parent resource.
     *
     * @generated from field: string name = 4;
     */
    r(this, "name", "");
    /**
     * The standard list filter.
     *
     * @generated from field: string filter = 1;
     */
    r(this, "filter", "");
    /**
     * The standard list page size.
     *
     * @generated from field: int32 page_size = 2;
     */
    r(this, "pageSize", 0);
    /**
     * The standard list page token.
     *
     * @generated from field: string page_token = 3;
     */
    r(this, "pageToken", "");
    i.util.initPartial(t, this);
  }
  static fromBinary(t, n) {
    return new m().fromBinary(t, n);
  }
  static fromJson(t, n) {
    return new m().fromJson(t, n);
  }
  static fromJsonString(t, n) {
    return new m().fromJsonString(t, n);
  }
  static equals(t, n) {
    return i.util.equals(m, t, n);
  }
};
r(m, "runtime", i), r(m, "typeName", "audiotool.longrunning.v1.ListOperationsRequest"), r(m, "fields", i.util.newFieldList(() => [
  {
    no: 4,
    name: "name",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  {
    no: 1,
    name: "filter",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  {
    no: 2,
    name: "page_size",
    kind: "scalar",
    T: 5
    /* ScalarType.INT32 */
  },
  {
    no: 3,
    name: "page_token",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  }
]));
let x = m;
const l = class l extends s {
  constructor(t) {
    super();
    /**
     * A list of operations that matches the specified filter in the request.
     *
     * @generated from field: repeated audiotool.longrunning.v1.Operation operations = 1;
     */
    r(this, "operations", []);
    /**
     * The standard List next-page token.
     *
     * @generated from field: string next_page_token = 2;
     */
    r(this, "nextPageToken", "");
    i.util.initPartial(t, this);
  }
  static fromBinary(t, n) {
    return new l().fromBinary(t, n);
  }
  static fromJson(t, n) {
    return new l().fromJson(t, n);
  }
  static fromJsonString(t, n) {
    return new l().fromJsonString(t, n);
  }
  static equals(t, n) {
    return i.util.equals(l, t, n);
  }
};
r(l, "runtime", i), r(l, "typeName", "audiotool.longrunning.v1.ListOperationsResponse"), r(l, "fields", i.util.newFieldList(() => [
  { no: 1, name: "operations", kind: "message", T: y, repeated: !0 },
  {
    no: 2,
    name: "next_page_token",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  }
]));
let v = l;
const c = class c extends s {
  constructor(t) {
    super();
    /**
     * The name of the operation resource to be cancelled.
     *
     * @generated from field: string name = 1;
     */
    r(this, "name", "");
    i.util.initPartial(t, this);
  }
  static fromBinary(t, n) {
    return new c().fromBinary(t, n);
  }
  static fromJson(t, n) {
    return new c().fromJson(t, n);
  }
  static fromJsonString(t, n) {
    return new c().fromJsonString(t, n);
  }
  static equals(t, n) {
    return i.util.equals(c, t, n);
  }
};
r(c, "runtime", i), r(c, "typeName", "audiotool.longrunning.v1.CancelOperationRequest"), r(c, "fields", i.util.newFieldList(() => [
  {
    no: 1,
    name: "name",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  }
]));
let P = c;
const f = class f extends s {
  constructor(a) {
    super(), i.util.initPartial(a, this);
  }
  static fromBinary(a, t) {
    return new f().fromBinary(a, t);
  }
  static fromJson(a, t) {
    return new f().fromJson(a, t);
  }
  static fromJsonString(a, t) {
    return new f().fromJsonString(a, t);
  }
  static equals(a, t) {
    return i.util.equals(f, a, t);
  }
};
r(f, "runtime", i), r(f, "typeName", "audiotool.longrunning.v1.CancelOperationResponse"), r(f, "fields", i.util.newFieldList(() => []));
let h = f;
const d = class d extends s {
  constructor(t) {
    super();
    /**
     * The name of the operation resource to be deleted.
     *
     * @generated from field: string name = 1;
     */
    r(this, "name", "");
    i.util.initPartial(t, this);
  }
  static fromBinary(t, n) {
    return new d().fromBinary(t, n);
  }
  static fromJson(t, n) {
    return new d().fromJson(t, n);
  }
  static fromJsonString(t, n) {
    return new d().fromJsonString(t, n);
  }
  static equals(t, n) {
    return i.util.equals(d, t, n);
  }
};
r(d, "runtime", i), r(d, "typeName", "audiotool.longrunning.v1.DeleteOperationRequest"), r(d, "fields", i.util.newFieldList(() => [
  {
    no: 1,
    name: "name",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  }
]));
let F = d;
const g = class g extends s {
  constructor(a) {
    super(), i.util.initPartial(a, this);
  }
  static fromBinary(a, t) {
    return new g().fromBinary(a, t);
  }
  static fromJson(a, t) {
    return new g().fromJson(a, t);
  }
  static fromJsonString(a, t) {
    return new g().fromJsonString(a, t);
  }
  static equals(a, t) {
    return i.util.equals(g, a, t);
  }
};
r(g, "runtime", i), r(g, "typeName", "audiotool.longrunning.v1.DeleteOperationResponse"), r(g, "fields", i.util.newFieldList(() => []));
let N = g;
const w = class w extends s {
  constructor(t) {
    super();
    /**
     * The status code, which should be an enum value of
     * [google.rpc.Code][google.rpc.Code].
     *
     * @generated from field: int32 code = 1;
     */
    r(this, "code", 0);
    /**
     * A developer-facing error message, which should be in English. Any
     * user-facing error message should be localized and sent in the
     * [google.rpc.Status.details][google.rpc.Status.details] field, or localized
     * by the client.
     *
     * @generated from field: string message = 2;
     */
    r(this, "message", "");
    /**
     * A list of messages that carry the error details.  There is a common set of
     * message types for APIs to use.
     *
     * @generated from field: repeated google.protobuf.Any details = 3;
     */
    r(this, "details", []);
    i.util.initPartial(t, this);
  }
  static fromBinary(t, n) {
    return new w().fromBinary(t, n);
  }
  static fromJson(t, n) {
    return new w().fromJson(t, n);
  }
  static fromJsonString(t, n) {
    return new w().fromJsonString(t, n);
  }
  static equals(t, n) {
    return i.util.equals(w, t, n);
  }
};
r(w, "runtime", i), r(w, "typeName", "audiotool.longrunning.v1.Status"), r(w, "fields", i.util.newFieldList(() => [
  {
    no: 1,
    name: "code",
    kind: "scalar",
    T: 5
    /* ScalarType.INT32 */
  },
  {
    no: 2,
    name: "message",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  { no: 3, name: "details", kind: "message", T: k, repeated: !0 }
]));
let T = w;
export {
  P as C,
  F as D,
  B as G,
  x as L,
  y as O,
  T as S,
  S as a,
  v as b,
  h as c,
  N as d
};
