// @audiotool/nexus/dist/synced-document-ywEybIAl.js
var di = Object.defineProperty;
var In = (i) => {
  throw TypeError(i);
};
var mi = (i, t, e) => t in i ? di(i, t, { enumerable: !0, configurable: !0, writable: !0, value: e }) : i[t] = e;
var o = (i, t, e) => mi(i, typeof t != "symbol" ? t + "" : t, e), It = (i, t, e) => t.has(i) || In("Cannot " + e);
var l = (i, t, e) => (It(i, t, "read from private field"), e ? e.call(i) : t.get(i)), b = (i, t, e) => t.has(i) ? In("Cannot add the same private member more than once") : t instanceof WeakSet ? t.add(i) : t.set(i, e), D = (i, t, e, n) => (It(i, t, "write to private field"), n ? n.call(i, e) : t.set(i, e), e), x = (i, t, e) => (It(i, t, "access private method"), e);
import { Message as w, proto3 as s, protoInt64 as Je, Any as Qe, createRegistry as fi, MethodKind as F, ScalarType as P } from "@bufbuild/protobuf";
import { P as q, e as Gn, b as $n, s as pi, a as hi, E as Et, g as gn, _ as yi } from "./get-schema-location-details-BVy-6cYL.js";
import { t as g, a as T, s as vn, b as qt } from "./lang-K-8hAzE4.js";
import { V as Fe, N as jn, h as gi, H as vt, s as vi, A as wi } from "./ticks-B57EbInS.js";
import { createClient as bi, ConnectError as Wn, Code as Be } from "@connectrpc/connect";
import { v5 as Ti } from "uuid";
import wn from "toposort";
import { O as xi } from "./operation_pb-CYWjPOnU.js";
const j = class j extends w {
  constructor(e) {
    super();
    /**
     * The id of the client. This is used to identify the client
     * (audiotool.document.v1.Client) and is used to distinguish between multiple clients of the same
     * source (e.g. user has opened the document on multiple devices). This name can be reused if a
     * client reconnects.
     *
     * @generated from field: string client_id = 1;
     */
    o(this, "clientId", "");
    /**
     * The source name of the client. This can be a user name in the form of 'users/{id}' but is not
     * guaranteed to be a user.
     *
     * @generated from field: string source_name = 2;
     */
    o(this, "sourceName", "");
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new j().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new j().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new j().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(j, e, n);
  }
};
o(j, "runtime", s), o(j, "typeName", "audiotool.document.v1.Client"), o(j, "fields", s.util.newFieldList(() => [
  {
    no: 1,
    name: "client_id",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  {
    no: 2,
    name: "source_name",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  }
]));
let Jt = j;
const W = class W extends w {
  constructor(e) {
    super();
    /**
     * The id of the document to attach to.
     *
     * @generated from field: string project_name = 1;
     */
    o(this, "projectName", "");
    /**
     * The client_id is used to identify the client.
     *
     * This is used to filter out messages so the client doesn't receive his own PutMetadata requests.
     *
     * @generated from field: string client_id = 2;
     */
    o(this, "clientId", "");
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new W().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new W().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new W().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(W, e, n);
  }
};
o(W, "runtime", s), o(W, "typeName", "audiotool.document.v1.AttachMetadataRequest"), o(W, "fields", s.util.newFieldList(() => [
  {
    no: 1,
    name: "project_name",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  {
    no: 2,
    name: "client_id",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  }
]));
let Ct = W;
const V = class V extends w {
  constructor(e) {
    super();
    /**
     * Creator can be a user in the form of 'users/{id}' but is not guaranteed to be a user.
     *
     * @generated from field: string source_name = 1;
     */
    o(this, "sourceName", "");
    /**
     * The client_id is used to identify the client. A user can have multiple clients as he could
     * have multiple devices or multiple tabs open.
     *
     * @generated from field: string client_id = 2;
     */
    o(this, "clientId", "");
    /**
     * The coordinates of the client can be blank.
     *
     * @generated from field: audiotool.document.v1.ClientCoordinates client_coordinates = 3;
     */
    o(this, "clientCoordinates");
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new V().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new V().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new V().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(V, e, n);
  }
};
o(V, "runtime", s), o(V, "typeName", "audiotool.document.v1.AttachMetadataResponse"), o(V, "fields", s.util.newFieldList(() => [
  {
    no: 1,
    name: "source_name",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  {
    no: 2,
    name: "client_id",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  { no: 3, name: "client_coordinates", kind: "message", T: Tt }
]));
let Lt = V;
const K = class K extends w {
  constructor(e) {
    super();
    /**
     * The project_name the metadata is for
     *
     * @generated from field: string project_name = 1;
     */
    o(this, "projectName", "");
    /**
     * The client_id is used to identify the client.
     *
     * This field is required.
     *
     * @generated from field: string client_id = 2;
     */
    o(this, "clientId", "");
    /**
     * The coordinates of the client can be blank.
     *
     * @generated from field: audiotool.document.v1.ClientCoordinates client_coordinates = 3;
     */
    o(this, "clientCoordinates");
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new K().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new K().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new K().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(K, e, n);
  }
};
o(K, "runtime", s), o(K, "typeName", "audiotool.document.v1.PutMetadataRequest"), o(K, "fields", s.util.newFieldList(() => [
  {
    no: 1,
    name: "project_name",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  {
    no: 2,
    name: "client_id",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  { no: 3, name: "client_coordinates", kind: "message", T: Tt }
]));
let Bt = K;
const R = class R extends w {
  constructor(t) {
    super(), s.util.initPartial(t, this);
  }
  static fromBinary(t, e) {
    return new R().fromBinary(t, e);
  }
  static fromJson(t, e) {
    return new R().fromJson(t, e);
  }
  static fromJsonString(t, e) {
    return new R().fromJsonString(t, e);
  }
  static equals(t, e) {
    return s.util.equals(R, t, e);
  }
};
o(R, "runtime", s), o(R, "typeName", "audiotool.document.v1.PutMetadataResponse"), o(R, "fields", s.util.newFieldList(() => []));
let zt = R;
const X = class X extends w {
  constructor(e) {
    super();
    /**
     * The x coordinate of the cursor.
     *
     * @generated from field: double x = 1;
     */
    o(this, "x", 0);
    /**
     * The y coordinate of the cursor.
     *
     * @generated from field: double y = 2;
     */
    o(this, "y", 0);
    /**
     * The context is a client specific information like:
     * - timeline
     * - mixer
     * - desktop
     * - ...
     *
     * The context may also be a descriptive information containing an element path like:
     *
     * - timeline.{id}.sth
     *
     * If the cursor left the window it also can be set to an empty string.
     *
     * @generated from field: string context = 3;
     */
    o(this, "context", "");
    /**
     * The height of the context - optional.
     *
     * @generated from field: double context_height = 4;
     */
    o(this, "contextHeight", 0);
    /**
     * The width of the context - optional.
     *
     * @generated from field: double context_width = 5;
     */
    o(this, "contextWidth", 0);
    /**
     * The x display value of the context - optional.
     *
     * @generated from field: double context_x = 6;
     */
    o(this, "contextX", 0);
    /**
     * The y display value of the context - optional.
     *
     * @generated from field: double context_y = 7;
     */
    o(this, "contextY", 0);
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new X().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new X().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new X().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(X, e, n);
  }
};
o(X, "runtime", s), o(X, "typeName", "audiotool.document.v1.ClientCoordinates"), o(X, "fields", s.util.newFieldList(() => [
  {
    no: 1,
    name: "x",
    kind: "scalar",
    T: 1
    /* ScalarType.DOUBLE */
  },
  {
    no: 2,
    name: "y",
    kind: "scalar",
    T: 1
    /* ScalarType.DOUBLE */
  },
  {
    no: 3,
    name: "context",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  {
    no: 4,
    name: "context_height",
    kind: "scalar",
    T: 1
    /* ScalarType.DOUBLE */
  },
  {
    no: 5,
    name: "context_width",
    kind: "scalar",
    T: 1
    /* ScalarType.DOUBLE */
  },
  {
    no: 6,
    name: "context_x",
    kind: "scalar",
    T: 1
    /* ScalarType.DOUBLE */
  },
  {
    no: 7,
    name: "context_y",
    kind: "scalar",
    T: 1
    /* ScalarType.DOUBLE */
  }
]));
let Tt = X;
const Q = class Q extends w {
  constructor(e) {
    super();
    /**
     * The document id is used to identify the document.
     *
     * @generated from field: string project_name = 1;
     */
    o(this, "projectName", "");
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new Q().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new Q().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new Q().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(Q, e, n);
  }
};
o(Q, "runtime", s), o(Q, "typeName", "audiotool.document.v1.GetCommitIndexRangeRequest"), o(Q, "fields", s.util.newFieldList(() => [
  {
    no: 1,
    name: "project_name",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  }
]));
let Ot = Q;
const Y = class Y extends w {
  constructor(e) {
    super();
    /**
     * The minimum commit_index currently held by the server.
     *
     * @generated from field: uint64 min_commit_index = 1;
     */
    o(this, "minCommitIndex", Je.zero);
    /**
     * The maximum commit_index currently held by the server.
     *
     * @generated from field: uint64 max_commit_index = 2;
     */
    o(this, "maxCommitIndex", Je.zero);
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new Y().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new Y().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new Y().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(Y, e, n);
  }
};
o(Y, "runtime", s), o(Y, "typeName", "audiotool.document.v1.GetCommitIndexRangeResponse"), o(Y, "fields", s.util.newFieldList(() => [
  {
    no: 1,
    name: "min_commit_index",
    kind: "scalar",
    T: 4
    /* ScalarType.UINT64 */
  },
  {
    no: 2,
    name: "max_commit_index",
    kind: "scalar",
    T: 4
    /* ScalarType.UINT64 */
  }
]));
let Ht = Y;
const Z = class Z extends w {
  constructor(e) {
    super();
    /**
     * The document id is used to identify the document.
     *
     * @generated from field: string project_name = 1;
     */
    o(this, "projectName", "");
    /**
     * The filter is used to filter the entities.
     *
     * @generated from field: string filter = 2;
     */
    o(this, "filter", "");
    /**
     * The commit_index for the entities to get.
     *
     * If the commit_index is 0, you will receive the latest entities.
     *
     * If the commit_index is > 0, you will receive the entities at the commit_index or an error if
     * the commit_index is out of bound.
     *
     * @generated from field: uint64 commit_index = 3;
     */
    o(this, "commitIndex", Je.zero);
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new Z().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new Z().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new Z().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(Z, e, n);
  }
};
o(Z, "runtime", s), o(Z, "typeName", "audiotool.document.v1.GetEntitiesRequest"), o(Z, "fields", s.util.newFieldList(() => [
  {
    no: 1,
    name: "project_name",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  {
    no: 2,
    name: "filter",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  {
    no: 3,
    name: "commit_index",
    kind: "scalar",
    T: 4
    /* ScalarType.UINT64 */
  }
]));
let Ut = Z;
const _ = class _ extends w {
  constructor(e) {
    super();
    /**
     * A list of all entities in the document, topo-sorted so all pointers
     * are always valid.
     *
     * @generated from field: repeated google.protobuf.Any entities = 1;
     */
    o(this, "entities", []);
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new _().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new _().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new _().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(_, e, n);
  }
};
o(_, "runtime", s), o(_, "typeName", "audiotool.document.v1.GetEntitiesResponse"), o(_, "fields", s.util.newFieldList(() => [
  { no: 1, name: "entities", kind: "message", T: Qe, repeated: !0 }
]));
let Gt = _;
const ee = class ee extends w {
  constructor(e) {
    super();
    /**
     * The id of the document to attach to.
     *
     * @generated from field: string project_name = 1;
     */
    o(this, "projectName", "");
    /**
     * The commit_index which the client has last seen.
     *
     * If the commit_index is 0, the client has not seen any transactions yet and the client will
     * receive a NoopResponse message.
     *
     * If the commit_index is > 0, the client has seen transactions and the client will receive all
     * transactions (Modification, Create, Delete) starting from the commit_index. The DocumentService
     * will only hold a specific amount of transactions in memory, so if the commit_index is too far
     * in the past, the client will receive an error (grpc.CodeOutOfRange). A Client should then call
     * Attach with commit_index 0.
     *
     * If the client is up to date with the server the server will send a NoopResponse message.
     *
     * @generated from field: uint64 commit_index = 2;
     */
    o(this, "commitIndex", Je.zero);
    /**
     * The id of the client. This is used to identify the client
     * (audiotool.document.v1.Client) and is used to distinguish between multiple clients of the same
     * source (e.g. user has opened the document on multiple devices). This name can be reused if a
     * client reconnects.
     *
     * @generated from field: string client_id = 3;
     */
    o(this, "clientId", "");
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new ee().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new ee().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new ee().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(ee, e, n);
  }
};
o(ee, "runtime", s), o(ee, "typeName", "audiotool.document.v1.AttachRequest"), o(ee, "fields", s.util.newFieldList(() => [
  {
    no: 1,
    name: "project_name",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  {
    no: 2,
    name: "commit_index",
    kind: "scalar",
    T: 4
    /* ScalarType.UINT64 */
  },
  {
    no: 3,
    name: "client_id",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  }
]));
let $t = ee;
const te = class te extends w {
  constructor(e) {
    super();
    /**
     * Possible responses; currently only transaction, but this might change.
     *
     * @generated from oneof audiotool.document.v1.AttachResponse.message
     */
    o(this, "message", { case: void 0 });
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new te().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new te().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new te().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(te, e, n);
  }
};
o(te, "runtime", s), o(te, "typeName", "audiotool.document.v1.AttachResponse"), o(te, "fields", s.util.newFieldList(() => [
  { no: 1, name: "transaction", kind: "message", T: B, oneof: "message" },
  { no: 2, name: "noop", kind: "message", T: Wt, oneof: "message" }
]));
let jt = te;
const ne = class ne extends w {
  constructor(e) {
    super();
    /**
     * The current commit_index of the server.
     *
     * @generated from field: uint64 server_commit_index = 1;
     */
    o(this, "serverCommitIndex", Je.zero);
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new ne().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new ne().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new ne().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(ne, e, n);
  }
};
o(ne, "runtime", s), o(ne, "typeName", "audiotool.document.v1.Noop"), o(ne, "fields", s.util.newFieldList(() => [
  {
    no: 1,
    name: "server_commit_index",
    kind: "scalar",
    T: 4
    /* ScalarType.UINT64 */
  }
]));
let Wt = ne;
const ie = class ie extends w {
  constructor(e) {
    super();
    /**
     * The document id is used to identify the document.
     *
     * @generated from field: string project_name = 1;
     */
    o(this, "projectName", "");
    /**
     * The transactions to apply to the document, in order.
     *
     * @generated from field: repeated audiotool.document.v1.Transaction transactions = 2;
     */
    o(this, "transactions", []);
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new ie().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new ie().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new ie().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(ie, e, n);
  }
};
o(ie, "runtime", s), o(ie, "typeName", "audiotool.document.v1.ApplyTransactionsRequest"), o(ie, "fields", s.util.newFieldList(() => [
  {
    no: 1,
    name: "project_name",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  { no: 2, name: "transactions", kind: "message", T: B, repeated: !0 }
]));
let Vt = ie;
const ae = class ae extends w {
  constructor(e) {
    super();
    /**
     * Maps transaction id to errors, if there are any.
     *
     * @generated from field: map<string, string> errors = 1;
     */
    o(this, "errors", {});
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new ae().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new ae().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new ae().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(ae, e, n);
  }
};
o(ae, "runtime", s), o(ae, "typeName", "audiotool.document.v1.ApplyTransactionsResponse"), o(ae, "fields", s.util.newFieldList(() => [
  { no: 1, name: "errors", kind: "map", K: 9, V: {
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  } }
]));
let Kt = ae;
const re = class re extends w {
  constructor(e) {
    super();
    /**
     * The id of the transaction. When calling `Modify`, this id must be a valid uuid.
     * Can be empty in the initial AttachResponse message.
     *
     * @generated from field: string id = 1;
     */
    o(this, "id", "");
    /**
     * An incremental number that is increased with each accepted transaction provided by the
     * document-service.
     *
     * @generated from field: uint64 commit_index = 2;
     */
    o(this, "commitIndex", Je.zero);
    /**
     * A list of modifications to the document. The order of the modifications is important.
     *
     * @generated from field: repeated audiotool.document.v1.Modification modifications = 3;
     */
    o(this, "modifications", []);
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new re().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new re().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new re().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(re, e, n);
  }
};
o(re, "runtime", s), o(re, "typeName", "audiotool.document.v1.Transaction"), o(re, "fields", s.util.newFieldList(() => [
  {
    no: 1,
    name: "id",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  {
    no: 2,
    name: "commit_index",
    kind: "scalar",
    T: 4
    /* ScalarType.UINT64 */
  },
  { no: 3, name: "modifications", kind: "message", T: He, repeated: !0 }
]));
let B = re;
const oe = class oe extends w {
  constructor(e) {
    super();
    /**
     * CUD message
     *
     * @generated from oneof audiotool.document.v1.Modification.modification
     */
    o(this, "modification", { case: void 0 });
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new oe().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new oe().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new oe().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(oe, e, n);
  }
};
o(oe, "runtime", s), o(oe, "typeName", "audiotool.document.v1.Modification"), o(oe, "fields", s.util.newFieldList(() => [
  { no: 2, name: "create", kind: "message", T: Rt, oneof: "modification" },
  { no: 3, name: "delete", kind: "message", T: Qt, oneof: "modification" },
  { no: 4, name: "update", kind: "message", T: Xt, oneof: "modification" }
]));
let He = oe;
const se = class se extends w {
  constructor(e) {
    super();
    /**
     * The entity to create
     *
     * @generated from field: google.protobuf.Any entity = 1;
     */
    o(this, "entity");
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new se().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new se().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new se().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(se, e, n);
  }
};
o(se, "runtime", s), o(se, "typeName", "audiotool.document.v1.Create"), o(se, "fields", s.util.newFieldList(() => [
  { no: 1, name: "entity", kind: "message", T: Qe }
]));
let Rt = se;
const ce = class ce extends w {
  constructor(e) {
    super();
    /**
     * Pointer is used to reference the entity + field which shall be updated.
     *
     * @generated from field: audiotool.document.v1.Pointer field = 1;
     */
    o(this, "field");
    /**
     * The poossible values for an update
     *
     * @generated from oneof audiotool.document.v1.Update.value
     */
    o(this, "value", { case: void 0 });
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new ce().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new ce().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new ce().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(ce, e, n);
  }
};
o(ce, "runtime", s), o(ce, "typeName", "audiotool.document.v1.Update"), o(ce, "fields", s.util.newFieldList(() => [
  { no: 1, name: "field", kind: "message", T: q },
  { no: 2, name: "double", kind: "scalar", T: 1, oneof: "value" },
  { no: 3, name: "float", kind: "scalar", T: 2, oneof: "value" },
  { no: 4, name: "int32", kind: "scalar", T: 5, oneof: "value" },
  { no: 5, name: "int64", kind: "scalar", T: 3, oneof: "value" },
  { no: 6, name: "uint32", kind: "scalar", T: 13, oneof: "value" },
  { no: 7, name: "uint64", kind: "scalar", T: 4, oneof: "value" },
  { no: 8, name: "sint32", kind: "scalar", T: 17, oneof: "value" },
  { no: 9, name: "sint64", kind: "scalar", T: 18, oneof: "value" },
  { no: 10, name: "fixed32", kind: "scalar", T: 7, oneof: "value" },
  { no: 11, name: "fixed64", kind: "scalar", T: 6, oneof: "value" },
  { no: 12, name: "sfixed32", kind: "scalar", T: 15, oneof: "value" },
  { no: 13, name: "sfixed64", kind: "scalar", T: 16, oneof: "value" },
  { no: 14, name: "bool", kind: "scalar", T: 8, oneof: "value" },
  { no: 15, name: "string", kind: "scalar", T: 9, oneof: "value" },
  { no: 16, name: "bytes", kind: "scalar", T: 12, oneof: "value" },
  { no: 17, name: "pointer", kind: "message", T: q, oneof: "value" }
]));
let Xt = ce;
const le = class le extends w {
  constructor(e) {
    super();
    /**
     * entity to remove
     *
     * @generated from field: string entity_id = 1;
     */
    o(this, "entityId", "");
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new le().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new le().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new le().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(le, e, n);
  }
};
o(le, "runtime", s), o(le, "typeName", "audiotool.document.v1.Delete"), o(le, "fields", s.util.newFieldList(() => [
  {
    no: 1,
    name: "entity_id",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  }
]));
let Qt = le;
const ue = class ue extends w {
  constructor(t) {
    super(), s.util.initPartial(t, this);
  }
  static fromBinary(t, e) {
    return new ue().fromBinary(t, e);
  }
  static fromJson(t, e) {
    return new ue().fromJson(t, e);
  }
  static fromJsonString(t, e) {
    return new ue().fromJsonString(t, e);
  }
  static equals(t, e) {
    return s.util.equals(ue, t, e);
  }
};
o(ue, "runtime", s), o(ue, "typeName", "audiotool.document.v1.GetWasmRequest"), o(ue, "fields", s.util.newFieldList(() => []));
let Yt = ue;
const de = class de extends w {
  constructor(e) {
    super();
    /**
     * data contains the wasm binary.
     *
     * @generated from field: bytes data = 1;
     */
    o(this, "data", new Uint8Array(0));
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new de().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new de().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new de().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(de, e, n);
  }
};
o(de, "runtime", s), o(de, "typeName", "audiotool.document.v1.GetWasmResponse"), o(de, "fields", s.util.newFieldList(() => [
  {
    no: 1,
    name: "data",
    kind: "scalar",
    T: 12
    /* ScalarType.BYTES */
  }
]));
let Zt = de;
const me = class me extends w {
  constructor(t) {
    super(), s.util.initPartial(t, this);
  }
  static fromBinary(t, e) {
    return new me().fromBinary(t, e);
  }
  static fromJson(t, e) {
    return new me().fromJson(t, e);
  }
  static fromJsonString(t, e) {
    return new me().fromJsonString(t, e);
  }
  static equals(t, e) {
    return s.util.equals(me, t, e);
  }
};
o(me, "runtime", s), o(me, "typeName", "audiotool.document.v1.GetWasmExecRequest"), o(me, "fields", s.util.newFieldList(() => []));
let _t = me;
const fe = class fe extends w {
  constructor(e) {
    super();
    /**
     * the js/go runtime script needed to execute the wasm binary on the client.
     *
     * @generated from field: string data = 1;
     */
    o(this, "data", "");
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new fe().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new fe().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new fe().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(fe, e, n);
  }
};
o(fe, "runtime", s), o(fe, "typeName", "audiotool.document.v1.GetWasmExecResponse"), o(fe, "fields", s.util.newFieldList(() => [
  {
    no: 1,
    name: "data",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  }
]));
let en = fe;
const pe = class pe extends w {
  constructor(e) {
    super();
    /**
     * The document id is used to identify the document.
     *
     * @generated from field: string project_name = 1;
     */
    o(this, "projectName", "");
    /**
     * The client id is used to identify the client.
     *
     * @generated from field: string client_id = 2;
     */
    o(this, "clientId", "");
    /**
     * The last ping duration in ms measured by the client.
     * if left empty, the server will not count the ping.
     *
     * @generated from field: uint32 last_ping_ms = 3;
     */
    o(this, "lastPingMs", 0);
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new pe().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new pe().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new pe().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(pe, e, n);
  }
};
o(pe, "runtime", s), o(pe, "typeName", "audiotool.document.v1.PingRequest"), o(pe, "fields", s.util.newFieldList(() => [
  {
    no: 1,
    name: "project_name",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  {
    no: 2,
    name: "client_id",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  {
    no: 3,
    name: "last_ping_ms",
    kind: "scalar",
    T: 13
    /* ScalarType.UINT32 */
  }
]));
let tn = pe;
const he = class he extends w {
  constructor(t) {
    super(), s.util.initPartial(t, this);
  }
  static fromBinary(t, e) {
    return new he().fromBinary(t, e);
  }
  static fromJson(t, e) {
    return new he().fromJson(t, e);
  }
  static fromJsonString(t, e) {
    return new he().fromJsonString(t, e);
  }
  static equals(t, e) {
    return s.util.equals(he, t, e);
  }
};
o(he, "runtime", s), o(he, "typeName", "audiotool.document.v1.PingResponse"), o(he, "fields", s.util.newFieldList(() => []));
let nn = he;
const ye = class ye extends w {
  constructor(e) {
    super();
    /**
     * The document id is used to identify the document.
     *
     * @generated from field: string project_name = 1;
     */
    o(this, "projectName", "");
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new ye().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new ye().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new ye().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(ye, e, n);
  }
};
o(ye, "runtime", s), o(ye, "typeName", "audiotool.document.v1.GetClientStatsRequest"), o(ye, "fields", s.util.newFieldList(() => [
  {
    no: 1,
    name: "project_name",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  }
]));
let an = ye;
const ge = class ge extends w {
  constructor(e) {
    super();
    /**
     * A repeated list of information about clients paricipating in the document session.
     *
     * @generated from field: repeated audiotool.document.v1.ClientInfo client_info = 1;
     */
    o(this, "clientInfo", []);
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new ge().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new ge().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new ge().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(ge, e, n);
  }
};
o(ge, "runtime", s), o(ge, "typeName", "audiotool.document.v1.GetClientStatsResponse"), o(ge, "fields", s.util.newFieldList(() => [
  { no: 1, name: "client_info", kind: "message", T: on, repeated: !0 }
]));
let rn = ge;
const ve = class ve extends w {
  constructor(e) {
    super();
    /**
     * the client's name. It is idempotent created from the Client message to identify the client.
     *
     * @generated from field: string name = 1;
     */
    o(this, "name", "");
    /**
     * The client.
     *
     * @generated from field: audiotool.document.v1.Client client = 2;
     */
    o(this, "client");
    /**
     * the client's ping
     *
     * @generated from field: uint32 ping_ms = 3;
     */
    o(this, "pingMs", 0);
    /**
     * set if client is offline
     *
     * @generated from field: bool offline = 4;
     */
    o(this, "offline", !1);
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new ve().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new ve().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new ve().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(ve, e, n);
  }
};
o(ve, "runtime", s), o(ve, "typeName", "audiotool.document.v1.ClientInfo"), o(ve, "fields", s.util.newFieldList(() => [
  {
    no: 1,
    name: "name",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  { no: 2, name: "client", kind: "message", T: Jt },
  {
    no: 3,
    name: "ping_ms",
    kind: "scalar",
    T: 13
    /* ScalarType.UINT32 */
  },
  {
    no: 4,
    name: "offline",
    kind: "scalar",
    T: 8
    /* ScalarType.BOOL */
  }
]));
let on = ve;
const we = class we extends w {
  constructor(t) {
    super(), s.util.initPartial(t, this);
  }
  static fromBinary(t, e) {
    return new we().fromBinary(t, e);
  }
  static fromJson(t, e) {
    return new we().fromJson(t, e);
  }
  static fromJsonString(t, e) {
    return new we().fromJsonString(t, e);
  }
  static equals(t, e) {
    return s.util.equals(we, t, e);
  }
};
o(we, "runtime", s), o(we, "typeName", "audiotool.document.v1.GetVersionRequest"), o(we, "fields", s.util.newFieldList(() => []));
let sn = we;
const be = class be extends w {
  constructor(e) {
    super();
    /**
     * The version of the document service.
     *
     * @generated from field: string version = 1;
     */
    o(this, "version", "");
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new be().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new be().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new be().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(be, e, n);
  }
};
o(be, "runtime", s), o(be, "typeName", "audiotool.document.v1.GetVersionResponse"), o(be, "fields", s.util.newFieldList(() => [
  {
    no: 1,
    name: "version",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  }
]));
let cn = be;
const Te = class Te extends w {
  constructor(t) {
    super(), s.util.initPartial(t, this);
  }
  static fromBinary(t, e) {
    return new Te().fromBinary(t, e);
  }
  static fromJson(t, e) {
    return new Te().fromJson(t, e);
  }
  static fromJsonString(t, e) {
    return new Te().fromJsonString(t, e);
  }
  static equals(t, e) {
    return s.util.equals(Te, t, e);
  }
};
o(Te, "runtime", s), o(Te, "typeName", "audiotool.document.v1.GetTimeRequest"), o(Te, "fields", s.util.newFieldList(() => []));
let ln = Te;
const xe = class xe extends w {
  constructor(e) {
    super();
    /**
     * The current time in seconds since the Unix epoch.
     *
     * @generated from field: int64 time = 1;
     */
    o(this, "time", Je.zero);
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new xe().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new xe().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new xe().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(xe, e, n);
  }
};
o(xe, "runtime", s), o(xe, "typeName", "audiotool.document.v1.GetTimeResponse"), o(xe, "fields", s.util.newFieldList(() => [
  {
    no: 1,
    name: "time",
    kind: "scalar",
    T: 3
    /* ScalarType.INT64 */
  }
]));
let un = xe;
const ke = class ke extends w {
  constructor(e) {
    super();
    /**
     * The project name of the project to render audio for.
     *
     * @generated from field: string project_name = 1;
     */
    o(this, "projectName", "");
    /**
     * The commit index of the project state to render.
     *
     * If 0, the latest state of the project will be used.
     *
     * @generated from field: uint64 commit_index = 2;
     */
    o(this, "commitIndex", Je.zero);
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new ke().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new ke().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new ke().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(ke, e, n);
  }
};
o(ke, "runtime", s), o(ke, "typeName", "audiotool.document.v1.RenderAudioRequest"), o(ke, "fields", s.util.newFieldList(() => [
  {
    no: 1,
    name: "project_name",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  {
    no: 2,
    name: "commit_index",
    kind: "scalar",
    T: 4
    /* ScalarType.UINT64 */
  }
]));
let dn = ke, wt, Vn;
const ki = (i) => {
  wt === void 0 && (Vn = i);
}, Si = async () => (await Kn()).createWasmDocumentState(), Di = async () => (await Kn()).createWasmConsolidator(), Kn = async () => {
  if (wt !== void 0)
    return await wt;
  const i = Vn ?? await Mi(), { promise: t, resolve: e } = Promise.withResolvers();
  wt = t;
  const n = i.executeRuntime(), a = i.loadModule();
  await n;
  const r = globalThis.Go ?? g("wasm wrapper initialization failed: Go not defined"), c = new r();
  try {
    await a;
    const m = await WebAssembly.instantiate(
      await a,
      c.importObject
    );
    c.run(m);
  } catch (m) {
    throw new Error("Failed to instantiate validator WASM", { cause: m });
  }
  delete globalThis.Go;
  const d = (
    // createDocumentState is attached to globalThis by the executed wasm code
    globalThis.createDocumentState ?? g("wasm initialization failed: createDocumentState not defined")
  ), u = (
    // createConsolidator is attached to globalThis by the executed wasm code
    globalThis.createConsolidator ?? g("wasm initialization failed: createConsolidator not defined")
  );
  return delete globalThis.createDocumentState, delete globalThis.createConsolidator, e({
    createWasmDocumentState: () => {
      const m = d();
      let f = !1;
      return {
        applyTransaction(y) {
          T(
            !f,
            "tried applying a transaction after document state was terminated"
          );
          const p = m.applyTransaction(y.toBinary());
          return p instanceof Object && "error" in p ? p.error : new B({
            modifications: p.rollbacks.map((k) => He.fromBinary(k))
          });
        },
        terminate() {
          f || (f = !0, m.delete());
        }
      };
    },
    createWasmConsolidator: () => {
      const m = u();
      let f = !1;
      return {
        consolidate(y, p, k) {
          T(
            !f,
            "tried consolidating after consolidator was terminated"
          );
          const S = m.consolidate(
            y.map((N) => N.toBinary()),
            p,
            k.map((N) => N.toBinary())
          );
          if (Array.isArray(S))
            return S.map((N) => B.fromBinary(N));
          throw new Error("wasm consolidator returned an error", {
            cause: (S == null ? void 0 : S.error) ?? "no error string provided by wasm"
          });
        },
        terminate() {
          f || (f = !0, m.delete());
        }
      };
    }
  }), await t;
}, Mi = async () => {
  if (typeof process < "u")
    throw new Error(
      "Can't load Wasm: server-side usage requires createDiskWasmLoader() from @audiotool/nexus/node"
    );
  const { createBrowserWasmLoader: i } = await import("./browser-wasm-loader-DqBeBlN9.js");
  return i();
}, Ii = (i) => {
  const t = i == null ? void 0 : i.typeUrl;
  if (t === void 0)
    return;
  const e = t.split("/").pop();
  return Gn[e];
}, Ni = (i) => Gn[i.getType().typeName] ?? g(), Ye = (i) => Qe.pack(i), Pi = fi(...Object.values($n)), at = (i) => {
  if (i !== void 0)
    return i.unpack(Pi);
}, Fi = (i) => at(i) ?? g("couldn't unpack any entity of type " + (i == null ? void 0 : i.typeUrl)), Ai = (i) => i.charAt(0).toLowerCase() + i.slice(1), Se = class Se extends w {
  constructor(e) {
    super();
    /**
     * The target of the preset. An existing entity of the same type is updated to match this entity.
     *
     * @generated from field: google.protobuf.Any target = 2;
     */
    o(this, "target");
    /**
     * Other related entities of the preset that are created fresh when applying the preset
     *
     * @generated from field: repeated google.protobuf.Any relatives = 3;
     */
    o(this, "relatives", []);
    s.util.initPartial(e, this);
  }
  static fromBinary(e, n) {
    return new Se().fromBinary(e, n);
  }
  static fromJson(e, n) {
    return new Se().fromJson(e, n);
  }
  static fromJsonString(e, n) {
    return new Se().fromJsonString(e, n);
  }
  static equals(e, n) {
    return s.util.equals(Se, e, n);
  }
};
o(Se, "runtime", s), o(Se, "typeName", "audiotool.document.v1.preset.v1.Preset"), o(Se, "fields", s.util.newFieldList(() => [
  { no: 2, name: "target", kind: "message", T: Qe },
  { no: 3, name: "relatives", kind: "message", T: Qe, repeated: !0 }
]));
let xt = Se;
const Ei = "###keepalive###", qi = (i) => {
  if (i === void 0)
    return;
  const t = (i == null ? void 0 : i.keepalive) ?? !1, e = new Headers(i == null ? void 0 : i.headers);
  e.set(Ei, t ? "true" : "false");
  const n = {
    ...i,
    headers: e
  };
  return delete n.keepalive, n;
}, Ji = (i, t, e) => {
  const r = ((u) => u instanceof Wn ? u.code === Be.Unavailable || u.code === Be.Unknown : !1), c = bi(i, t), d = {};
  return Object.entries(i.methods).forEach(([u, m]) => {
    if (m.kind !== F.Unary) {
      d[u] = c[u];
      return;
    }
    const f = c[u];
    d[u] = async (y, p) => {
      var k, S, N, Mn;
      for (p = qi(p); ; )
        try {
          const Ae = await f(
            y,
            p
          );
          return (k = p == null ? void 0 : p.callIsOk) == null || k.setValue(!0), Ae;
        } catch (Ae) {
          if ((S = p == null ? void 0 : p.callIsOk) == null || S.setValue(!1), !(Ae instanceof Error && r(Ae)))
            return new Error(`${i.typeName}.${u} threw error`, {
              cause: Ae
            });
          if ((N = p == null ? void 0 : p.onRetry) == null || N.call(p, Ae), ((p == null ? void 0 : p.logIfRetrying) ?? !0) && console.warn(
            `${i.typeName}.${u} call failed, retrying in 2000 ms. Error:`,
            Ae
          ), await vn(
            2e3 + Math.random() * 2e3 * 0.1,
            // pass in signal for earlier return in case of abort
            p == null ? void 0 : p.signal
          ), ((Mn = p == null ? void 0 : p.signal) == null ? void 0 : Mn.aborted) ?? !1)
            return new Error(`${i.typeName}.${u} was aborted`, {
              cause: Ae
            });
          continue;
        }
    };
  }), d;
}, bn = (i, t) => Ci[i](t), Ci = {
  // non-numbers:
  // already boolean
  bool: (i) => i,
  // already string
  string: (i) => i,
  // already pointer (NexusLocation)
  pointer: (i) => i,
  // unused & nothing to truncate
  bytes: (i) => i,
  // floats
  // single-precision float
  float: (i) => Math.fround(i),
  // double precision float
  double: (i) => i,
  // 32 bit integers
  // 32 bit signed integer
  int32: (i) => Pt(i),
  // 32 bit unsigned integer
  uint32: (i) => qn(i),
  // alternative signed 32 bit integer
  sint32: (i) => Pt(i),
  // alternative unsigned 32 bit integer
  fixed32: (i) => qn(i),
  // 64 bit signed integer, bigint
  // alternative signed 32 bit integer
  sfixed32: (i) => Pt(i),
  // 64 bit integers
  // 64 bit signed integer, bigint
  int64: (i) => Nt(i),
  // 64 bit unsigned integer, bigint
  uint64: (i) => En(i),
  // alternative signed 64 bit integer
  sint64: (i) => Nt(i),
  // alternative unsigned 64 bit integer
  fixed64: (i) => En(i),
  // alternative signed 64 bit integer
  sfixed64: (i) => Nt(i)
}, Li = 2147483647, Bi = -2147483648, zi = 4294967295, Nn = 18446744073709551615n, Pn = 0n, Fn = -9223372036854775808n, An = 9223372036854775807n, En = (i) => i < Pn ? Pn : i > Nn ? Nn : i, Nt = (i) => i > An ? An : i < Fn ? Fn : i, qn = (i) => Math.max(0, Math.min(zi, Math.trunc(i))), Pt = (i) => Math.max(Bi, Math.min(Li, Math.trunc(i)));
class Ze {
  /** @internal */
  constructor(t, e) {
    o(this, "location");
    /** The array of fields contained in this array field. */
    o(this, "array");
    this.location = t, this.array = e;
  }
}
var $e;
class G {
  /** @internal */
  constructor(t, e, n, a) {
    /** The location of this field within the Nexus document. */
    o(this, "location");
    /** The value of the field; private with a getter to prevent accidenal overwrites. */
    b(this, $e);
    /** Whether the field is mutable. Updating immutable fields results in a transaction error.
     *
     * If the type fo the field is known at compile time, updating the field using the {@link TransactionBuilder}
     * or listing to updates of the field using the {@link NexusEventManager} will result in type errors.
     */
    o(this, "mutable");
    /** @internal The type of the field in the protobuf message. Used internally.  */
    o(this, "_protoType");
    this.location = t, D(this, $e, e), this._protoType = n, this.mutable = a;
  }
  /** Get the value of the field. To set, use transactions in the document. */
  get value() {
    return l(this, $e);
  }
  /**
   * @internal
   *
   * Set the value of this field. Used by {@link NexusDocument}, don't use directly! */
  _setValue(t) {
    D(this, $e, bn(this._protoType, t));
  }
}
$e = new WeakMap();
const Rn = (i, t) => i === t ? !0 : i instanceof Uint8Array && t instanceof Uint8Array ? i.length !== t.length ? !1 : i.every((e, n) => e === t[n]) : !1;
class $ {
  /** @internal */
  constructor(t, e) {
    /** the fields in this object */
    o(this, "fields");
    o(this, "location");
    this.location = e, this.fields = t;
  }
  /** @internal Returns the field with the given field number. Throws if it can't find it. */
  _getField(t) {
    return Object.values(this.fields).find(
      (e) => e.location.fieldIndex.at(-1) === t
    ) ?? g(
      `can't find field with number ${t} on NexusObject ${this}`
    );
  }
}
class Ge extends $ {
  /** @internal */
  constructor(e, n) {
    super(n, e);
    /** The id of this entity. */
    o(this, "id");
    /** The entity type key of this entity. */
    o(this, "entityType");
    this.id = e.entityId, this.entityType = e.entityType;
  }
  /**
   * @internal
   *
   * Returns a field with a specific field index. Throws if it can't find it.
   */
  _resolveField(e) {
    if (e.length === 0)
      throw "tried resolving empty field path";
    const [n, ...a] = e;
    let r = this._getField(n);
    for (const c of a) {
      if (r instanceof $) {
        r = r._getField(c);
        continue;
      }
      if (r instanceof Ze) {
        r = r.array[c];
        continue;
      }
      throw `tried resolving field number ${c} on non-object field ${r}`;
    }
    return r;
  }
}
const Oi = (...i) => {
  const t = new Fe(!0), e = () => t.setValue(i.every((n) => n.getValue()));
  return i.forEach((n) => n.subscribe(() => e())), e(), t;
}, Hi = (i, t, e) => {
  const n = new jn(), a = new Fe(0), r = new Fe(!0);
  n.subscribe((d) => {
    r.setValue(d);
  });
  const c = qt(
    async (d) => {
      const u = Date.now(), m = await i.ping(
        {
          clientId: (e == null ? void 0 : e.clientId) ?? "undefined",
          projectName: t,
          // required by backend
          lastPingMs: Math.floor(a.getValue())
        },
        {
          onRetry: () => n.notify(!1),
          signal: d
        }
      );
      if (!d.aborted) {
        if (m instanceof Error)
          throw m;
        n.notify(!0), a.setValue(Date.now() - u);
      }
    },
    1e3,
    { immediateTrigger: !0 }
  );
  return {
    pingCallResults: n,
    connectionOk: r,
    pingMs: a,
    terminate: () => {
      n.terminate(), c.terminate(), r.terminate(), a.terminate();
    }
  };
}, Ui = (i, t, e) => {
  const n = new Fe(!0);
  let a = 0n;
  const r = new AbortController();
  let c = new AbortController(), d = !1;
  const u = () => (c.abort(), c = new AbortController(), i.attach(
    { projectName: t, commitIndex: a },
    {
      signal: AbortSignal.any([
        c.signal,
        r.signal
      ])
    }
  )[Symbol.asyncIterator]());
  let m = u();
  const f = 1e3;
  return {
    nextTransactionIterator: async function* () {
      for (; ; )
        try {
          const { value: p, done: k } = await m.next();
          if (k ?? !1)
            throw new Error("document service attach stream closed");
          switch (n.setValue(!0), p.message.case) {
            case "noop": {
              if (!d) {
                d = !0, yield new B({ id: crypto.randomUUID() });
                continue;
              }
              continue;
            }
            case "transaction": {
              a = p.message.value.commitIndex, d = !0, yield p.message.value;
              continue;
            }
            default:
              throw new Error(
                `received attach response with unknown message case: ${p.message.case}`
              );
          }
        } catch (p) {
          if (r.signal.aborted)
            return;
          if (!(p instanceof Wn))
            throw p;
          switch (p.code) {
            case Be.Canceled:
            // thrown if abort controller is aborted
            case Be.Aborted:
            case Be.Unavailable:
            case Be.Unknown: {
              if (n.setValue(!1), await vn(
                f + Math.random() * f * 0.1,
                // pass in master abort controller for early termination
                r.signal
              ), r.signal.aborted)
                return;
              m = u();
              continue;
            }
            // else throw
            default:
              throw p.code === Be.OutOfRange ? new Error(
                "local document state too far in the past, must reload tab"
              ) : p;
          }
        }
    }(),
    terminate: () => r.abort(),
    reconnect: () => c.abort(),
    connectionOk: n
  };
}, Gi = (i, t) => {
  const e = new Fe(!0), n = [], a = $i();
  let r;
  return (async () => {
    for (; ; ) {
      const d = n.splice(0);
      if (d.length === 0) {
        if (r !== void 0) {
          r.resolve();
          return;
        }
        await a.wait;
        continue;
      }
      const u = await i.applyTransactions(
        {
          projectName: t,
          transactions: d.map(([m]) => m)
        },
        {
          callIsOk: e
        }
        // no signal passed: we want to send everything
      );
      if (u instanceof Error)
        throw new Error("error sending transaction", { cause: u });
      d.forEach(([m, f]) => {
        f(u.errors[m.id] ?? void 0);
      });
    }
  })(), {
    sendNextTransaction: async (d) => {
      if (r !== void 0)
        throw new Error("tried sending transaction after termination");
      const { promise: u, resolve: m } = Promise.withResolvers();
      return n.push([d, m]), a.signal(), u;
    },
    terminate: async () => {
      if (r !== void 0) {
        await r.promise;
        return;
      }
      r = Promise.withResolvers(), a.signal(), e.terminate(), await r.promise, n.length > 0 && console.error(
        "invariant violation: nextBatch not empty after termination; have:",
        n.length,
        "transactions left"
      );
    },
    connectionOk: e
  };
}, $i = () => {
  let { promise: i, resolve: t } = Promise.withResolvers();
  return {
    get wait() {
      return i;
    },
    signal: () => {
      t(), { promise: i, resolve: t } = Promise.withResolvers();
    }
  };
}, ji = (i, t, e) => {
  const n = Hi(i, t, {}), a = Ui(
    i,
    t
  ), r = Gi(
    i,
    t
  ), c = Oi(
    n.connectionOk,
    a.connectionOk,
    r.connectionOk
  );
  return n.pingCallResults.subscribe((d) => {
    d || a.reconnect();
  }), {
    receiveNextTransaction: a.nextTransactionIterator,
    sendNextTransaction: r.sendNextTransaction,
    pingMs: n.pingMs,
    connectionOk: c,
    terminate: async () => {
      n.terminate(), a.terminate(), await r.terminate(), c.terminate();
    }
  };
}, Wi = (i, t, e, n) => {
  const a = ji(
    i,
    e
  ), r = t(), c = [], d = [], u = [];
  (async () => {
    for await (const p of a.receiveNextTransaction)
      c.push(p);
  })();
  const f = new Fe(!0);
  a.connectionOk.subscribe((p) => {
    f.setValue(!p);
  }, !0);
  let y = !1;
  return {
    blocked: f,
    send: (p) => {
      if (y)
        throw new Error(
          "tried sending a transaction after gateway was terminated"
        );
      p = p.clone(), p.id = crypto.randomUUID(), d.push(p), a.sendNextTransaction(p).then((k) => {
        k !== void 0 && u.push(p.id);
      });
    },
    synchronize: () => {
      if (y)
        return [];
      c.length > 0;
      const p = r.consolidate(
        c,
        u,
        d
      );
      return c.length = 0, u.length = 0, d.length = 0, p;
    },
    terminate: async () => {
      y = !0, await a.terminate(), c.length = 0, u.length = 0, d.length = 0, f.terminate(), r.terminate();
    }
  };
}, Xn = async () => {
  const i = await Si();
  return {
    validate: (t) => {
      const e = i.applyTransaction(
        new B({ modifications: [t] })
      );
      if (!(e instanceof B))
        return e;
    },
    terminate: () => {
      i.terminate();
    }
  };
};
var je;
const Ee = class Ee {
  constructor(t = void 0, e = void 0, n = []) {
    o(this, "entityId");
    o(this, "fieldIndex");
    o(this, "entityType");
    b(this, je);
    if (this.entityId = t ?? "", this.fieldIndex = n, this.entityType = e, t !== void 0 && t !== "" && e === void 0)
      throw new Error("entity type is required if id is set");
  }
  isEmpty() {
    return this.entityId.length === 0;
  }
  equals(t) {
    return this.entityId === t.entityId && this.fieldIndex.length === t.fieldIndex.length && this.fieldIndex.every((e, n) => e === t.fieldIndex[n]);
  }
  /** @internal */
  equalsPointer(t) {
    return t === void 0 ? this.isEmpty() : this.entityId === t.entityId && this.fieldIndex.length === t.fieldIndex.length && this.fieldIndex.every((e, n) => e === t.fieldIndex[n]);
  }
  /** @internal Returns a copy of this object with an appended field number. */
  withAppendedFieldNumber(t) {
    return new Ee(this.entityId, this.entityType, [
      ...this.fieldIndex,
      t
    ]);
  }
  /** @internal */
  withFieldIndex(t) {
    return new Ee(this.entityId, this.entityType, [...t]);
  }
  /** @internal */
  toPointerMessage() {
    return new q({
      entityId: this.entityId,
      fieldIndex: this.fieldIndex.slice()
    });
  }
  get [gi]() {
    return l(this, je) === void 0 && D(this, je, Vi(this.entityId, this.fieldIndex)), l(this, je);
  }
  /** Returns a human readable string representation */
  toString() {
    if (this.isEmpty())
      return "[empty location]";
    const t = pi(this);
    return `[location: ${this.entityId}${t}]`;
  }
  /** @internal clones the location, updating only the id */
  withId(t) {
    return new Ee(t, this.entityType, this.fieldIndex.slice());
  }
  /** @internal */
  static fromPointerMessage(t, e) {
    if (e.entityId === void 0 || e.entityId === "")
      return new Ee();
    const n = t(e.entityId);
    if (n === void 0)
      throw new Error(
        `entity type for ${e.entityId} is undefined for a set pointer: ${e.toJsonString()}`
      );
    return new Ee(e.entityId, n, e.fieldIndex.slice());
  }
  /** @internal */
  static fromSchemaPath(t, e) {
    const { entityType: n, fieldIndex: a } = hi(e);
    return new Ee(t, n, a);
  }
};
je = new WeakMap();
let v = Ee;
const Vi = (i, t) => Ti(
  `${i}/${t.join(",")}`,
  "4f4aaf81-65f2-4239-b1df-e34a0729f6b8"
), Qn = (i, t, e, n) => {
  const a = {};
  return e.getType().fields.list().forEach((r) => {
    if (r.name === "id" && r.no === 1)
      return;
    const c = n.withAppendedFieldNumber(r.no), d = e[r.localName];
    let u;
    if (!r.repeated)
      u = Jn(
        i,
        t,
        c,
        d,
        r
      );
    else {
      const m = d.map(
        (f, y) => Jn(
          i,
          t,
          c.withAppendedFieldNumber(y),
          f,
          r
        )
      );
      u = new Ze(c, m);
    }
    a[r.localName] = u;
  }), a;
}, Jn = (i, t, e, n, a) => {
  switch (a.kind) {
    case "message": {
      if (n instanceof q)
        return new G(
          e,
          v.fromPointerMessage(i, n),
          "pointer",
          !Cn(t, e.fieldIndex)
        );
      if (a.T === Et)
        return new $({}, e);
      T(n !== void 0, `undefined value for field ${e}`);
      const r = Qn(
        i,
        t,
        n,
        e
      );
      return new $(r, e);
    }
    case "scalar":
      return new G(
        e,
        n,
        Ki[a.T],
        !Cn(t, e.fieldIndex)
      );
    default:
      throw new Error(`unsupported field kind: ${a.kind}`);
  }
}, Ki = {
  [P.DOUBLE]: "double",
  [P.FIXED32]: "fixed32",
  [P.FIXED64]: "fixed64",
  [P.FLOAT]: "float",
  [P.INT32]: "int32",
  [P.INT64]: "int64",
  [P.SFIXED32]: "sfixed32",
  [P.SFIXED64]: "sfixed64",
  [P.SINT32]: "sint32",
  [P.SINT64]: "sint64",
  [P.UINT32]: "uint32",
  [P.UINT64]: "uint64",
  [P.STRING]: "string",
  [P.BOOL]: "bool",
  [P.BYTES]: "bytes"
}, Cn = (i, t) => {
  const e = gn({ entityType: i, fieldIndex: t });
  if ((e == null ? void 0 : e.type) !== "primitive")
    throw "Expected primitive field details, got: " + (e == null ? void 0 : e.type);
  return (e == null ? void 0 : e.immutable) ?? !1;
}, Ri = (i, t) => {
  const e = Ni(t), n = Qn(
    i,
    e,
    t,
    new v(t.id, e, [])
  );
  return new Ge(new v(t.id, e, []), n);
}, Xi = (i, t) => {
  switch (i.case) {
    case "pointer":
      return v.fromPointerMessage(t, i.value);
    case "sfixed64":
    case "fixed64":
    case "int64":
    case "sint64":
    case "uint64":
    case "sfixed32":
    case "fixed32":
    case "bool":
    case "float":
    case "double":
    case "int32":
    case "sint32":
    case "uint32":
    case "string":
      return i.value;
    case "bytes":
      throw "unexpected update message type 'bytes'";
    case void 0:
      return g("update message type is undefined");
  }
}, Ln = (i, t, e) => {
  let n = i.get(e);
  return n === void 0 ? !1 : (n = n.filter((a) => !a.equals(t)), n.length === 0 ? i.delete(e) : i.set(e, n), !0);
}, Bn = (i, t, e) => {
  const n = [...i.get(e) ?? [], t];
  i.set(e, n);
}, kt = (i, t) => {
  Object.values(i.fields).forEach((e) => {
    if (e instanceof G && e.value instanceof v) {
      e.value.isEmpty() || t(e.location, e.value);
      return;
    }
    if (e instanceof Ze) {
      if (e.array[0] instanceof $) {
        e.array.forEach((n) => kt(n, t));
        return;
      }
      if (e.array[0] instanceof G && e.array[0].value instanceof v) {
        e.array.filter(
          (n) => !n.value.isEmpty()
        ).forEach(
          (n) => t(e.location, n.value)
        );
        return;
      }
      return;
    }
    e instanceof $ && kt(e, t);
  });
}, Qi = (i, t, e, n) => {
  const a = i._resolveField(t.fieldIndex) ?? g(`can't find updated field ${t}`);
  if (!(a instanceof G))
    throw "received update on non-primitive field";
  if (!(e instanceof v)) {
    a._setValue(e);
    return;
  }
  const r = a.value;
  r.isEmpty() || n == null || n.onStopPointingTo(a.location, r), a._setValue(e);
  const c = e;
  c.isEmpty() || n == null || n.onStartPointingTo(a.location, c);
}, Tn = (i) => {
  const t = (i == null ? void 0 : i.entities) ?? /* @__PURE__ */ new Map(), e = (i == null ? void 0 : i.references) ?? new vt(), n = {
    onCreate: () => {
    },
    onDelete: () => {
    },
    onStartPointingTo: () => {
    },
    onStopPointingTo: () => {
    },
    onUpdate: () => {
    },
    ...(i == null ? void 0 : i.callbacks) ?? {}
  }, a = /* @__PURE__ */ new Map(), r = (m) => {
    var f;
    return ((f = t.get(m)) == null ? void 0 : f.entityType) ?? a.get(m);
  }, c = (m) => {
    const f = Ri(
      r,
      Fi(
        m.entity ?? g("received empty create modification")
      )
    );
    t.set(f.id, f), kt(f, (y, p) => {
      Bn(e, y, p), n.onStartPointingTo(y, p);
    }), n.onCreate(f);
  }, d = (m) => {
    const f = m.entityId, y = t.get(f) ?? g("can't find deleted entity");
    t.delete(f), kt(y, (p, k) => {
      Ln(e, p, k), n.onStopPointingTo(p, k);
    }), n.onDelete(y);
  }, u = (m) => {
    const f = m.field ?? g("received update without pointer"), y = v.fromPointerMessage(r, f), p = t.get(y.entityId) ?? g("can't find updated entity"), k = Xi(m.value, r);
    Qi(p, y, k, {
      onStopPointingTo: (S, N) => {
        Ln(e, S, N) || g(), n.onStopPointingTo(S, N);
      },
      onStartPointingTo: (S, N) => {
        Bn(e, S, N), n.onStartPointingTo(S, N);
      }
    }), n.onUpdate(y, k);
  };
  return {
    entities: t,
    references: e,
    applyModification(m) {
      const f = m.modification;
      switch (f.case) {
        case "create": {
          c(f.value);
          break;
        }
        case "delete": {
          d(f.value);
          break;
        }
        case "update": {
          u(f.value);
          break;
        }
      }
    },
    getStats() {
      return {
        entities: t.size,
        references: e.values().reduce((m, f) => m + f.length, 0)
      };
    },
    _addEntityTypeForId(m, f) {
      a.set(m, f);
    }
  };
}, _e = {
  terminableFrom: (...i) => ({
    terminate: () => {
      for (const t of i)
        t();
    }
  })
};
var z, De, C, O, H, ct;
const Ue = class Ue {
  constructor(t = Tn()) {
    b(this, z, new vt());
    b(this, De, /* @__PURE__ */ new Map());
    /** contains either: entity id, *, or entity type */
    b(this, C, /* @__PURE__ */ new Map());
    b(this, O, new vt());
    b(this, H, new vt());
    b(this, ct);
    D(this, ct, t), Ue.debugWindowInstance ?? (Ue.debugWindowInstance = this);
  }
  /**
   * Subscribe to the event that an entity of a specific type is created.
   *
   * @returns A terminable that when terminated will stop dispatching new onCreate events. Cleanup functions that were returned
   * during entity creation will still be called on removal of the entity.
   */
  onCreate(t, e) {
    const n = (r) => {
      const c = e(
        r
      );
      c !== void 0 && this.onRemove(r, c);
    };
    return tt(l(this, De), t).add(n), _e.terminableFrom(
      () => {
        var r;
        return (r = l(this, De).get(t)) == null ? void 0 : r.delete(n);
      }
    );
  }
  /** @internal */
  _dispatchCreate(t) {
    [
      ...l(this, De).get("*") ?? [],
      ...l(this, De).get(t.entityType) ?? []
    ].forEach((e) => e(t));
  }
  /**
   * Subscribe to updates of a mutable primitive field in the nexus document.
   *
   * @returns A terminable that when terminated will stop dispatching new onUpdate events.
   */
  onUpdate(t, e, n = !0) {
    const a = e;
    return tt(l(this, z), t.location).add(a), n && e(t.value), _e.terminableFrom(
      () => {
        var c;
        return (c = l(this, z).get(t.location)) == null ? void 0 : c.delete(a);
      }
    );
  }
  /** @internal */
  _dispatchUpdate(t, e) {
    var n;
    (n = l(this, z).get(t)) == null || n.forEach((a) => a(e));
  }
  /**
   * Subscribe to an event where an entity is removed.
   *
   * @example
   * ```ts
   * const tm = await nexus.modify(t => t.create("tonematrix", {}))
   * nexus.events.onRemove(tm, (tm) => console.debug("tonematrix", tm.id, "removed"))
   * ```
   *
   *
   * @returns A terminable that when terminated will stop dispatching new onRemove events.
   */
  onRemove(t, e) {
    const n = e;
    return tt(
      l(this, C),
      typeof t == "string" ? t : t.id
    ).add(n), _e.terminableFrom(
      () => {
        var r;
        return (r = l(this, C).get(typeof t == "string" ? t : t.id)) == null ? void 0 : r.delete(n);
      }
    );
  }
  /** @internal */
  _dispatchRemove(t) {
    [
      ...l(this, C).get("*") ?? [],
      ...l(this, C).get(t.id) ?? [],
      ...l(this, C).get(t.entityType) ?? []
    ].forEach((e) => e(t)), l(this, C).delete(t.id), [...l(this, O).keys()].filter((e) => e.entityId === t.id).forEach((e) => l(this, O).delete(e)), [...l(this, H).keys()].filter((e) => e.entityId === t.id).forEach((e) => l(this, H).delete(e)), [...l(this, z).keys()].filter((e) => e.entityId === t.id).forEach((e) => l(this, z).delete(e));
  }
  /** Subscribe to the event that some pointer in the document starts pointing to a given location.
   *
   * @example
   * ```ts
   * const tm = await nexus.modify(t => t.create("tonematrix", {}))
   * nexus.events.onPointingTo(tm.fields.audioOutput, (from) =>
   *  console.debug(
   *    "pointing from field",
   *    from.toString(),
   *    "which is entity",
   *    nexus.queryEntities.getEntity(from.entityId)?.id
   * ))
   * ```
   *
   * If the pointer is the result of an entity being created, then `onCreate` is called before this callback.
   *
   * @returns A terminable that when terminated will stop dispatching new onPointingTo events.
   */
  onPointingTo(t, e, n = !0) {
    var r;
    return tt(l(this, O), t).add(e), n && ((r = l(this, ct).references.get(t)) == null || r.forEach((c) => e(c))), _e.terminableFrom(
      () => {
        var c;
        return (c = l(this, O).get(t)) == null ? void 0 : c.delete(e);
      }
    );
  }
  /** @internal */
  _dispatchPointingTo(t, e) {
    var n;
    (n = l(this, O).get(t)) == null || n.forEach((a) => a(e));
  }
  /** Subscribe to the event that some pointer in the document stops pointing to a given location.
   *
   * @example
   * ```ts
   * const tm = await nexus.modify(t => t.create("tonematrix", {}))
   * nexus.events.onStopPointingTo(tm.fields.audioOutput, (from) =>
   *  console.debug("pointing from field", from.toString(), "which is entity", nexus.queryEntities.getEntity(from.entityId)?.id
   * ))
   * ```
   *
   * If the pointer is the result of an entity being removed, then `onRemove` is called after this callback.
   *
   * @returns A terminable that when terminated will stop dispatching new onStopPointingTo events.
   */
  onStopPointingTo(t, e) {
    return tt(l(this, H), t).add(e), _e.terminableFrom(
      () => {
        var a;
        return (a = l(this, H).get(t)) == null ? void 0 : a.delete(e);
      }
    );
  }
  /** @internal */
  _dispatchStopPointingTo(t, e) {
    var n;
    (n = l(this, H).get(t)) == null || n.forEach((a) => a(e));
  }
  /**
   * @internal
   *
   * Removes all event listeners.*/
  _clear() {
    l(this, De).clear(), l(this, z).clear(), l(this, C).clear(), l(this, O).clear(), l(this, H).clear();
  }
  /**
   * @internal
   * Some stats for debugging. Static for easy access. This method can get quite slow, 6.5ms measured
   * on large documents. Don't call too often. */
  getStats() {
    return {
      numCreateListeners: et(l(this, De)),
      numUpdateListeners: et(l(this, z)),
      numRemoveListeners: et(l(this, C)),
      numPointingToListeners: et(l(this, O)),
      numStopPointingToListeners: et(l(this, H))
    };
  }
};
z = new WeakMap(), De = new WeakMap(), C = new WeakMap(), O = new WeakMap(), H = new WeakMap(), ct = new WeakMap(), /**
 * @internal
 * For debugging purposes, the first instance of the nexus event manager ever created. */
o(Ue, "debugWindowInstance");
let mn = Ue;
const et = (i) => i.values().reduce((t, e) => t + e.size, 0), tt = (i, t) => i.get(t) ?? i.set(t, /* @__PURE__ */ new Set()).get(t) ?? g(), Yn = () => {
  let i = !1, t = !1;
  return {
    synchronize: () => t ? "done" : (t = !0, [new B()]),
    send: () => {
      if (i)
        throw new Error("Gateway terminated");
    },
    blocked: new Fe(!1),
    terminate: async () => {
      i = !0;
    }
  };
}, Zn = () => ({
  validate: () => {
  },
  terminate: () => {
  }
});
var We, ze, Ve, Ke, Pe, it, _n;
const Sn = class Sn {
  constructor({
    getFields: t,
    getRefs: e,
    filterFields: n,
    documentLock: a
  }) {
    b(this, Pe);
    b(this, We);
    b(this, ze);
    b(this, Ve);
    b(this, Ke);
    D(this, We, t), D(this, ze, e), D(this, Ve, n), D(this, Ke, a);
  }
  /** Only keep fields that are marked with target type appearing in `targetTypes`. */
  ofTargetTypes(...t) {
    return x(this, Pe, it).call(this, (e) => gn(e.location).targetTypes.some(
      (n) => t.includes(n)
    ));
  }
  /** Only keep fields that aren't pointed to by any other field in the nexus document. */
  notPointedTo() {
    return x(this, Pe, it).call(this, (t) => {
      const e = l(this, ze).call(this).get(t.location);
      return e === void 0 || e.length === 0;
    });
  }
  pointedToBy(t) {
    return x(this, Pe, it).call(this, (e) => {
      const n = l(this, ze).call(this).get(e.location);
      return n !== void 0 && n.some((a) => a.equals(t));
    });
  }
  primitiveFields() {
    return x(this, Pe, it).call(this, (t) => t instanceof G);
  }
  /** Returns all primitive fields selected using this query */
  get() {
    return l(this, We).call(this).filter((t) => l(this, Ve).call(this, t));
  }
  /** Returns the first primitive field of the result, or undefined if the query is empty.
   */
  getOne() {
    return this.get()[0];
  }
  /** Returns all primitive fields selected using this query, as
   * part of a ComparableMap that maps entity ids to fields for that entity.
   */
  getByEntity() {
    x(this, Pe, _n).call(this);
    const t = /* @__PURE__ */ new Map();
    return this.get().forEach((e) => {
      t.has(e.location.entityId) || t.set(e.location.entityId, []), (t.get(e.location.entityId) ?? g()).push(e);
    }), t;
  }
};
We = new WeakMap(), ze = new WeakMap(), Ve = new WeakMap(), Ke = new WeakMap(), Pe = new WeakSet(), it = function(t) {
  return new Sn({
    getFields: l(this, We),
    getRefs: l(this, ze),
    filterFields: (e) => l(this, Ve).call(this, e) && t(e),
    documentLock: l(this, Ke)
  });
}, _n = function() {
  var t;
  if (!(((t = l(this, Ke)) == null ? void 0 : t.locked) ?? !0))
    throw new Error("Document is not locked");
};
let fn = Sn;
var I, Me, Oe, M, J, Le;
const Dn = class Dn {
  /** Don't construct this class, it is constructed by NexusDocument. */
  constructor(t) {
    b(this, M);
    b(this, I);
    /** This function filters & transforms entities. When modifying a query,
     * (e.g. with `ofTypes`, we basically "append" another function
     * to this one that filters and casts the entities.
     *
     * When calling `run`, we get all entities and filter them using this function
     * before returning all existing entities.
     */
    b(this, Me);
    b(this, Oe);
    D(this, I, (t == null ? void 0 : t.documentState) ?? Tn()), D(this, Me, (t == null ? void 0 : t.filterEntities) ?? ((e) => e)), D(this, Oe, t == null ? void 0 : t.documentLock);
  }
  /** Returns all entities selected by this query, in undefined order.
   */
  get() {
    return x(this, M, Le).call(this), [...l(this, Me).call(this, l(this, I).entities).values()];
  }
  /** Returns the first entity returned by `get()`, if any.
   *
   *  Since the order of entities in `get()` is undefined, which of the selected
   * entity this method returns is also undefined. This method is intended to be used
   * if it's known that the query will return at most one entity.
   */
  getOne() {
    return x(this, M, Le).call(this), this.get()[0];
  }
  /** Of all selected entities, return the one with id `id`, if it exists. */
  getEntity(t) {
    return x(this, M, Le).call(this), l(this, Me).call(this, l(this, I).entities).get(t);
  }
  /**
   * Of all selected entities, return the one with id `id`. Throw if it doesn't
   * exist.
   */
  mustGetEntity(t) {
    return this.getEntity(t) ?? g(`can't find entity with uuid ${t}`);
  }
  /**
   * Get an entity as a specific type, if it exists and has the type matching one
   * of the provided types.
   */
  getEntityAs(t, ...e) {
    x(this, M, Le).call(this), T(e.length > 0, "must provide at least one type");
    const n = this.getEntity(t);
    if (n !== void 0 && e.includes(n.entityType))
      return n;
  }
  /**
   * Get an entity as a specific type, if it has the type matching one of the
   * provided types. Throw if it doesn't exist.
   */
  mustGetEntityAs(t, ...e) {
    x(this, M, Le).call(this), T(e.length > 0, "must provide at least one type");
    const n = this.mustGetEntity(t);
    return e.includes(n.entityType) ? n : g(
      `entity with uuid ${t} is not of any of the provided types ${e}`
    );
  }
  /** Only keep entities whose id appears in `ids`. */
  withIds(...t) {
    return x(this, M, J).call(this, (e) => Ce(e, t));
  }
  /** Return the `FieldQuery<NexusField>` that starts with all fields of all
   * currently selected entities.
   */
  fields() {
    return new fn({
      getFields: () => [
        ...l(this, Me).call(this, l(this, I).entities).values()
      ].flatMap((t) => ra(t)),
      getRefs: () => l(this, I).references,
      filterFields: () => !0,
      documentLock: l(this, Oe)
    });
  }
  /** Only keep entities whose messages are marked with a target type appearing
   * in `targetTypes`. Target types of fields of entities are ignored.
   */
  ofTargetTypes(...t) {
    return x(this, M, J).call(this, (e) => {
      const n = /* @__PURE__ */ new Map();
      return e.forEach((a, r) => {
        gn(a.location).targetTypes.some(
          (c) => t.includes(c)
        ) && n.set(r, a);
      }), n;
    });
  }
  /** Check if a specific entity is contained in the current query. */
  has(t) {
    x(this, M, Le).call(this);
    const e = typeof t == "string" ? t : t.id;
    return l(this, Me).call(this, l(this, I).entities).has(e);
  }
  /** Only keep entities whose type string appears in `types`. */
  ofTypes(...t) {
    return x(this, M, J).call(this, (e) => Yi(e, t));
  }
  /** Omit entities whose type string appears in `types`. */
  notOfTypes(...t) {
    return x(this, M, J).call(this, (e) => Zi(e, t));
  }
  /** Only keep entities that have some fields that point to:
   * * `entityOfType`: some field of entities of a set of types
   * * `locations`: specific locations
   * * `entities`: some field of specific entities
   *
   * Passing an empty list to any of these methods will result in an empty query result.
   *
   * Use e.g. like:
   * ```
   * nexus.entities.pointingTo.entitiesOfType("tb303").get()
   * ```
   */
  get pointingTo() {
    return {
      entityOfType: (...t) => x(this, M, J).call(this, (e) => Ce(
        e,
        na(
          l(this, I).references,
          l(this, I).entities,
          ...t
        )
      )),
      locations: (...t) => x(this, M, J).call(this, (e) => Ce(
        e,
        _i(l(this, I).references, t)
      )),
      entities: (...t) => x(this, M, J).call(this, (e) => Ce(
        e,
        ea(l(this, I).references, t)
      ))
    };
  }
  /** Only keep entities that are themselves, or have fields that are, pointed to by:
   * * `entityOfType`: some field of entities of a set of types
   * * `locations`: specific locations
   * * `entities`: some field of specific entities
   *
   * Passing an empty list to any of these methods will result in an empty query result.
   *
   * Use e.g. as:
   * ```
   * nexus.entities.pointedToBy.entitiesOfType("tb303").get()
   * ```
   */
  get pointedToBy() {
    return {
      entityOfType: (...t) => x(this, M, J).call(this, (e) => Ce(
        e,
        ia(
          l(this, I).references,
          l(this, I).entities,
          ...t
        )
      )),
      locations: (...t) => x(this, M, J).call(this, (e) => Ce(
        e,
        ta(l(this, I).references, t)
      )),
      entities: (...t) => x(this, M, J).call(this, (e) => Ce(
        e,
        aa(l(this, I).references, t)
      ))
    };
  }
};
I = new WeakMap(), Me = new WeakMap(), Oe = new WeakMap(), M = new WeakSet(), /** Return a copy of this class, with a new function "appended" to the
 * filter pipeline. The filter can change the type of the entity, or
 * remove entities from the query.
 */
J = function(t) {
  return new Dn({
    documentState: l(this, I),
    filterEntities: (e) => t(l(this, Me).call(this, e)),
    documentLock: l(this, Oe)
  });
}, Le = function() {
  var t;
  if (!(((t = l(this, Oe)) == null ? void 0 : t.locked) ?? !0))
    throw new Error("EntityQuery method called without locking the document");
};
let St = Dn;
const Yi = (i, t) => {
  const e = /* @__PURE__ */ new Map();
  return [...i.values()].filter((n) => t.includes(n.entityType)).forEach((n) => e.set(n.id, n)), e;
}, Zi = (i, t) => {
  const e = /* @__PURE__ */ new Map();
  return [...i.values()].filter((n) => !t.includes(n.entityType)).forEach((n) => e.set(n.id, n)), e;
}, Ce = (i, t) => {
  const e = /* @__PURE__ */ new Map();
  return t.forEach((n) => {
    i.get(n) !== void 0 && e.set(n, i.get(n));
  }), e;
}, _i = (i, t) => t.flatMap((e) => {
  var n;
  return ((n = i.get(e)) == null ? void 0 : n.map((a) => a.entityId)) ?? [];
}), ea = (i, t) => [...i.entries()].filter(([e, n]) => t.some((a) => e.entityId === a)).flatMap(([e, n]) => n.map((a) => a.entityId)), ta = (i, t) => [...i.entries()].filter(
  ([e, n]) => t.some((a) => n.some((r) => a.equals(r)))
).map(([e, n]) => e.entityId), na = (i, t, ...e) => [...i.entries()].filter(([n, a]) => {
  const r = t.get(n.entityId);
  return r !== void 0 && e.includes(r.entityType);
}).flatMap(([n, a]) => a.map((r) => r.entityId)), ia = (i, t, ...e) => [...i.entries()].filter(
  ([n, a]) => a.some((r) => {
    const c = t.get(r.entityId);
    return c !== void 0 && e.includes(c.entityType);
  })
).map(([n, a]) => n.entityId), aa = (i, t) => [...i.entries()].filter(
  ([e, n]) => t.some((a) => n.some((r) => r.entityId === a))
).map(([e, n]) => e.entityId), ra = (i) => {
  const t = (e) => {
    if (e instanceof G)
      return [e];
    if (e instanceof Ze)
      return [e, ...e.array.flatMap((n) => t(n))];
    if (e instanceof $) {
      const n = Object.values(
        e.fields
      ).flatMap((a) => t(a));
      return e instanceof Ge ? n : [e, ...n];
    }
    throw new Error(`field of unknown type: ${e}`);
  };
  return t(i);
}, Dt = (i) => new He({
  modification: {
    case: "create",
    value: {
      entity: i instanceof Qe ? i : Ye(i)
    }
  }
}).clone(), oa = (i, t, e) => new He({
  modification: {
    case: "update",
    value: {
      field: i.toPointerMessage(),
      value: {
        case: t,
        value: e
      }
    }
  }
}).clone(), sa = (i) => new He({
  modification: {
    case: "delete",
    value: { entityId: i }
  }
}).clone(), h = {
  positionX: 0,
  positionY: 0,
  presetName: ""
}, ca = {
  ...h,
  displayName: "AudioDevice",
  gain: 0.7079399824142456,
  panning: 0,
  isActive: !0
}, la = {
  ...h,
  displayName: "Merger",
  blendModeIndex: 1,
  mergeCoords: {
    x: 0.3333333432674408,
    y: 0
  }
}, ua = {
  ...h,
  displayName: "Audio Splitter",
  blendModeIndex: 1,
  splitCoords: {
    x: 0.6666666865348816,
    y: 0
  }
}, da = {
  ...h,
  displayName: "Auto Filter",
  thresholdGain: 0.5,
  attackMs: 62.5,
  sustainMs: 62.5,
  releaseMs: 62.5,
  filterModeIndex: 1,
  cutoffFrequencyHz: 300,
  filterModulationDepth: 1,
  filterResonance: 1.4142135381698608,
  gain: 1,
  mix: 1,
  isActive: !0,
  presetName: ""
}, ma = {
  ...h,
  displayName: "Band Splitter",
  filterLowHz: 360,
  filterHighHz: 3600,
  highGain: 1,
  midGain: 1,
  lowGain: 1,
  presetName: ""
}, fa = {
  key: 36,
  transposeOctaves: 0,
  isActive: !0,
  doesSlide: !1,
  isAccented: !1
}, pa = {
  ...h,
  steps: Array.from({ length: 99 }, () => fa),
  length: 16,
  groove: new v()
}, ha = {
  ...h,
  displayName: "Bassline",
  gain: 0.70794,
  tuneSemitones: 0,
  cutoffFrequencyHz: 220,
  filterResonance: 1,
  filterEnvelopeModulationDepth: 0.10000000149011612,
  filterDecay: 0,
  accent: 1,
  waveformIndex: 1,
  patternIndex: 0,
  isActive: !0,
  presetName: ""
}, ya = {
  bassdrumIsActive: !1,
  snaredrumIsActive: !1,
  tomCongaLowIsActive: !1,
  tomCongaMidIsActive: !1,
  tomCongaHighIsActive: !1,
  rimClavesIsActive: !1,
  clapMaracasIsActive: !1,
  cowbellIsActive: !1,
  cymbalIsActive: !1,
  openHihatIsActive: !1,
  closedHihatIsActive: !1,
  isAccented: !1
}, ga = {
  ...h,
  length: 16,
  stepScaleIndex: 3,
  groove: new v(),
  steps: Array.from({ length: 64 }, () => ya)
}, va = {
  ...h,
  displayName: "Beatbox 8",
  gain: 0.70794,
  accentAmount: 0.5,
  patternIndex: 0,
  isActive: !0,
  bassdrum: {
    gain: 1,
    tone: 0.6,
    decay: 0.2750000059604645
  },
  snaredrum: {
    gain: 1,
    tone: 0.5,
    snappy: 1
  },
  tomCongaLow: {
    gain: 1,
    tuning: 0,
    instrumentTypeIndex: 1
  },
  tomCongaMid: {
    gain: 1,
    tuning: 0,
    instrumentTypeIndex: 1
  },
  tomCongaHigh: {
    gain: 1,
    tuning: 0.6000000238418579,
    instrumentTypeIndex: 1
  },
  rimClaves: {
    gain: 1,
    instrumentTypeIndex: 1
  },
  clapMaracas: {
    gain: 1,
    instrumentTypeIndex: 1
  },
  cowbell: {
    gain: 1
  },
  cymbal: {
    gain: 1,
    tone: 0,
    decay: 0
  },
  openHihat: {
    gain: 1,
    decay: 0.8220000267028809
  },
  closedHihat: {
    gain: 1
  },
  presetName: ""
}, wa = {
  bassdrumStepIndex: 0,
  snaredrumStepIndex: 0,
  tomLowStepIndex: 0,
  tomMidStepIndex: 0,
  tomHighStepIndex: 0,
  rimStepIndex: 0,
  clapStepIndex: 0,
  closedHihatStepIndex: 0,
  openHihatStepIndex: 0,
  crashStepIndex: 0,
  rideStepIndex: 0
}, ba = {
  ...h,
  length: 16,
  stepScaleIndex: 3,
  groove: new v(),
  steps: Array.from({ length: 64 }, () => wa)
}, Ta = {
  ...h,
  displayName: "Beatbox 9",
  gain: 0.7079399824142456,
  accentAmount: 0.5,
  patternIndex: 0,
  isActive: !0,
  bassdrum: {
    gain: 1,
    tone: Math.fround(0.6),
    attack: 1,
    decay: 0.3149999976158142
  },
  snaredrum: {
    gain: 1,
    tune: Math.fround(0.7),
    tone: 0.1837099939584732,
    snappy: 1
  },
  tomLow: {
    gain: 1,
    tune: 0,
    decay: Math.fround(0.671)
  },
  tomMid: {
    gain: 1,
    tune: 0,
    decay: Math.fround(0.671)
  },
  tomHigh: {
    gain: 1,
    tune: 0,
    decay: Math.fround(0.671)
  },
  rim: {
    gain: 1
  },
  clap: {
    gain: 1
  },
  hihat: {
    gain: 1,
    closedDecay: 0.5,
    openDecay: 0.5
  },
  crash: {
    gain: 1,
    tune: 0.5
  },
  ride: {
    gain: 1,
    tune: 0.5
  },
  presetName: ""
}, xa = {
  ...h,
  displayName: "Centroid",
  postGain: 1,
  panning: 0,
  aux1: {
    sendGain: 1
  },
  aux2: {
    sendGain: 1
  }
}, ka = {
  displayName: "Channel",
  postGain: 1,
  panning: 0,
  orderAmongChannels: 0,
  isMuted: !1,
  isSoloed: !1,
  useAuxPreMode: !1,
  preGain: 1,
  eqHighGainDb: 0,
  eqMidFrequency: 1e3,
  eqMidGainDb: 0,
  eqLowGainDb: 0,
  aux1SendGain: 0,
  aux2SendGain: 0
}, Sa = {
  tempoBpm: 125,
  baseFrequencyHz: 440,
  signatureNumerator: 4,
  signatureDenominator: 4,
  durationTicks: 1966080
}, Da = {
  ...h,
  displayName: "Crossfader",
  postGain: 1,
  crossfade: 0,
  panning: 0,
  blendModeIndex: 1,
  channelA: {
    preGain: 1,
    eqLowFrequencyHz: 260,
    eqLowGainDb: 0,
    lowKillEnabled: !1,
    eqMidFrequencyHz: 1500,
    eqMidGainDb: 0,
    midKillEnabled: !1,
    eqHighFrequencyHz: 4200,
    eqHighGainDb: 0,
    highKillEnabled: !1
  },
  channelB: {
    preGain: 1,
    eqLowFrequencyHz: 260,
    eqLowGainDb: 0,
    lowKillEnabled: !1,
    eqMidFrequencyHz: 1500,
    eqMidGainDb: 0,
    midKillEnabled: !1,
    eqHighFrequencyHz: 4200,
    eqHighGainDb: 0,
    highKillEnabled: !1
  },
  presetName: ""
}, Ma = {
  ...h,
  displayName: "Curve",
  isActive: !0,
  gainDb: 0,
  spectrumModeIndex: 3,
  peak1: {
    isEnabled: !1,
    centerFrequencyHz: 240,
    gainDb: 0,
    q: 1
  },
  peak2: {
    isEnabled: !0,
    centerFrequencyHz: 1e3,
    gainDb: 0,
    q: 1
  },
  peak3: {
    isEnabled: !1,
    centerFrequencyHz: 4e3,
    gainDb: 0,
    q: 1
  },
  lowPass: {
    cutoffFrequencyHz: 18e3,
    q: 0.71,
    isEnabled: !1,
    filterSlopeIndex: 1
  },
  highPass: {
    cutoffFrequencyHz: 40,
    q: 0.71,
    isEnabled: !1,
    filterSlopeIndex: 1
  },
  lowShelf: {
    centerFrequencyHz: 80,
    gainDb: 0,
    isEnabled: !0
  },
  highShelf: {
    centerFrequencyHz: 1e4,
    gainDb: 0,
    isEnabled: !0
  },
  presetName: ""
}, Ia = {
  colorIndex: 0
}, Na = {
  colorIndex: 0
}, Pa = {
  ...h,
  displayName: "Exciter",
  toneFrequencyHz: 3500,
  powerFactor: 0.5,
  mix: 1,
  isActive: !0,
  presetName: ""
}, Fa = "56ada375-bc78-4912-b447-d80e06457a80", Aa = {
  ...h,
  displayName: "Gakki",
  soundfontId: Fa,
  gain: 0.6,
  presetName: ""
}, Ea = {
  ...h,
  displayName: "",
  pluginPath: "",
  state: new Uint8Array(0),
  isActive: !0
}, qa = {
  ...h,
  displayName: "Graphical EQ",
  filter1: {
    gainDb: 0,
    frequencyHz: 65.406,
    q: 0.7386,
    stereoSeparation: 0
  },
  filter2: {
    gainDb: 0,
    frequencyHz: 8372.02,
    q: 0.7386,
    stereoSeparation: 0
  },
  mix: 1,
  gainDb: 0,
  isActive: !0,
  presetName: ""
}, Ja = {
  ...h,
  displayName: "Gravity",
  thresholdDb: -10,
  ratio: 4,
  kneeDbRange: 3,
  makeupGainDb: 0,
  attackMs: 5,
  releaseTimeNormalized: 0.5,
  rmsWindowMs: 5,
  releaseIsSynced: !1,
  isActive: !0,
  presetName: ""
}, Ca = {
  functionIndex: 1,
  durationTicks: 1920,
  impact: 0,
  displayName: ""
}, yt = {
  gain: 0,
  panning: 0,
  detuneFactor: 1,
  frequencyOffsetHz: 0,
  waveformIndex: 1,
  usePitchEnvelope: !1,
  modulationFactorA: 0,
  modulationFactorB: 0,
  modulationFactorC: 0,
  modulationFactorD: 0,
  velocityAmplitudeModulationDepth: 0,
  envelopeMainAmplitudeModulationDepth: 0,
  envelope2AmplitudeModulationDepth: 0,
  envelope3AmplitudeModulationDepth: 0,
  lfo1AmplitudeModulationDepth: 0,
  lfo2AmplitudeModulationDepth: 0
}, Ft = {
  isSynced: !1,
  attackTimeNormalized: 0,
  attackSlopeFactor: -0.5,
  decayTimeNormalized: 0.25,
  decaySlopeFactor: 0.5,
  decayIsLooped: !1,
  sustainFactor: 0.5,
  releaseTimeNormalized: 0.25,
  releaseSlopeFactor: 0.5
}, zn = {
  isSynced: !0,
  doesRestart: !1,
  rateNormalized: 0.5,
  offsetFactor: 0,
  delayTimeNormalized: 0,
  blendTimeNormalized: 0,
  waveformIndex: 1
}, La = {
  ...h,
  displayName: "Heisenberg",
  isActive: !0,
  microTuning: new v(),
  tuneSemitones: 0,
  gain: 0.7079460024833679,
  playModeIndex: 3,
  glideMs: 0,
  velocityFactor: 1,
  unisonoCount: 1,
  unisonoDetuneSemitones: Math.fround(1e-3),
  unisonoStereoSpreadFactor: 0.5,
  operatorDetuneModeIndex: 1,
  operatorA: {
    ...yt,
    gain: 1
  },
  operatorB: yt,
  operatorC: yt,
  operatorD: yt,
  envelopeMain: Ft,
  envelope2: Ft,
  envelope3: Ft,
  pitchEnvelope: {
    isSynced: !1,
    loopDecayIndex: 1,
    attackRangeFactor: 1,
    attackTimeNormalized: 0.5,
    attackSlopeFactor: 0,
    decayRangeFactor: 0,
    decayTimeNormalized: 0.5,
    decaySlopeFactor: 0,
    sustainRangeFactor: 0,
    releaseTimeNormalized: 0.5,
    releaseSlopeFactor: 0,
    releaseRangeFactor: 0,
    semitoneRange: 12
  },
  lfo1: zn,
  lfo2: zn,
  filter: {
    cutoffFrequencyHz: 11025,
    resonance: 0.7071067690849304,
    filterType: -1,
    orderIndex: 1,
    velocityCutoffModulationDepth: 0,
    envelopeMainCutoffModulationDepth: 0,
    envelope2CutoffModulationDepth: 0,
    envelope3CutoffModulationDepth: 0,
    lfo1CutoffModulationDepth: 0,
    lfo2CutoffModulationDepth: 0,
    keyboardTrackingAmount: 0
  },
  presetName: ""
}, Ba = {
  ...h,
  displayName: "Helmholtz",
  microTuning: new v(),
  gain: 0.7079399824142456,
  decayTime: 0.75,
  mix: 1,
  filters: [
    {
      isActive: !0,
      gain: 1,
      panning: 0,
      frequencyNote: 60,
      frequencyTuneSemitones: 0
    },
    {
      isActive: !0,
      gain: 1,
      panning: 0,
      frequencyNote: 42,
      frequencyTuneSemitones: 0
    },
    {
      isActive: !0,
      gain: 1,
      panning: 0,
      frequencyNote: 65,
      frequencyTuneSemitones: 0
    },
    {
      isActive: !1,
      gain: 1,
      panning: 0,
      frequencyNote: 60,
      frequencyTuneSemitones: 0
    },
    {
      isActive: !1,
      gain: 1,
      panning: 0,
      frequencyNote: 60,
      frequencyTuneSemitones: 0
    }
  ],
  isActive: !0,
  presetName: ""
}, za = {
  ...h,
  displayName: "Kobolt",
  postGain: 1,
  channels: [
    ...Array.from({ length: 16 }, () => ({
      gain: 1,
      panning: 0
    }))
  ]
}, Oa = {
  isActive: !1,
  modulationDepth: 1
}, Ha = {
  isExclusive: !1,
  isMuted: !1,
  steps: Array.from({ length: 128 }, () => Oa)
}, Ua = {
  ...h,
  length: 16,
  stepScaleIndex: 1,
  groove: new v(),
  channelPatterns: Array.from(
    { length: 9 },
    () => Ha
  )
}, Ga = {
  sample: new v(),
  startTrimFactor: 0,
  startTrimModulationDepth: 0,
  endTrimFactor: 1,
  endTrimModulationDepth: 0,
  pitchSemitones: 0,
  pitchModulationDepth: 0,
  filterTypeIndex: 1,
  cutoffFrequencyHz: 6e3,
  cutoffModulationDepth: 0,
  resonance: 0,
  resonanceModulationDepth: 0,
  envelopePeakRatio: 0,
  envelopeRatioModulationDepth: 0,
  envelopeSlope: 1,
  envelopeSlopeModulationDepth: 0,
  panning: 0,
  panningModulationDepth: 0,
  gain: 0.1,
  gainModulationDepth: 1
}, $a = {
  ...h,
  displayName: "Machiniste",
  channels: new Array(9).fill(Ga),
  mainOutputGain: 0.70794,
  globalModulationDepth: 1,
  patternIndex: 0,
  isActive: !0,
  presetName: ""
}, ja = {
  overrideVelocity: !1,
  stepVelocity: 1,
  isMuted: !1,
  isTied: !1,
  isChord: !1
}, Wa = {
  ...h,
  length: 16,
  groove: new v(),
  steps: Array.from(
    { length: 64 },
    () => ja
  )
}, Va = {
  ...h,
  displayName: "Matrix Arpeggiator",
  isActive: !0,
  velocity: 1,
  stepLengthIndex: 7,
  repeat: 1,
  gateRatio: 1,
  arpeggiationModeIndex: 1,
  randomSeed: 1e3,
  octaves: 1,
  holdNotes: !1,
  holdNotesUntilNote: 0,
  ignorePatternStepParameters: !1,
  patternIsSynced: !1,
  patternIndex: 0,
  presetName: ""
}, Ka = {
  displayName: "Micro Tuning",
  semitones: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
}, Ra = {
  ...h,
  displayName: "Minimixer",
  gain: 1,
  auxSendGain: 1,
  auxIsPreGain: !0,
  channel1: {
    gain: 1,
    panning: 0,
    auxSendGain: 0,
    auxIsPreGain: !1,
    isMuted: !1,
    isSoloed: !1
  },
  channel2: {
    gain: 1,
    panning: 0,
    auxSendGain: 0,
    auxIsPreGain: !1,
    isMuted: !1,
    isSoloed: !1
  },
  channel3: {
    gain: 1,
    panning: 0,
    auxSendGain: 0,
    auxIsPreGain: !1,
    isMuted: !1,
    isSoloed: !1
  },
  channel4: {
    gain: 1,
    panning: 0,
    auxSendGain: 0,
    auxIsPreGain: !1,
    isMuted: !1,
    isSoloed: !1
  }
}, ft = {
  orderAmongStrips: 0,
  displayName: "",
  colorIndex: 0
}, pt = {
  panning: 0,
  postGain: 1,
  isMuted: !1,
  isSoloed: !1
}, ht = {
  highPassCutoffFrequencyHz: 20,
  lowPassCutoffFrequencyHz: 2e4,
  isActive: !0
}, ei = {
  attackMs: 15,
  releaseMs: 100,
  makeupGainDb: 0,
  detectionModeIndex: 1,
  ratio: 2,
  thresholdDb: -10,
  isActive: !1
}, ti = {
  lowShelfFrequencyHz: 60,
  lowShelfGainDb: 0,
  lowMidFrequencyHz: 500,
  lowMidGainDb: 0,
  highMidFrequencyHz: 4800,
  highMidGainDb: 0,
  highShelfFrequencyHz: 12e3,
  highShelfGainDb: 0,
  isActive: !0
}, Xa = {
  displayParameters: ft,
  preGain: 1,
  trimFilter: ht,
  faderParameters: pt
}, Qa = {
  gain: 1
}, Ya = {
  displayParameters: ft,
  preGain: 0.39810699224472046,
  doesPhaseReverse: !1,
  trimFilter: ht,
  compressor: ei,
  eq: ti,
  auxSendsAreActive: !0,
  faderParameters: pt
}, Za = {
  displayParameters: ft,
  preGain: 1,
  trimFilter: ht,
  feedbackFactor: 0.30000001192092896,
  stepCount: 3,
  stepLengthIndex: 1,
  faderParameters: pt
}, _a = {
  displayParameters: ft,
  trimFilter: ht,
  compressor: ei,
  eq: ti,
  auxSendsAreActive: !0,
  faderParameters: pt
}, er = {
  positionX: 0,
  positionY: 0,
  doBypassInserts: !1,
  panning: 0,
  postGain: 1,
  limiterEnabled: !1,
  isMuted: !1
}, tr = {
  displayParameters: ft,
  preGain: 1,
  trimFilter: ht,
  roomSizeFactor: 0.800000011920929,
  preDelayTimeMs: 160,
  dampFactor: 0.10000000149011612,
  faderParameters: pt
}, nr = {}, ir = {}, ar = {
  ...h,
  displayName: "Note Splitter",
  channels: [
    {
      velocityModulation: 0,
      isMuted: !1
    },
    {
      velocityModulation: 0,
      isMuted: !1
    },
    {
      velocityModulation: 0,
      isMuted: !1
    }
  ],
  presetName: ""
}, rr = {
  ...h,
  displayName: "Panorama",
  leftFactor: 1,
  rightFactor: 1,
  leftPanning: -1,
  rightPanning: 1,
  isActive: !0,
  presetName: ""
}, or = {
  ...h,
  displayName: "Pulsar Delay",
  preDelayLeftTimeSemibreveIndex: 1,
  preDelayLeftTimeMs: 0,
  preDelayLeftPanning: -1,
  preDelayRightTimeSemibreveIndex: 3,
  preDelayRightTimeMs: 0,
  preDelayRightPanning: 1,
  feedbackDelayTimeSemibreveIndex: 1,
  feedbackDelayTimeMs: 0,
  lfoSpeedHz: 5,
  lfoModulationDepthMs: 0,
  feedbackFactor: 0.7,
  stereoCrossFactor: 1,
  filterMinHz: 20,
  filterMaxHz: 2e4,
  dryGain: 1,
  wetGain: 0.7,
  isActive: !0,
  presetName: ""
}, nt = {
  isActive: !0,
  panning: 0,
  gain: 1
}, At = {
  tuneSemitones: 0,
  tuneOctaves: 0,
  waveform: 0
}, sr = {
  ...h,
  displayName: "Pulverisateur",
  isActive: !0,
  microTuning: new v(),
  gain: 0.707946,
  glideTimeMs: 0,
  tuneSemitones: 0,
  playModeIndex: 2,
  oscillatorA: {
    channel: nt,
    oscillator: At
  },
  oscillatorB: {
    channel: {
      ...nt,
      isActive: !1
    },
    oscillator: At,
    hardSyncToOscillatorA: !1
  },
  oscillatorC: {
    channel: {
      ...nt,
      isActive: !1
    },
    oscillator: At,
    doesTrackKeyboard: !0
  },
  noise: {
    channel: {
      ...nt,
      isActive: !1
    },
    color: 1
  },
  audio: {
    channel: { ...nt, isActive: !1 },
    drive: 0
  },
  filter: {
    modeIndex: 1,
    cutoffFrequencyHz: 15500,
    resonance: 0,
    filterSpacing: 0,
    keyboardTrackingAmount: 0
  },
  lfo: {
    waveform: 0,
    rateIsSynced: !1,
    rateNormalized: 0.25,
    restartOnNote: !1,
    targetsOscillatorAPitch: !1,
    targetsOscillatorBPitch: !1,
    targetsOscillatorCPitch: !1,
    targetsFilterCutoff: !1,
    targetsPulseWidth: !1,
    modulationDepth: 0
  },
  filterEnvelope: {
    attackMs: 1,
    decayMs: 500,
    decayIsLooped: !1,
    sustainFactor: 0,
    releaseMs: 1,
    modulationDepth: 0
  },
  amplitudeEnvelope: {
    attackMs: 1,
    decayMs: 500,
    decayIsLooped: !1,
    sustainFactor: 0,
    releaseMs: 1
  },
  presetName: ""
}, gt = {
  thresholdDb: -10,
  ratio: 4,
  kneeDb: 3,
  attackMs: 5,
  releaseMs: 50,
  makeupGainDb: 0,
  isCompressorActive: !0,
  isMuted: !1,
  isSoloed: !1
}, cr = {
  ...h,
  displayName: "Quantum",
  gainDb: 0,
  spectrumModeIndex: 3,
  rmsWindowMs: 5,
  bands: [
    gt,
    gt,
    gt,
    gt
  ],
  splitFrequencyHz: [120, 2e3, 1e4],
  isActive: !0,
  presetName: ""
}, lr = {
  ...h,
  displayName: "Quasar",
  preDelayMs: 0,
  lowPassFrequencyHz: 2e4,
  highPassFrequencyHz: 20,
  filterSlopeIndex: 1,
  dryGain: 0.5011870265007019,
  wetGain: 0.5011870265007019,
  plateDecay: 0.75,
  plateDamp: 0,
  inputDiffusion: 0,
  tankDiffusion: 0,
  vibratoDepth: 0,
  vibratoFrequencyHz: 1,
  isActive: !0,
  presetName: ""
}, bt = (i, t) => Array.from({ length: i }, (e, n) => t(n)), ur = {
  gain: 1,
  panning: 0,
  mix: 1,
  mixMode: 1,
  isMuted: !1,
  isSoloed: !1
}, dr = {
  isOn: !1,
  isEnd: !1
}, On = {
  steps: Array.from({ length: 256 }, () => dr)
}, mr = {
  ...h,
  length: 16,
  groove: new v(),
  channelPatterns: bt(5, () => On),
  effectOrder: bt(7, (i) => i),
  effectPatterns: bt(7, () => On)
}, fr = {
  ...h,
  displayName: "Rasselbock",
  patternIndex: 0,
  isActive: !0,
  channelConfigs: bt(5, () => ur),
  shuffleConfig: {
    intervalIndex: 3,
    seed: 16777215,
    isMuted: !1,
    isSoloed: !1
  },
  speedConfig: {
    speedRatioIndex: 6,
    isMuted: !1,
    isSoloed: !1
  },
  stopConfig: {
    durationIndex: 5,
    doesSpinback: !1,
    isMuted: !1,
    isSoloed: !1
  },
  gateConfig: {
    intervalDurationIndex: 5,
    durationFactor: 0.5,
    isMuted: !1,
    isSoloed: !1
  },
  stutterConfig: {
    intervalDurationIndex: 5,
    scaleFactor: 1,
    pitchSemitones: 0,
    isMuted: !1,
    isSoloed: !1
  },
  scratchConfig: {
    rateBars: 4,
    modulationDepth: 1,
    modulationOffset: 1,
    modulationShapeIndex: 1,
    isMuted: !1,
    isSoloed: !1
  },
  reverseConfig: {
    isMuted: !1,
    isSoloed: !1
  },
  presetName: ""
}, pr = {
  ...h,
  displayName: "Ring Modulator",
  gain: 1,
  isActive: !0
}, hr = {
  sampleName: "",
  uploadStartTime: 0n
}, Hn = {
  dispersion: 0.5,
  vaporisation: 0.2,
  brightness: 0,
  metal: 0,
  separation: 1,
  harmonicsCount: 32,
  combFilterAmount: 0,
  combFilterRate: 0.15000000596046448,
  combFilterWidth: 0
}, yr = {
  ...h,
  displayName: "Space",
  isActive: !0,
  microTuning: new v(),
  gain: 0.70794,
  stereoDetuneShift: 0,
  tuneSemitones: 0,
  tuneASemitones: 0,
  tuneBSemitones: 0,
  glideMs: 0,
  mixAB: -1,
  lfoMixModulationDepth: 0,
  lfoGainModulationDepth: 0,
  lfoStereoDetuneShiftModulationDepth: 0,
  lfoPanningModulationDepth: 0,
  envelopeMixModulationDepth: 0,
  envelopeTuneModulationDepth: 0,
  envelopeLfoRateModulationDepth: 0,
  envelopeLfoAmountModulationDepth: 0,
  velocityGainModulationDepth: 0,
  velocityMixModulationDepth: 0,
  keyboardMixModulationDepth: 0,
  notePlayModeIndex: 3,
  modulationEnvelopeHasRelease: !1,
  lfo: {
    waveformIndex: 1,
    rateNormalized: 0.25,
    phaseOffset: 0,
    isSynced: !1,
    doesRetrigger: !0
  },
  amplitudeEnvelope: {
    isSynced: !1,
    attackTimeNormalized: 0,
    attackSlopeFactor: -0.5,
    decayTimeNormalized: 0.25,
    decaySlopeFactor: 0.5,
    decayIsLooped: !1,
    sustainFactor: 0.5,
    releaseTimeNormalized: 0.25,
    releaseSlopeFactor: 0.5
  },
  modulationEnvelope: {
    isSynced: !1,
    attackTimeNormalized: 0,
    attackSlopeFactor: -0.5,
    decayTimeNormalized: 0.25,
    decaySlopeFactor: 0.5,
    decayIsLooped: !1,
    sustainFactor: 0.5,
    releaseTimeNormalized: 0.25,
    releaseSlopeFactor: 0.5
  },
  soundA: Hn,
  soundB: Hn,
  presetName: ""
}, gr = {
  ...h,
  displayName: "Spitfire LABS",
  expression: 0,
  dynamics: 0,
  reverb: 0,
  release: 0,
  tightness: 0,
  vibrato: 0,
  simpleMix: 0,
  stereoPan: 0,
  adsrAttack: 0,
  adsrDecay: 0,
  adsrSustain: 0,
  adsrRelease: 0,
  globalGain: 0,
  globalPan: 0,
  globalTune: 0,
  stereoFlip: 0,
  stereoSpread: 0,
  variation: 0,
  delay: 0,
  amount: 0,
  distortion: 0,
  lushVerb: 0,
  pedalVol: 0,
  pedalDyn: 0,
  length: 0,
  timeMachine: 0,
  stretch: 0,
  softPedal: 0,
  response: 0,
  mallet: 0,
  stopMute: 0,
  direction: 0,
  lowPassFilter: 0,
  portamento: 0,
  generalPurpose1: 0,
  generalPurpose2: 0,
  generalPurpose3: 0,
  generalPurpose4: 0,
  generalPurpose5: 0,
  generalPurpose6: 0,
  generalPurpose7: 0,
  generalPurpose8: 0,
  generalPurpose9: 0,
  speed: 0,
  compression: 0,
  scale: 0,
  depth: 0,
  noiseFx: 0,
  grainSpeed: 0,
  state: new Uint8Array(0)
}, vr = {
  ...h,
  displayName: "Stereo Enhancer",
  isActive: !0,
  channelsAreInverted: !1,
  frequencyHz: 11e3,
  stereoWidth: 0.25,
  presetName: ""
}, wr = {
  ...h,
  displayName: "Chorus",
  delayTimeMs: 20,
  feedbackFactor: 0,
  lfoFrequencyHz: 0.33329999446868896,
  lfoModulationDepth: 1,
  spreadFactor: 0,
  isActive: !0,
  presetName: ""
}, br = {
  ...h,
  displayName: "Compressor",
  attackMs: 5,
  releaseMs: 25,
  makeupGainDb: 0,
  detectionModeIndex: 1,
  ratio: 0.4000000059604645,
  thresholdDb: -15,
  isActive: !0,
  presetName: ""
}, Tr = {
  ...h,
  displayName: "Crusher",
  preGain: 1,
  downsamplingFactor: 0,
  postGain: 1,
  bits: 8,
  mix: 1,
  isActive: !0,
  presetName: ""
}, xr = {
  ...h,
  displayName: "Delay",
  stepCount: 3,
  stepLengthIndex: 1,
  feedbackFactor: Math.fround(0.4),
  mix: Math.fround(0.2),
  isActive: !0,
  presetName: ""
}, kr = {
  ...h,
  displayName: "Flanger",
  delayTimeMs: 3,
  feedbackFactor: 1,
  lfoFrequencyHz: 0.03999999910593033,
  lfoModulationDepth: 1,
  isActive: !0,
  presetName: ""
}, Sr = {
  ...h,
  displayName: "Gate",
  attackMs: 10,
  releaseMs: 50,
  postGain: 1,
  isInverted: !1,
  holdMs: 100,
  thresholdGain: 0.6665999889373779,
  isActive: !0,
  presetName: ""
}, Dr = {
  ...h,
  displayName: "Parametric EQ",
  frequencyHz: 3600,
  bandwidthFactor: 0.15000000596046448,
  postGainDb: 0,
  isActive: !0,
  presetName: ""
}, Mr = {
  ...h,
  displayName: "Phaser",
  minFrequencyHz: 240,
  maxFrequencyHz: 3e3,
  feedbackFactor: 0.7,
  lfoFrequencyHz: 0.6,
  mix: 1,
  isActive: !0,
  presetName: ""
}, Ir = {
  ...h,
  displayName: "Pitch Delay",
  stepCount: 3,
  stepLengthIndex: 1,
  feedbackFactor: 0.6660000085830688,
  tuneFactor: 0.2,
  mix: 0.7,
  isActive: !0,
  presetName: ""
}, Nr = {
  ...h,
  displayName: "Reverb",
  roomSizeFactor: 0.8,
  preDelayTimeMs: 160,
  feedbackFactor: 0.6669999957084656,
  dampFactor: 0.1,
  mix: 0.2,
  isActive: !0,
  presetName: ""
}, Pr = {
  ...h,
  displayName: "Slope",
  filterModeIndex: 1,
  frequencyHz: 300,
  resonanceFactor: 0,
  bandWidthHz: 0,
  mix: 1,
  isActive: !0,
  presetName: ""
}, Fr = {
  ...h,
  displayName: "Stereo Detune",
  detuneSemitones: 0.25,
  delayTimeMs: 14,
  isActive: !0,
  presetName: ""
}, Ar = {
  ...h,
  displayName: "Tube",
  drive: 12,
  tone: 0,
  postGain: 1,
  isActive: !0,
  presetName: ""
}, Mt = {
  positionTicks: 0,
  durationTicks: 15360,
  collectionOffsetTicks: 0,
  loopOffsetTicks: 0,
  loopDurationTicks: 15360,
  isEnabled: !0,
  colorIndex: 0,
  displayName: ""
}, Er = {
  region: Mt,
  gain: 1,
  fadeInDurationTicks: 10,
  fadeInSlope: 0,
  fadeOutDurationTicks: 10,
  fadeOutSlope: 0,
  timestretchMode: 2,
  pitchShiftSemitones: 0
}, qr = {
  orderAmongTracks: 0,
  isEnabled: !0,
  groove: new v()
}, Jr = {}, Cr = {
  positionTicks: 0,
  value: 0,
  slope: 0,
  interpolation: 1,
  isSecond: !1
}, Lr = {
  region: Mt
}, Br = {
  orderAmongTracks: 0,
  isEnabled: !0
}, zr = {
  isEnabled: !0
}, Or = {
  positionTicks: 0,
  durationTicks: 960,
  pitch: 60,
  velocity: 0.699999988079071,
  doesSlide: !1
}, Hr = {}, Ur = {
  region: Mt
}, Gr = {
  orderAmongTracks: 0,
  isEnabled: !0,
  groove: new v()
}, $r = {
  region: Mt,
  patternIndex: 0,
  restart: !1
}, jr = {
  orderAmongTracks: 0,
  isEnabled: !0
}, Wr = {
  ...h,
  displayName: "Tiny Gain",
  gain: 1,
  isMuted: !1,
  isActive: !0
}, Vr = {
  notes: Array.from({ length: 16 }, () => !1)
}, Kr = {
  ...h,
  groove: new v(),
  steps: Array.from({ length: 16 }, () => Vr)
}, Rr = {
  ...h,
  displayName: "Tone Matrix",
  patternIndex: 0,
  microTuning: new v(),
  isActive: !0,
  presetName: ""
}, Xr = {
  ...h,
  displayName: "Waveshaper",
  preGain: 1,
  mix: 1,
  autoDrive: 0,
  attackMs: 5,
  releaseMs: 50,
  thresholdGain: 1,
  invertEnvelope: !1,
  finalSlope: 0,
  finalY: 1,
  isActive: !0,
  disableOversampling: !1,
  presetName: ""
}, Qr = {
  x: 0,
  y: 0,
  slope: 0
}, Yr = {
  audioDevice: ca,
  audioMerger: la,
  audioRegion: Er,
  audioSplitter: ua,
  audioTrack: qr,
  autoFilter: da,
  automationCollection: Jr,
  automationEvent: Cr,
  automationRegion: Lr,
  automationTrack: Br,
  bandSplitter: ma,
  bassline: ha,
  basslinePattern: pa,
  beatbox8: va,
  beatbox8Pattern: ga,
  beatbox9: Ta,
  beatbox9Pattern: ba,
  centroid: xa,
  centroidChannel: ka,
  config: Sa,
  crossfader: Da,
  curve: Ma,
  desktopAudioCable: Ia,
  desktopNoteCable: Na,
  exciter: Pa,
  gakki: Aa,
  genericVst3PluginBeta: Ea,
  graphicalEQ: qa,
  gravity: Ja,
  groove: Ca,
  heisenberg: La,
  helmholtz: Ba,
  kobolt: za,
  machiniste: $a,
  machinistePattern: Ua,
  matrixArpeggiator: Va,
  matrixArpeggiatorPattern: Wa,
  microTuningOctave: Ka,
  minimixer: Ra,
  mixerAux: Xa,
  mixerAuxRoute: Qa,
  mixerChannel: Ya,
  mixerDelayAux: Za,
  mixerGroup: _a,
  mixerMaster: er,
  mixerReverbAux: tr,
  mixerSideChainCable: nr,
  mixerStripGrouping: ir,
  note: Or,
  noteCollection: Hr,
  noteRegion: Ur,
  noteSplitter: ar,
  noteTrack: Gr,
  panorama: rr,
  patternRegion: $r,
  patternTrack: jr,
  pulsar: or,
  pulverisateur: sr,
  quantum: cr,
  quasar: lr,
  rasselbock: fr,
  rasselbockPattern: mr,
  ringModulator: pr,
  sample: hr,
  space: yr,
  spitfireLabsVst3Plugin: gr,
  stereoEnhancer: vr,
  stompboxChorus: wr,
  stompboxCompressor: br,
  stompboxCrusher: Tr,
  stompboxDelay: xr,
  stompboxFlanger: kr,
  stompboxGate: Sr,
  stompboxParametricEqualizer: Dr,
  stompboxPhaser: Mr,
  stompboxPitchDelay: Ir,
  stompboxReverb: Nr,
  stompboxSlope: Pr,
  stompboxStereoDetune: Fr,
  stompboxTube: Ar,
  tempoAutomationTrack: zr,
  tinyGain: Wr,
  tonematrix: Rr,
  tonematrixPattern: Kr,
  waveshaper: Xr,
  waveshaperAnchor: Qr
}, xn = (i) => {
  const t = Yr[i];
  if (t === void 0)
    throw new Error(`No defaults found for entity type ${i}`);
  return ni(
    i,
    $n[i],
    t
  );
}, ni = (i, t, e) => new t(
  Object.fromEntries(
    t.fields.list().map((n) => n.name === "id" && n.no === 1 ? ["id", crypto.randomUUID()] : [
      n.localName,
      ii(
        `${i}:${n.no}`,
        n,
        e
      )
    ])
  )
), ii = (i, t, e) => {
  const n = yi(i);
  switch (n.type) {
    case "object": {
      if (T(t.kind === "message", "field is not a message"), t.T === Et)
        return {};
      if (t.T === q)
        return new q();
      const a = t.repeated ? e : e == null ? void 0 : e[t.localName];
      return ni(
        i,
        t.T,
        a ?? g(`No default value for field ${t.localName}`)
      );
    }
    case "array":
      return T(t.repeated, "field is not repeated"), new Array(n.length).fill(void 0).map((a, r) => {
        var c;
        return t.kind === "message" && t.T === Et ? {} : ii(
          `${i}:[]`,
          t,
          ((c = e == null ? void 0 : e[t.localName]) == null ? void 0 : c[r]) ?? g(`No default value for field ${t.localName}[${r}]`)
        );
      });
    case "primitive":
      return t.kind === "message" && t.T === q ? new q() : typeof e != "object" ? e : e[t.localName] ?? g(`No default value for field ${t.localName}`);
  }
}, ai = (i) => pn(i), pn = (i) => {
  if (i instanceof Ze)
    return i.array.map((t) => pn(t));
  if (i instanceof $) {
    const t = {};
    return Object.entries(i.fields).forEach(([e, n]) => {
      n instanceof $ && Object.entries(n.fields).length === 0 || (t[e] = pn(n));
    }), t;
  }
  if (i instanceof G)
    return i.value instanceof v, i.value;
  throw new Error(`unknown value ${i} `);
}, rt = (i, t) => {
  const e = Ai(i.getType().name);
  ri(e, i, t);
}, ri = (i, t, e) => {
  t.getType().fields.list().forEach((n) => {
    if (!(n.localName in e) || e[n.localName] === void 0)
      return;
    const a = t;
    if (!n.repeated) {
      a[n.localName] = Un(
        i,
        n,
        a[n.localName],
        e[n.localName]
      );
      return;
    }
    const r = e[n.localName], c = a[n.localName];
    T(
      Array.isArray(r),
      `tried overriding repeated field ${n.localName} of message ${t.getType().typeName} with non-array field`
    ), T(
      r.length === c.length,
      `tried overriding repeated field ${n.localName} with array of different length`
    ), r.forEach((d, u) => {
      c[u] = Un(i, n, c[u], d);
    });
  });
}, Un = (i, t, e, n) => {
  switch (t.kind) {
    case "message": {
      if (t.T.name === q.name) {
        const a = n;
        return new q({
          fieldIndex: a.fieldIndex.slice(),
          entityId: a.entityId
        });
      }
      return ri(
        i,
        e,
        n
      ), e;
    }
    case "scalar":
      return n;
  }
}, hn = (i, t = !0) => {
  const e = xn(i.entityType);
  return rt(
    e,
    ai(i)
  ), t || (e.id = i.id), e;
}, Zr = (i, t) => {
  const e = hn(i);
  return t !== void 0 && rt(e, t), {
    modification: Dt(e),
    entityId: e.id
  };
}, _r = (i, t) => {
  const e = (n) => {
    if (n instanceof v)
      return t(n);
    if (n instanceof Array)
      return n.map((a) => e(a));
    if (typeof n == "object" && n != null) {
      const a = {};
      return Object.entries(n).forEach(([r, c]) => {
        a[r] = e(c);
      }), a;
    }
    return n;
  };
  return e(i);
}, eo = (...i) => {
  const t = i.map(
    (u) => u instanceof Ge ? { entity: u } : u
  ), e = /* @__PURE__ */ new Map();
  t.forEach(
    (u) => e.set(u.entity.id, crypto.randomUUID())
  );
  const n = /* @__PURE__ */ new Map();
  t.forEach((u) => n.set(u.entity.id, u));
  const a = /* @__PURE__ */ new Map(), r = [];
  [...n.values()].forEach(({ entity: u, overwrites: m }) => {
    const f = ai(u), y = xn(u.entityType);
    y.id = e.get(u.id) ?? g();
    const p = _r(f, (k) => {
      const S = e.get(k.entityId);
      return S !== void 0 ? (r.push([y.id, S]), new v(S, k.entityType, [
        ...k.fieldIndex
      ])) : k;
    });
    rt(y, p), m !== void 0 && rt(y, m), a.set(y.id, [u.entityType, y]);
  });
  const c = wn(r).reverse(), d = [...a.keys()].filter(
    (u) => !c.includes(u)
  );
  return {
    modifications: [...c, ...d].map((u) => {
      const [, m] = a.get(u) ?? g();
      return Dt(m);
    }),
    uuidMap: e
  };
}, ot = (i, t) => {
  if (t instanceof q && i._protoType !== "pointer")
    throw new Error(
      `Expected value of type ${i._protoType}, but got Pointer`
    );
  return oa(
    i.location,
    i._protoType,
    t instanceof v ? t.toPointerMessage() : t
  );
}, U = (i, t) => {
  i === "gakki" && (t = {
    ...t,
    // @ts-ignore
    soundfontId: t.soundfontId ?? "ce79731c-f100-4f54-9ccc-2d2c60269483"
  });
  const e = xn(i);
  return rt(e, t), {
    modification: Dt(e),
    entityId: e.id
  };
}, kn = (i) => sa(i), to = (i, t) => {
  const e = /* @__PURE__ */ new Set(), n = [];
  e.add(i), e.forEach((r) => {
    t.pointingTo.entities(r).get().forEach((c) => {
      n.push([c.id, r]), e.add(c.id);
    });
  });
  let a;
  return n.length > 0 ? a = wn(n) : a = [i], a.map((r) => kn(r));
}, no = ["positionX", "positionY", "displayName"], io = (i, t, e) => si(i, t, e !== void 0, e), oi = (i, t) => {
  if (i instanceof G)
    return ro(i, t);
  if (i instanceof Ze)
    return ao(i, t);
  if (i instanceof $)
    return si(i, t);
  throw new Error("unknown field type");
}, si = (i, t, e = !1, n) => (T(typeof t == "object", "msg is not an object"), Object.entries(i.fields).flatMap(([a, r]) => {
  if (e && no.includes(a))
    return [];
  if (e && a === "presetName" && r instanceof G && r._protoType === "string") {
    const c = n ?? "";
    return r.value !== c ? [ot(r, c)] : [];
  }
  return oi(
    r,
    t[a]
  );
})), ao = (i, t) => (T(t instanceof Array, "field is not an array"), T(i.array.length === t.length, "field has different lengths"), [...i.array.entries()].flatMap(
  ([e, n]) => oi(n, t[e])
)), ro = (i, t) => {
  if (T(typeof i.value == typeof t, "field has different types"), i.value instanceof v)
    return T(
      t instanceof q,
      "tried updating a field that's not a pointer field with a NexusLocation"
    ), i.value.equalsPointer(t) ? [] : [ot(i, t)];
  const e = bn(i._protoType, t);
  return Rn(i.value, e) ? [] : [ot(i, e)];
}, st = (i, t) => {
  if (i instanceof q) {
    t(i);
    return;
  }
  if (i instanceof Array) {
    i.forEach((e) => st(e, t));
    return;
  }
  if (typeof i == "object") {
    Object.values(i).forEach(
      (e) => st(e, t)
    );
    return;
  }
  if (!["string", "number", "boolean", "bigint"].includes(typeof i))
    throw new Error(`unknown type ${typeof i}: ${i}`);
}, oo = (i, t) => {
  const e = /* @__PURE__ */ new Map(), n = at(i.target) ?? g();
  e.set(n.id, t), e.set("", "");
  const a = i.relatives.map((r) => at(r) ?? g());
  return a.forEach((r) => {
    if (e.has(r.id))
      throw new Error("duplicate entity id");
    e.set(r.id, crypto.randomUUID());
  }), [n, ...a].forEach((r) => {
    r ?? (r = g("can't happen")), r.id = e.get(r.id) ?? g(), st(r, (c) => {
      e.has(c.entityId) || console.error("entity", r.constructor.name, "pointer", c), c.entityId = e.get(c.entityId) ?? g("preset contains pointer to unknown entity");
    });
  }), new xt({
    relatives: a.map((r) => Ye(r)),
    target: Ye(n)
  });
}, so = (i, t, e, n) => (t.preset = oo(t.preset, e.id), [
  // remove entities pointing to target entity, we'll create new ones in next step
  ...i.ofTypes(...t.entitiesToRemovePointingToMain).pointingTo.entities(e.id).get().map((a) => kn(a.id)),
  // build new entities pointing to target
  ...t.preset.relatives.map(
    (a) => Dt(a)
  ),
  // update existing target entity
  ...io(
    e,
    at(t.preset.target) ?? g(),
    n
  )
]), co = (i, t, e) => {
  const n = [], a = (e == null ? void 0 : e.displayName) ?? i.displayName ?? "Sample inserted via Nexus SDK", r = po(
    i,
    e,
    t.bpm
  ), c = mo(
    r,
    e
  ), [d, u] = lo(
    e == null ? void 0 : e.attachTo,
    a,
    t.nextTrackOrder,
    t.nextMixerChannelOrder
  );
  n.push(...u);
  const [m, f] = uo(
    i.name,
    c,
    r,
    d
  );
  return n.push(...f), { modifications: n, regionId: m };
}, lo = (i, t, e, n) => {
  const a = () => {
    const { modification: d, entityId: u } = U(
      "audioDevice",
      {
        displayName: t
      }
    );
    return [u, d];
  }, r = (d) => {
    const { modification: u, entityId: m } = U(
      "audioTrack",
      {
        player: new v(d, "audioTrack"),
        orderAmongTracks: e
      }
    );
    return [m, u];
  }, c = (d) => {
    const { modification: u, entityId: m } = U("mixerChannel", {
      displayParameters: {
        displayName: t,
        orderAmongStrips: n
      }
    }), { modification: f } = U(
      "desktopAudioCable",
      {
        fromSocket: v.fromSchemaPath(
          d,
          "/audioDevice/audioOutput"
        ),
        toSocket: v.fromSchemaPath(
          m,
          "/mixerChannel/audioInput"
        )
      }
    );
    return [u, f];
  };
  if (i === void 0) {
    const [d, u] = a(), [m, f] = r(d), y = c(d);
    return [m, [u, f, ...y]];
  }
  if (i.entityType === "audioDevice") {
    const [d, u] = r(i.id);
    return [d, [u]];
  }
  if (i.entityType === "audioTrack")
    return [i.id, []];
  throw new Error("Invalid attachTo type");
}, uo = (i, t, e, n) => {
  const { modification: a, entityId: r } = U("sample", {
    sampleName: i
  }), { modification: c, entityId: d } = U("automationCollection", {}), { modification: u } = U(
    "automationEvent",
    {
      collection: new v(d, "automationCollection"),
      positionTicks: 0,
      value: 0,
      interpolation: 2
    }
  ), { modification: m } = U(
    "automationEvent",
    {
      collection: new v(d, "automationCollection"),
      positionTicks: e,
      value: 1
    }
  ), { modification: f, entityId: y } = U("audioRegion", {
    track: new v(n, "audioTrack"),
    sample: new v(r, "sample"),
    playbackAutomationCollection: new v(
      d,
      "automationCollection"
    ),
    region: t
  });
  return [
    y,
    [
      a,
      c,
      u,
      m,
      f
    ]
  ];
}, mo = (i, t) => {
  var m, f;
  const e = ho(
    t,
    i
  ), n = ((m = t == null ? void 0 : t.region) == null ? void 0 : m.positionTicks) ?? 0;
  if (n < 0)
    throw new Error(
      `insertSample: region.positionTicks must be >= 0, got ${n}`
    );
  const a = ((f = t == null ? void 0 : t.region) == null ? void 0 : f.durationTicks) ?? Math.max(0, i - e);
  if (a < 0)
    throw new Error(
      `insertSample: region.durationTicks must be >= 0, got ${a}`
    );
  const r = t == null ? void 0 : t.loop;
  let c, d, u;
  if (r === void 0 || r === !1)
    d = 0, c = e, u = a;
  else if (r === !0)
    d = 0, c = e, u = Math.max(0, i - e);
  else {
    const y = r.startTicks ?? 0;
    if (c = y, d = y - e, d < 0)
      throw new Error(
        `insertSample: sample.offsetTicks (${e}) must be <= loop.startTicks (${y})`
      );
    u = r.durationTicks ?? Math.max(0, i - y);
  }
  return {
    positionTicks: n,
    durationTicks: a,
    collectionOffsetTicks: c,
    loopOffsetTicks: d,
    loopDurationTicks: u,
    isEnabled: !0,
    colorIndex: 0
  };
}, fo = (i, t, e) => {
  if (typeof t == "number")
    return t;
  if (t === "project") {
    if (e === void 0)
      throw new Error(
        'insertSample: sample.bpm = "project" requires a project bpm, but no Config entity with tempoBpm is available'
      );
    return e;
  }
  if (i !== void 0 && i > 0)
    return i;
  if (e !== void 0)
    return e;
  throw new Error(
    "insertSample: cannot resolve bpm — no sample.bpm override, no bpm on the sample, and no project bpm available"
  );
}, po = (i, t, e) => {
  var r, c;
  const n = (r = t == null ? void 0 : t.sample) == null ? void 0 : r.musicDurationTicks;
  if (n !== void 0)
    return n;
  const a = fo(
    i.bpm,
    (c = t == null ? void 0 : t.sample) == null ? void 0 : c.bpm,
    e
  );
  return vi(i.durationSeconds, a);
}, ho = (i, t) => {
  var a;
  const e = (a = i == null ? void 0 : i.sample) == null ? void 0 : a.offsetTicks;
  if (e !== void 0) {
    if (e < 0)
      throw new Error(
        `insertSample: sample.offsetTicks must be >= 0, got ${e}`
      );
    if (e > t)
      throw new Error(
        `insertSample: sample.offsetTicks (${e}) must be <= sample duration (${t})`
      );
    return e;
  }
  const n = i == null ? void 0 : i.loop;
  return n && typeof n == "object" && n.startTicks !== void 0 ? n.startTicks : 0;
}, yo = (i) => {
  i.relatives = go(
    i.relatives.map((e) => at(e) ?? g())
  ).map((e) => Ye(e));
  const t = Ii(i.target) ?? g("invalid preset");
  return {
    preset: i,
    entitiesToRemovePointingToMain: ci[t] ?? g("preset with unexpected main entity type")
  };
}, go = (i) => {
  const t = new Map(i.map((r) => [r.id, r])), e = [];
  i.forEach((r) => {
    st(r, (c) => {
      t.has(c.entityId) && e.push([r.id, c.entityId]);
    });
  });
  const n = wn(e).reverse().map((r) => t.get(r) ?? g()), a = i.filter((r) => !n.includes(r));
  return [...n, ...a];
}, ci = {
  // synthesizers
  bassline: ["basslinePattern", "groove"],
  pulverisateur: ["microTuningOctave"],
  heisenberg: ["microTuningOctave"],
  tonematrix: ["tonematrixPattern", "groove", "microTuningOctave"],
  space: ["microTuningOctave"],
  // drum machines
  machiniste: ["machinistePattern", "sample", "groove"],
  beatbox8: ["beatbox8Pattern", "groove"],
  beatbox9: ["beatbox9Pattern", "groove"],
  // pedals
  stompboxChorus: [],
  stompboxCompressor: [],
  stompboxCrusher: [],
  stompboxDelay: [],
  stompboxFlanger: [],
  stompboxGate: [],
  stompboxParametricEqualizer: [],
  stompboxPhaser: [],
  stompboxPitchDelay: [],
  stompboxReverb: [],
  stompboxSlope: [],
  stompboxStereoDetune: [],
  stompboxTube: [],
  // other effects
  quasar: [],
  rasselbock: ["rasselbockPattern"],
  pulsar: [],
  quantum: [],
  curve: [],
  graphicalEQ: [],
  gravity: [],
  autoFilter: [],
  waveshaper: ["waveshaperAnchor"],
  helmholtz: [],
  stereoEnhancer: [],
  exciter: [],
  ringModulator: void 0,
  panorama: [],
  tinyGain: void 0,
  // mixer
  minimixer: void 0,
  kobolt: void 0,
  centroid: void 0,
  crossfader: [],
  bandSplitter: [],
  audioSplitter: void 0,
  audioMerger: void 0,
  matrixArpeggiator: ["matrixArpeggiatorPattern", "groove"],
  noteSplitter: [],
  // no presets
  audioDevice: void 0,
  spitfireLabsVst3Plugin: void 0,
  mixerMaster: void 0,
  // no presets
  beatbox8Pattern: void 0,
  beatbox9Pattern: void 0,
  waveshaperAnchor: void 0,
  basslinePattern: void 0,
  tonematrixPattern: void 0,
  machinistePattern: void 0,
  rasselbockPattern: void 0,
  matrixArpeggiatorPattern: void 0,
  config: void 0,
  desktopAudioCable: void 0,
  desktopNoteCable: void 0,
  groove: void 0,
  microTuningOctave: void 0,
  sample: void 0,
  audioRegion: void 0,
  automationEvent: void 0,
  automationCollection: void 0,
  automationRegion: void 0,
  automationTrack: void 0,
  note: void 0,
  noteCollection: void 0,
  noteRegion: void 0,
  noteTrack: void 0,
  patternRegion: void 0,
  patternTrack: void 0,
  tempoAutomationTrack: void 0,
  mixerAux: void 0,
  mixerChannel: void 0,
  mixerAuxRoute: void 0,
  mixerDelayAux: void 0,
  mixerGroup: void 0,
  mixerReverbAux: void 0,
  mixerStripGrouping: void 0,
  mixerSideChainCable: void 0,
  audioTrack: void 0,
  centroidChannel: void 0,
  genericVst3PluginBeta: void 0,
  gakki: []
}, vo = (i, t) => {
  const e = hn(i, !1);
  let n;
  {
    const a = t.ofTypes(
      ...ci[i.entityType] ?? g()
    );
    n = [
      // select entities that point from/tom target entity
      ...a.pointingTo.entities(i.id).get(),
      ...a.pointedToBy.entities(i.id).get()
    ].map((r) => [
      // select entities that point from/to any relatives entity
      ...a.pointingTo.entities(r.id).get(),
      ...a.pointedToBy.entities(r.id).get(),
      r
    ]).flat().map((r) => hn(r, !1));
  }
  return wo([e, ...n]), new xt({
    relatives: n.map((a) => Ye(a)),
    target: Ye(e)
  });
}, wo = (i) => {
  const t = /* @__PURE__ */ new Map();
  t.set("", "");
  const e = (n) => (t.set(n, t.get(n) ?? crypto.randomUUID()), t.get(n) ?? g());
  i.forEach((n) => {
    const a = e(n.id);
    n.id = a, st(n, (r) => {
      r.entityId = e(r.entityId);
    });
  });
}, bo = (i) => {
  let t = !1;
  const e = {
    create: (n, a) => {
      if (t)
        throw new A("create");
      const { modification: r, entityId: c } = U(n, a);
      return i.applyModification(r, !0), i.query.mustGetEntity(c);
    },
    clone: (n, a) => {
      if (t)
        throw new A("clone");
      const { modification: r, entityId: c } = Zr(n, a);
      return i.applyModification(r, !0), i.query.mustGetEntity(c);
    },
    cloneLinked: (...n) => {
      if (t)
        throw new A("cloneLinked");
      const { modifications: a, uuidMap: r } = eo(...n);
      return a.forEach((c) => i.applyModification(c, !0)), n.map((c) => {
        const d = c instanceof Ge ? c.id : c.entity.id, u = r.get(d) ?? g();
        return i.query.mustGetEntity(u);
      });
    },
    update: (n, a) => {
      if (t)
        throw new A("update");
      if (a = bn(n._protoType, a), Rn(a, n.value))
        return;
      const r = ot(n, a);
      i.applyModification(r, !0);
    },
    tryUpdate: (n, a) => {
      if (t)
        throw new A("tryUpdate");
      const r = ot(n, a);
      return i.applyModification(r, !1) ?? void 0;
    },
    remove: (n) => {
      if (t)
        throw new A("remove");
      const a = n instanceof Ge ? n.id : n, r = kn(a);
      i.applyModification(r, !0);
    },
    removeWithDependencies: (n) => {
      if (t)
        throw new A("removeWithDependencies");
      const a = n instanceof Ge ? n.id : n;
      to(a, i.query).forEach((c) => i.applyModification(c, !0));
    },
    applyPresetTo: (n, a) => {
      if (t)
        throw new A("applyPreset");
      const r = yo(a.data);
      {
        const d = n.entityType;
        T(
          a.entityType === d,
          `attempted to apply preset for entity of type ${a.entityType} to entity of type ${d}`
        );
      }
      so(
        i.query,
        r,
        n,
        a._presetName
      ).forEach((d) => i.applyModification(d, !0));
    },
    createDeviceFromPreset: (n) => {
      if (t)
        throw new A("createDeviceFromPreset");
      const a = e.create(
        n.entityType,
        {}
      );
      return e.applyPresetTo(
        a,
        n
      ), a;
    },
    createPresetFor: (n) => {
      if (t)
        throw new A("createPresetFor");
      return vo(n, i.query);
    },
    insertSample: (n, a) => {
      var d;
      if (t)
        throw new A("insertSample");
      const { modifications: r, regionId: c } = co(
        n,
        {
          nextTrackOrder: To(i.query),
          nextMixerChannelOrder: xo(i.query),
          bpm: (d = i.query.ofTypes("config").getOne()) == null ? void 0 : d.fields.tempoBpm.value
        },
        a
      );
      return r.forEach((u) => i.applyModification(u, !0)), i.query.mustGetEntityAs(c, "audioRegion");
    },
    _addModification: (n) => {
      if (t)
        throw new A("_addModification");
      i.applyModification(n, !0);
    },
    send() {
      var n;
      if (t)
        throw new A("send");
      t = !0, (n = i.finish) == null || n.call(i);
    },
    entities: i.query
  };
  return e;
};
class A extends Error {
  constructor(t) {
    super(
      `Tried calling method ${t} on transaction that was already sent`
    );
  }
}
const To = (i) => i.ofTypes("audioTrack", "patternTrack", "noteTrack", "automationTrack").get().reduce(
  (t, e) => Math.max(t, e.fields.orderAmongTracks.value),
  -1
  // could be lower but doesn't matter
) + 1, xo = (i) => i.ofTypes(
  "mixerChannel",
  "mixerAux",
  "mixerDelayAux",
  "mixerGroup",
  "mixerReverbAux"
).get().reduce(
  (t, e) => Math.max(
    t,
    e.fields.displayParameters.fields.orderAmongStrips.value
  ),
  -1
) + 1;
var Ie, E, L, Ne, qe, lt, Re, ut, Xe, dt, mt, yn;
class li {
  constructor(t) {
    b(this, mt);
    /** Public field to query the entities in the document.
     *
     * ## Careful
     *
     * This method of querying the nexus document doesn't require a transaction lock. Any time a transaction
     * lock is taken, all previous query results should be considered out of date.
     *
     * ### Example incorrect usage
     *
     * The `await nexus.modify` awaits other transactions to finish before the callback is executed.
     * These transactions could be removing the output entity, and so by the type `t.update` is called,
     * the update transaction would be invalid, resulting in an exception:
     *
     * ```
     * const output = nexus.queryEntitiesWithoutLock.ofTypes("output").getOne() ?? ..
     * await nexus.modify(t => {
     *     // updating a field of an entity that was potentially removed during the await
     *     t.update(output.fields.name, "hello")
     * })
     * ```
     *
     * ### Example Correct usage
     *
     * It's better to use the `entities` field in the {@link TransactionBuilder} that is of the same type.
     * In the example above, better write
     *
     * ```
     * await nexus.modify(t => {
     *     const output = t.entities.ofTypes("output").getOne() ?? ..
     *     t.update(output.fields.name, "hello")
     * })
     * ```
     */
    o(this, "queryEntitiesWithoutLock");
    /** Manages subscriptions to nexus events, such as entity update & creation */
    o(this, "events");
    /**
     * This is a notifier that notifies of all modifications applied on the nexus
     * document, locally created or incoming from the backend.
     *
     * If the transaction is incoming `local` is `false`.
     */
    o(this, "onModification", new jn());
    /** Keeps track of all entities currently in the document. */
    b(this, Ie);
    /**
     * Lock that protects transactions. Only one transaction can be created at a time.
     */
    b(this, E, new wi());
    /** The connection to the backend */
    b(this, L);
    /** Validates any and all modifications */
    b(this, Ne);
    /**
     * This flag is set to true if {@link takeTransactions} is called. Before that, no transactions
     * from the backend are processed, and no transactions are allowed to be created from the frontend.
     *
     * This is to make sure that all `onCreate` callbacks are attached before any entity is created.
     */
    b(this, qe, !1);
    /**
     * This flag is set to true during {@link stop} to exit early if {@link stop} is called twice.
     */
    b(this, lt, !1);
    /**
     * This flag is set to true after {@link stop} is called and causes all further calls to {@link createTransaction} and {@link modify} to throw an error.
     */
    b(this, Re, !1);
    /**
     * The number of modifications applied to the document for incoming transactions before control
     * is yield to the scheduler.
     */
    b(this, ut);
    /** The delay between two gateway.synchronize() calls to sync with backend. */
    b(this, Xe);
    b(this, dt);
    D(this, Xe, (t == null ? void 0 : t.synchronizeEveryMs) ?? 16), D(this, ut, (t == null ? void 0 : t.incomingModificationsBatchSize) ?? 20), D(this, Ie, Tn({
      callbacks: {
        onStartPointingTo: (e, n) => {
          this.events._dispatchPointingTo(n, e);
        },
        onStopPointingTo: (e, n) => {
          this.events._dispatchStopPointingTo(n, e);
        },
        onUpdate: (e, n) => {
          this.events._dispatchUpdate(e, n);
        },
        onCreate: (e) => {
          this.events._dispatchCreate(e);
        },
        onDelete: (e) => {
          this.events._dispatchRemove(e);
        }
      }
    })), this.queryEntitiesWithoutLock = new St({
      documentState: l(this, Ie),
      documentLock: void 0
    }), this.events = new mn(l(this, Ie));
  }
  /**
   * This flag is set to true if {@link takeTransactions} is called. Before that, no transactions
   * from the backend are processed, and no transactions are allowed to be created from the frontend.
   * Attempting to call {@link createTransaction} before this is set to true will throw an error.
   *
   * This is to make sure that all `onCreate` callbacks are attached before any entity is created.
   */
  get transactionsAllowed() {
    return l(this, qe);
  }
  /**
   * Wait to acquire the document lock, then returns a {@link TransactionBuilder} that can be used to modify the nexus document
   * in a single transaction.
   *
   * To finish the transaction, call `send()`, which will release the lock, and send the modification to the backend. After this
   * method is called, no further methods can be called of the builder.
   *
   * The backend and other clients will only see changes after `send` is called, however locally, all changes are immediately applied.
   *
   * Note that every transaction by default is undoable, unless the flag in {@link TransactionOptions} is set to false.
   * @returns: {@link TransactionBuilder}
   */
  async createTransaction(t, e = !0) {
    T(l(this, qe), "Transactions not allowed yet"), T(l(this, L) !== void 0, "Gateway not initialized"), T(l(this, Ne) !== void 0, "Validator not initialized"), T(!l(this, Re), "Document stopped");
    const n = e ? await l(this, E).acquire() : void 0, a = [], r = new St({
      documentState: l(this, Ie),
      documentLock: l(this, E)
    }), c = (t == null ? void 0 : t.actionId) ?? Symbol();
    return bo({
      applyModification: (d, u) => {
        const m = x(this, mt, yn).call(this, d, {
          local: !0,
          throwIfInvalid: u,
          actionId: c
        });
        if (m !== void 0)
          return m;
        a.push(d);
      },
      finish: () => {
        var d;
        return a.length === 0 ? (n == null || n.release(), []) : ((d = l(this, L)) == null || d.send(new B({ modifications: a })), n == null || n.release(), a);
      },
      query: r
    });
  }
  /**
   * A helper method for small transactions.
   *
   * Writing
   * ```ts
   * const foo = await nexus.modify(t => fn(t))
   * ```
   *
   * is shorthand for
   *
   * ```ts
   * const t = await this.createTransaction(opts)
   * const foo = await fn(t)
   * t.send()
   * ```
   *
   * Note that every transaction by default is undo-able, unless the flag in {@link TransactionOptions} is set to false.
   */
  async modify(t, e) {
    const n = await this.createTransaction(e), a = await t(n);
    return n.send(), a;
  }
  /**
   * This function connects the document with the backend and starts syncing its state.
   * The returned promise resolves once the document has synced up with the backend,
   * and allows creating local transaction using {@link createTransaction}.
   *
   * It should be called exactly once when all `onCreate` callbacks are registered.
   */
  async takeTransactions(t) {
    if (l(this, qe))
      throw new Error("called `takeTransactions()` twice");
    D(this, Ne, (t == null ? void 0 : t.validator) ?? Zn()), D(this, L, (t == null ? void 0 : t.gateway) ?? Yn());
    let e;
    l(this, L).blocked.subscribe(async (m) => {
      m ? e === void 0 && (e = await l(this, E).acquire()) : (e == null || e.release(), e = void 0);
    });
    const n = await l(this, E).acquire();
    D(this, qe, !0);
    const {
      promise: a,
      resolve: r
    } = Promise.withResolvers(), c = qt(async () => {
      const m = (l(this, L) ?? g()).synchronize();
      if (m === "done")
        throw new Error(
          "Gateway returned 'done' before returning the initial transaction."
        );
      if (m.length === 0)
        return;
      const f = m.flatMap((y) => y.modifications).length;
      await this._applyIncomingTransactions(m, {
        _takeTransactionLock: !1
      }), c.terminate(), r(f > 0);
    }, l(this, Xe)), d = await a;
    if ((t == null ? void 0 : t.templateTransaction) !== void 0)
      if (d)
        console.error(
          "Could not apply template transaction to document: Project state from backend is not empty."
        );
      else {
        const m = await t.templateTransaction, f = await this.createTransaction(void 0, !1);
        m.modifications.forEach((y) => f._addModification(y)), f.send();
      }
    n.release();
    const u = qt(async (m) => {
      if (m.aborted)
        return;
      const f = await l(this, E).acquire(), y = (l(this, L) ?? g()).synchronize();
      if (y === "done") {
        u.terminate(), f.release();
        return;
      }
      await this._applyIncomingTransactions(y, { _takeTransactionLock: !1 }), f.release();
    }, l(this, Xe));
    D(this, dt, () => u.terminate());
  }
  /** Apply a transaction that doesn't originate from this document. Yields to the browser
   * scheduler every few modifications applied to make sure we don't block the main thread
   * for too long for big transactions. Throws if the transaction lock isn't taken.
   *
   */
  async _applyIncomingTransactions(t, e) {
    T(l(this, L) !== void 0, "Gateway not initialized"), T(l(this, Ne) !== void 0, "Validator not initialized"), T(!l(this, Re), "Document stopped");
    const n = (e == null ? void 0 : e._takeTransactionLock) ?? !0 ? await l(this, E).acquire() : T(
      l(this, E).locked,
      "if takeTransactionLock is false, the lock must be held."
    );
    for (const a of t)
      for (const [r, c] of a.modifications.entries())
        x(this, mt, yn).call(this, c, {
          local: !1,
          throwIfInvalid: !0
        }), r % l(this, ut) === 0 && await vn(0);
    n == null || n.release();
  }
  /** For debugging purposes */
  getStats() {
    return l(this, Ie).getStats();
  }
  /** Stop the document from syncing. This will have the following effect:
   * First, all pending `modify` and `createTransaction` calls will finish.
   * The modifications they create will be synced with the backend, and the document is locked down.
   *
   * After this, calling `modify` or `createTransaction` will throw an error. The only property that
   * can still be accessed is `queryEntities`. The document can be thrown away safely after this.
   */
  async terminate() {
    var t, e, n;
    if (!l(this, qe))
      throw new Error("Can't stop a document that hasn't started yet.");
    l(this, lt) || (D(this, lt, !0), await l(this, E).acquire(), (t = l(this, dt)) == null || t.call(this), l(this, E).acquire = () => g(
      "invariant violated: cannot get transaction lock after document stopped"
    ), D(this, Re, !0), await ((e = l(this, L)) == null ? void 0 : e.terminate()), (n = l(this, Ne)) == null || n.terminate(), this.events._clear());
  }
}
Ie = new WeakMap(), E = new WeakMap(), L = new WeakMap(), Ne = new WeakMap(), qe = new WeakMap(), lt = new WeakMap(), Re = new WeakMap(), ut = new WeakMap(), Xe = new WeakMap(), dt = new WeakMap(), mt = new WeakSet(), /**
 * Apply a modification to the document after validating it, and dispatch events to
 * {@link onModification} listeners. Throws if the modification is invalid, unless `throwIfInvalid` is false;
 * in that case, the function returns `string` on error and doesn't change the document state. If application
 * succeeds, always returns undefined.
 */
yn = function(t, {
  local: e = !0,
  throwIfInvalid: n = !0,
  actionId: a
}) {
  T(l(this, Ne) !== void 0, "Validator not initialized"), T(
    l(this, E).locked,
    "tried applying modification without lock"
  );
  const r = l(this, Ne).validate(t);
  if (r !== void 0) {
    if (n)
      throw new Error(`modification failed validation: ${r}`);
    return r;
  }
  l(this, Ie).applyModification(t), this.onModification.notify({ modification: t, local: e, actionId: a });
};
const ko = {
  typeName: "audiotool.document.v1.DocumentService",
  methods: {
    /**
     * Attach will provide a stream of modifications to the document.
     * for the Document will be sent to the respone.
     *
     *  The first AttachResponse contains a single transaction that reflects the
     * "current" document state; it's empty if the document is new, or contains
     * a list of `Create` modifications for all entities in the document otherwise (in topological order
     * so all pointers are always valid).
     *
     * There is no need to combine Attach and GetEntities.
     *
     * @generated from rpc audiotool.document.v1.DocumentService.Attach
     */
    attach: {
      name: "Attach",
      I: $t,
      O: jt,
      kind: F.ServerStreaming
    },
    /**
     * ApplyTransactions is a non-streaming version of `Modify`. It takes multiple
     * transactions per message as input, acts as though `Modify` was called with each transaction
     * in a separate message, collects the errors, and returns them in a single message.
     *
     * This method exists because gRPC-web doesn't support client -> server streaming. To guarantee
     * the order of transactions across method calls, each method call's reply must be awaited before
     * sending the next. To keep transaction throughput reasonably high, this method allows batching
     * multiple transactions in a single call.
     *
     * @generated from rpc audiotool.document.v1.DocumentService.ApplyTransactions
     */
    applyTransactions: {
      name: "ApplyTransactions",
      I: Vt,
      O: Kt,
      kind: F.Unary
    },
    /**
     * A single shot of entities in time.
     *
     * @generated from rpc audiotool.document.v1.DocumentService.GetEntities
     */
    getEntities: {
      name: "GetEntities",
      I: Ut,
      O: Gt,
      kind: F.Unary
    },
    /**
     * GetWasm will return the wasm binary for validation.
     *
     * @generated from rpc audiotool.document.v1.DocumentService.GetWasm
     */
    getWasm: {
      name: "GetWasm",
      I: Yt,
      O: Zt,
      kind: F.Unary
    },
    /**
     * GetWasmExec will return the wasm execution script (go).
     *
     * @generated from rpc audiotool.document.v1.DocumentService.GetWasmExec
     */
    getWasmExec: {
      name: "GetWasmExec",
      I: _t,
      O: en,
      kind: F.Unary
    },
    /**
     * Ping will return a response as fast as the server could process the request.
     *
     * The client_id is used to identify the client.
     * The last_ping_ms is the time in ms since of the last ping duration and will be used for the
     * GetClientStats information.
     *
     * @generated from rpc audiotool.document.v1.DocumentService.Ping
     */
    ping: {
      name: "Ping",
      I: tn,
      O: nn,
      kind: F.Unary
    },
    /**
     * GetClientStats will return the stats of all clients.
     *
     * @generated from rpc audiotool.document.v1.DocumentService.GetClientStats
     */
    getClientStats: {
      name: "GetClientStats",
      I: an,
      O: rn,
      kind: F.Unary
    },
    /**
     * Returns the version of the service implementing.
     *
     * @generated from rpc audiotool.document.v1.DocumentService.GetVersion
     */
    getVersion: {
      name: "GetVersion",
      I: sn,
      O: cn,
      kind: F.Unary
    },
    /**
     * Returns the current time in seconds since the Unix epoch of the server.
     *
     * @generated from rpc audiotool.document.v1.DocumentService.GetTime
     */
    getTime: {
      name: "GetTime",
      I: ln,
      O: un,
      kind: F.Unary
    },
    /**
     * AttachMetadata is used to receive metadata of the clients.
     *
     * Metadata contains independent messages which can be used to sync information about client
     * states.
     * Some examples are:
     *
     * - cursor position
     * - zoom level
     * - play/pause/transport
     * - ...
     *
     * @generated from rpc audiotool.document.v1.DocumentService.AttachMetadata
     */
    attachMetadata: {
      name: "AttachMetadata",
      I: Ct,
      O: Lt,
      kind: F.ServerStreaming
    },
    /**
     * PutMetadata is used to send metadata of the client to the document.
     *
     * @generated from rpc audiotool.document.v1.DocumentService.PutMetadata
     */
    putMetadata: {
      name: "PutMetadata",
      I: Bt,
      O: zt,
      kind: F.Unary
    },
    /**
     * GetCommitIndexRange returns the minimum and maximum commit_index currently held by the server
     * for a given document.
     *
     * @generated from rpc audiotool.document.v1.DocumentService.GetCommitIndexRange
     */
    getCommitIndexRange: {
      name: "GetCommitIndexRange",
      I: Ot,
      O: Ht,
      kind: F.Unary
    },
    /**
     * RenderAudio starts rendering the project to audio files and caching them on the
     * hermes instance for download via HTTP.
     *
     * Returns immediately with an [audiotool.longrunning.v1.Operation] where `done` is false.
     * Poll [audiotool.longrunning.v1.OperationService.GetOperation] (or subscribe to operation
     * updates via your client stack) until `done` is true. On success, `response` unpacks to
     * [RenderAudioResponse]. Metadata, when set, uses [RenderAudioInfo].
     *
     * @generated from rpc audiotool.document.v1.DocumentService.RenderAudio
     */
    renderAudio: {
      name: "RenderAudio",
      I: dn,
      O: xi,
      kind: F.Unary
    }
  }
}, ui = (i, t, e, n, a) => ({
  connected: t,
  createTransaction: () => i.createTransaction(),
  modify: (r) => i.modify(r),
  queryEntities: i.queryEntitiesWithoutLock,
  events: i.events,
  start: async () => i.takeTransactions({ validator: e, gateway: n }),
  stop: async () => i.terminate(),
  dawUrl: a
}), qo = async (i) => {
  i != null && i.wasm && ki(i.wasm);
  const t = (i == null ? void 0 : i.validated) ?? !0 ? await Xn() : Zn(), e = Yn(), n = new li(), a = new Fe(!0), r = ui(
    n,
    a,
    t,
    e,
    ""
    // offline documents have no project URL
  );
  return await r.start(), r;
}, Jo = async (i, t, e, n) => {
  var S;
  let a = Xn();
  const r = await i.openSession({
    projectName: t
  });
  if (r instanceof Error)
    throw new Error("Couldn't open session", { cause: r });
  const c = Ji(
    ko,
    await n.createTransport({
      baseUrl: ((S = r.session) == null ? void 0 : S.documentServiceUrl) ?? g("backend returned no document service url"),
      useBinaryFormat: !0,
      getToken: e
    })
  ), d = await a, u = await Di(), m = Wi(
    c,
    () => u,
    t
  ), f = new li(), y = new Fe(!1), k = `https://beta.audiotool.com/studio?project=${t.replace(/^projects\//, "")}`;
  return m.blocked.subscribe((N) => y.setValue(!N), !0), ui(f, y, d, m, k);
};
export {
  Ei as K,
  Ge as N,
  xt as P,
  Ji as a,
  Ii as b,
  qo as c,
  Jo as d,
  ki as i
};
