// @audiotool/nexus/dist/ticks-B57EbInS.js
var v = (r) => {
  throw TypeError(r);
};
var E = (r, e, t) => e.has(r) || v("Cannot " + t);
var s = (r, e, t) => (E(r, e, "read from private field"), t ? t.call(r) : e.get(r)), n = (r, e, t) => e.has(r) ? v("Cannot add the same private member more than once") : e instanceof WeakSet ? e.add(r) : e.set(r, t), l = (r, e, t, i) => (E(r, e, "write to private field"), i ? i.call(r, t) : e.set(r, t), t);
var f;
class T {
  constructor() {
    n(this, f, /* @__PURE__ */ new Set());
  }
  // A set allows us to remove while iterating
  subscribe(e) {
    return s(this, f).add(e), { terminate: () => s(this, f).delete(e) };
  }
  notify(e) {
    s(this, f).forEach((t) => t(e));
  }
  terminate() {
    s(this, f).clear();
  }
}
f = new WeakMap();
var d, o, g;
class A {
  constructor(e) {
    n(this, d, new T());
    n(this, o);
    n(this, g, k(s(this, d)));
    l(this, o, e);
  }
  getValue() {
    return s(this, g).call(this), s(this, o);
  }
  setValue(e) {
    s(this, o) !== e && (l(this, o, e), s(this, d).notify(e));
  }
  subscribe(e, t = !1) {
    return t && e(s(this, o)), s(this, d).subscribe(e);
  }
  terminate() {
    s(this, d).terminate();
  }
}
d = new WeakMap(), o = new WeakMap(), g = new WeakMap();
var c, y, h;
class B {
  constructor(e) {
    n(this, c, new T());
    n(this, y, k(s(this, c)));
    n(this, h);
    l(this, h, e ?? /* @__PURE__ */ new Map());
  }
  /** Subscribe to updates of the set */
  subscribe(e, t = !1) {
    return t && e(s(this, h)), s(this, c).subscribe(() => e(s(this, h)));
  }
  /** Get the underlying map. Readonly because changing it won't trigger updates. */
  getValue() {
    return s(this, y).call(this), s(this, h);
  }
  set(e, t) {
    return s(this, h).set(e, t), s(this, c).notify(), this;
  }
  delete(e) {
    const t = s(this, h).delete(e);
    return t && s(this, c).notify(), t;
  }
  clear() {
    const e = s(this, h).size > 0;
    s(this, h).clear(), e && s(this, c).notify();
  }
}
c = new WeakMap(), y = new WeakMap(), h = new WeakMap();
const k = (r) => () => {
};
var m, w, b;
class _ {
  /** if `warnAfterMs` is set, the lock will emit a warning if a call to `lock.acquire()
   * tok more than `warnAfterMs` milliseconds.
   */
  constructor(e) {
    n(this, m, !1);
    n(this, w, []);
    n(this, b);
    l(this, b, (e == null ? void 0 : e.warnAfterMs) ?? void 0);
  }
  /**
   * Wait until no other async thread holds a lock, then returns a lock.
   *
   * Once the lock is held, all other threads that call `acquire()` will have to wait until
   * the lock is released.
   *
   *
   * Release the lock with `lock.release()`.
   *
   * Example:
   * ```ts
   * const lock = new AsyncLock()
   * ...
   * const l = await lock.acquire()
   * // do stuff
   * l.release()
   * ```
   *
   */
  async acquire() {
    let e = () => {
    };
    s(this, b) !== void 0 && (e = N(
      s(this, b),
      `Waited for lock.acquire() for more than ${s(this, b)} ms, deadlock?`
    )), s(this, m) && await new Promise((i) => s(this, w).push(i)), e(), l(this, m, !0);
    let t = !1;
    return {
      release: () => {
        var i;
        if (t)
          throw new Error("Lock already released");
        t = !0, s(this, w).length > 0 ? (i = s(this, w).shift()) == null || i() : l(this, m, !1);
      }
    };
  }
  /**
   * Execute a function after acquiring a lock, and release the lock after the function is done.
   *
   * Example:
   * ```ts
   * const v = await lock.runAcquired(() => {
   *  // do something
   *  return 42
   * })
   * ```
   *
   * This function is **safe against exceptions**. If the function throws an exception, the lock
   * is released before the exception is thrown.
   *
   */
  async runAcquired(e) {
    const t = await this.acquire();
    let i;
    try {
      return i = await e(), t.release(), i;
    } catch (u) {
      throw t.release(), u;
    }
  }
  /** Weather the lock is currently taken. Because javascript is single-threaded, it is safe
   * to do e.g.:
   * ```ts
   * if (!lock.locked) {
   *   lock.acquire()
   *   // do something
   *   lock.release()
   * }
   *
   * ```
   *
   * without `await`ing the `lock.acquire()`.
   */
  get locked() {
    return s(this, m);
  }
}
m = new WeakMap(), w = new WeakMap(), b = new WeakMap();
const N = (r, e) => {
  const t = setTimeout(() => {
    console.warn(e), console.trace("Waited here:");
  }, r);
  return () => clearTimeout(t);
}, S = Symbol();
var a;
class q {
  constructor() {
    /** Maps the hash of a key to both key and value */
    n(this, a, /* @__PURE__ */ new Map());
  }
  clear() {
    s(this, a).clear();
  }
  delete(e) {
    const t = e[S];
    return s(this, a).delete(t);
  }
  forEach(e, t) {
    s(this, a).forEach(([i, u]) => {
      e(u, i, this);
    }, t);
  }
  entries() {
    return s(this, a).entries().map(([e, [t, i]]) => [t, i]);
  }
  get(e) {
    var t;
    return (t = s(this, a).get(e[S])) == null ? void 0 : t[1];
  }
  has(e) {
    return s(this, a).has(e[S]);
  }
  keys() {
    return s(this, a).values().map(([e, t]) => e);
  }
  set(e, t) {
    const i = e[S];
    return s(this, a).set(i, [e, t]), this;
  }
  values() {
    return s(this, a).values().map(([e, t]) => t);
  }
  get size() {
    return s(this, a).size;
  }
  [Symbol.iterator]() {
    return this.entries();
  }
  get [Symbol.toStringTag]() {
    return "StringHashMap";
  }
}
a = new WeakMap();
const M = 15360, p = {
  /** How many ticks pass in "1 whole note" or 4x1/4th notes in a 4/4th beat, independent of tempo. */
  SemiBreve: M,
  /** How many ticks pass in 1 quarter note in a 4/4th bar, independent of tempo. */
  Beat: 3840,
  /** How many ticks pass in 1/16th note in a 4/4th bar, independent of tempo. */
  SemiQuaver: 960,
  /**
   * Number of ticks in `n` bars at the given time signature.
   *
   * `signature` is a string of the form `"X/Y"` where `X` is the number of
   * beats per bar and `Y` is the note value of one beat (4 = quarter note,
   * 8 = eighth note, etc.). Defaults to `"4/4"`.
   *
   * Examples:
   * - `Ticks.Bars(3)`        → 3 bars in 4/4 = `3 * SemiBreve` = 46080
   * - `Ticks.Bars(2, "3/4")` → 2 bars in 3/4 = `2 * 3 * Beat`  = 23040
   * - `Ticks.Bars(2, "6/8")` → 2 bars in 6/8 = `2 * 6 * (SemiBreve / 8)` = 23040
   */
  Bars(r, e = "4/4") {
    const t = e.split("/");
    if (t.length !== 2)
      throw new Error(`Invalid time signature: "${e}" (expected "X/Y")`);
    const i = Number(t[0]), u = Number(t[1]);
    if (!Number.isFinite(i) || !Number.isFinite(u) || i <= 0 || u <= 0)
      throw new Error(`Invalid time signature: "${e}" (expected "X/Y")`);
    return r * i * (M / u);
  }
}, P = (r, e) => {
  const t = e / 60, i = r * t;
  return Math.round(i * p.Beat);
}, I = (r, e) => {
  const t = e / 60;
  return r / p.Beat / t;
};
export {
  _ as A,
  q as H,
  B as M,
  T as N,
  p as T,
  A as V,
  S as h,
  P as s,
  I as t
};
