// @audiotool/nexus/dist/utils.js
var o = (s) => {
  throw TypeError(s);
};
var n = (s, t, i) => t.has(s) || o("Cannot " + i);
var e = (s, t, i) => (n(s, t, "read from private field"), i ? i.call(s) : t.get(s)), h = (s, t, i) => t.has(s) ? o("Cannot add the same private member more than once") : t instanceof WeakSet ? t.add(s) : t.set(s, i);
import { N as c } from "./ticks-B57EbInS.js";
import { A as p, H as g, M as T, T as A, V as N, s as b, t as k } from "./ticks-B57EbInS.js";
import { t as x } from "./lang-K-8hAzE4.js";
var a, r;
class l {
  constructor(t) {
    h(this, a, new c());
    h(this, r, /* @__PURE__ */ new Set());
    t.forEach((i) => e(this, r).add(i));
  }
  getValue() {
    return Array.from(e(this, r));
  }
  replace(t) {
    e(this, r).clear(), t.forEach((i) => e(this, r).add(i)), e(this, a).notify(this.getValue());
  }
  add(t) {
    e(this, r).has(t) || (e(this, r).add(t), e(this, a).notify(this.getValue()));
  }
  delete(t) {
    e(this, r).has(t) && (e(this, r).delete(t), e(this, a).notify(this.getValue()));
  }
  has(t) {
    return e(this, r).has(t);
  }
  clear() {
    e(this, r).clear(), e(this, a).notify(this.getValue());
  }
  subscribe(t, i = !1) {
    return i && t(this.getValue()), e(this, a).subscribe(t);
  }
  terminate() {
    e(this, a).terminate();
  }
}
a = new WeakMap(), r = new WeakMap();
const y = (s, t) => Array.from({ length: s }, (i, f) => t(f));
export {
  p as AsyncLock,
  g as HashMap,
  T as MapValueNotifier,
  c as Notifier,
  l as SetNotifier,
  A as Ticks,
  N as ValueNotifier,
  y as createTypedArray,
  b as secondsToTicks,
  x as throw_,
  k as ticksToSeconds
};
